<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>invasimap â€¢ invasimap</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="â€image/svg+xmlâ€" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/PT_Sans-0.4.10/font.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="invasimap">
<meta name="description" content="A Novel Framework to visualise trait dispersion and assess species invasiveness or site invasibility">
<meta property="og:description" content="A Novel Framework to visualise trait dispersion and assess species invasiveness or site invasibility">
<meta property="og:image" content="https://macsands.github.io/invasimap/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">invasimap</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.9</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Articles</h6></li>
    <li><a class="dropdown-item" href="articles/getting-started.html">getting-started</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/macSands/invasimap" aria-label="GitHub repository"><span class="fa fab fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="invasimap">invasimap<a class="anchor" aria-label="anchor" href="#invasimap"></a>
</h1>
</div>
<div class="section level2">
<h2 id="a-novel-framework-to-visualise-trait-dispersion-and-assess-species-invasiveness-or-site-invasibility">A Novel Framework to visualise trait dispersion and assess species invasiveness or site invasibility<a class="anchor" aria-label="anchor" href="#a-novel-framework-to-visualise-trait-dispersion-and-assess-species-invasiveness-or-site-invasibility"></a>
</h2>
<!-- badges: start -->

<hr>
</div>
<div class="section level2">
<h2 id="id_1-introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#id_1-introduction"></a>
</h2>
<p>Biological invasions threaten global biodiversity. Invasive alien species (IAS) can expand rapidly and transform ecosystems. Because invasion dynamics arise from multiple drivers, from species interactions and traits to environmental gradients, we need rigorous, reproducible workflows to diagnose how species establish and spread. Recent theory, including trait-mediated ecological networks and invasion fitness, offers a coherent basis for integrating traits, environment, and biotic interactions (Hui et al., 2016; Hui et al., 2021). In the <code>invasimap</code> framework, <strong>invasion fitness</strong> is the <strong>low-density per-capita growth rate</strong> of an invader in a resident community. The low-density condition reflects a realistic introduction stage where the invader is rare and not self-limited, so establishment potential depends on environmental suitability and interspecific interactions. Positive values indicate growth from rarity; negative values indicate likely exclusion. This aligns with mutual invasibility and adaptive-dynamics invasion criteria.</p>
<div class="section level4">
<h4 id="formal-model-and-notation">Formal model and notation<a class="anchor" aria-label="anchor" href="#formal-model-and-notation"></a>
</h4>
<p>We compute <strong>invasion fitness</strong> for invader <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math> as:<br><a id="def-lambda"></a> <math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>ğ›Œ</mi><mrow><mi>ğ¢</mi><mi>ğ¬</mi></mrow></msub><mspace width="0.278em"></mspace><mo mathvariant="bold">=</mo><mspace width="0.278em"></mspace><msub><mi>ğ«</mi><mrow><mi>ğ¢</mi><mi>ğ¬</mi></mrow></msub><mspace width="0.278em"></mspace><mo mathvariant="bold">âˆ’</mo><mspace width="0.278em"></mspace><msubsup><mi>ğ‚</mi><mrow><mi>ğ¢</mi><mi>ğ¬</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup></mrow><annotation encoding="application/x-tex">
\mathbf{\lambda_{is} \;=\; r_{is} \;-\; C^{\mathrm{(raw)}}_{is}}
</annotation></semantics></math></p>
<p>where:</p>
<ul>
<li>Predicted <strong>growth potential</strong> (intrinsic growth or abundance proxy for invader <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math> from the trait-environment model) is:Â Â <a id="def-r"></a><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>r</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">r_{is}</annotation></semantics></math> Â </li>
<li>
<strong>Total raw competitive penalty</strong> is: <a id="def-Craw"></a> <math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><mspace width="0.278em"></mspace><mo>=</mo><mspace width="0.278em"></mspace><munder><mo>âˆ‘</mo><mrow><mi>j</mi><mo>â‰ </mo><mi>i</mi></mrow></munder><msub><mi>I</mi><mrow><mi>i</mi><mi>j</mi><mi>s</mi></mrow></msub></mrow><annotation encoding="application/x-tex">
C^{\mathrm{(raw)}}_{is} \;=\; \sum_{j \neq i} I_{ijs}
</annotation></semantics></math>
</li>
<li>
<strong>Impact tensor</strong> (resident <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math> on invader <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> in site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>) is: <a id="def-I"></a> <math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>I</mi><mrow><mi>i</mi><mi>j</mi><mi>s</mi></mrow></msub><mspace width="0.278em"></mspace><mo>=</mo><mspace width="0.278em"></mspace><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mspace width="0.278em"></mspace><msub><mi>K</mi><mi>e</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><mo>;</mo><msub><mi>Ïƒ</mi><mi>e</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mspace width="0.278em"></mspace><msubsup><mi>N</mi><mrow><mi>j</mi><mi>s</mi></mrow><mo>*</mo></msubsup><mi>.</mi></mrow><annotation encoding="application/x-tex">
I_{ijs} \;=\; \alpha_{ij} \; K_e(\Delta_{js}; \sigma_e) \; N^{*}_{js}.
</annotation></semantics></math>
</li>
<li>
<strong>Competition kernel in trait space</strong> is: <a id="def-alpha"></a> <math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mspace width="0.278em"></mspace><mo>=</mo><mspace width="0.278em"></mspace><mo>exp</mo><mspace width="-0.167em"></mspace><mrow><mo stretchy="true" form="prefix">(</mo><mo>âˆ’</mo><mfrac><msubsup><mi>d</mi><mrow><mi>i</mi><mi>j</mi></mrow><mn>2</mn></msubsup><mrow><mn>2</mn><msubsup><mi>Ïƒ</mi><mi>t</mi><mn>2</mn></msubsup></mrow></mfrac><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
\alpha_{ij} \;=\; \exp\!\left(-\frac{d_{ij}^2}{2\sigma_t^2}\right)
</annotation></semantics></math> with <a id="def-dij"></a><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>d</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>âˆˆ</mo><mrow><mo stretchy="true" form="prefix">[</mo><mn>0</mn><mo>,</mo><mn>1</mn><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">d_{ij} \in [0,1]</annotation></semantics></math> as the trait dissimilarity (e.g., Gower) and <a id="def-sigmat"></a><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Ïƒ</mi><mi>t</mi></msub><mo>&gt;</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\sigma_t &gt; 0</annotation></semantics></math> as the trait bandwidth. While, the <strong>generalised trait distance/similarity</strong>Â  <a id="def-gall"></a> <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>g</mi><mrow><mi>i</mi><mi>j</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">a</mi><mi mathvariant="normal">l</mi><mi mathvariant="normal">l</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><annotation encoding="application/x-tex">g^{\mathrm{(all)}}_{ij}</annotation></semantics></math> is calculated as the the pairwise trait relationship (distance, similarity, or kernel value, depending on <code>kernel</code> choice) between species <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math> across the entire species set, not restricted to invader-resident pairs. Â </li>
<li>
<strong>Environmental filtering kernel</strong> is: <a id="def-Ke"></a> <a id="def-delta"></a> <a id="def-sigmae"></a> <a id="def-E"></a> <a id="def-theta"></a> <math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>K</mi><mi>e</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><mo>;</mo><msub><mi>Ïƒ</mi><mi>e</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mspace width="0.278em"></mspace><mo>=</mo><mspace width="0.278em"></mspace><mo>exp</mo><mspace width="-0.167em"></mspace><mrow><mo stretchy="true" form="prefix">(</mo><mo>âˆ’</mo><mfrac><msubsup><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow><mn>2</mn></msubsup><mrow><mn>2</mn><msubsup><mi>Ïƒ</mi><mi>e</mi><mn>2</mn></msubsup></mrow></mfrac><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo><mspace width="1.0em"></mspace><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><mspace width="0.278em"></mspace><mo>=</mo><mspace width="0.278em"></mspace><mrow><mi mathvariant="normal">G</mi><mi mathvariant="normal">o</mi><mi mathvariant="normal">w</mi><mi mathvariant="normal">e</mi><mi mathvariant="normal">r</mi></mrow><mspace width="-0.167em"></mspace><mo minsize="1.2" maxsize="1.2" stretchy="false" form="prefix">(</mo><msub><mi>ğ„</mi><mi>s</mi></msub><mo>,</mo><msub><mi>Î¸</mi><mi>j</mi></msub><mo minsize="1.2" maxsize="1.2" stretchy="false" form="postfix">)</mo><mo>,</mo></mrow><annotation encoding="application/x-tex">
K_e(\Delta_{js}; \sigma_e) \;=\; \exp\!\left(-\frac{\Delta_{js}^2}{2\sigma_e^2}\right),
\quad \Delta_{js} \;=\; \mathrm{Gower}\!\big(\mathbf{E}_s, \theta_j\big),</annotation></semantics></math> whereÂ Â <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ğ„</mi><mi>s</mi></msub><annotation encoding="application/x-tex">\mathbf{E}_s</annotation></semantics></math> is the site environment vector and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î¸</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\theta_j</annotation></semantics></math> is the abundance-weighted environmental optimum of resident <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>.Â Â <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Ïƒ</mi><mi>e</mi></msub><mo>&gt;</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\sigma_e &gt; 0</annotation></semantics></math> is the environmental bandwidth. Â </li>
<li>In the <strong>resident context</strong>, <a id="def-Nstar"></a><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>N</mi><mrow><mi>j</mi><mi>s</mi></mrow><mo>*</mo></msubsup><annotation encoding="application/x-tex">N^{*}_{js}</annotation></semantics></math> is the predicted equilibrium or typical abundance of resident <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math> at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>, obtained from the fitted trait-environment model: <math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>N</mi><mrow><mi>j</mi><mi>s</mi></mrow><mo>*</mo></msubsup><mspace width="0.278em"></mspace><mo>=</mo><mspace width="0.278em"></mspace><msub><mi>f</mi><mtext mathvariant="normal">env</mtext></msub><mspace width="-0.167em"></mspace><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>ğ„</mi><mi>s</mi></msub><mo>,</mo><msub><mi>Î¸</mi><mi>j</mi></msub><mo>,</mo><msub><mi>ğ³</mi><mi>j</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>,</mo></mrow><annotation encoding="application/x-tex">
N^{*}_{js} \;=\; f_{\text{env}}\!\left(\mathbf{E}_s, \theta_j, \mathbf{z}_j\right),
</annotation></semantics></math> where <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>f</mi><mtext mathvariant="normal">env</mtext></msub><annotation encoding="application/x-tex">f_{\text{env}}</annotation></semantics></math> denotes the fitted response function (e.g., GLMM/GAM), <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ğ„</mi><mi>s</mi></msub><annotation encoding="application/x-tex">\mathbf{E}_s</annotation></semantics></math> are site covariates, <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î¸</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\theta_j</annotation></semantics></math> summarizes species-level optima/traits, and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ğ³</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\mathbf{z}_j</annotation></semantics></math> are any additional predictors.<br>
Â </li>
<li>Site-level <strong>invasibility</strong> and species-level <strong>invasiveness</strong> summaries of the fitness matrix <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> are derived as follows:
<ul>
<li><p>The proportion of invaders with positive fitness at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math> is:Â Â <a id="def-Vs"></a> <math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>V</mi><mi>s</mi></msub><mspace width="0.278em"></mspace><mo>=</mo><mspace width="0.278em"></mspace><mfrac><mn>1</mn><mi>I</mi></mfrac><munder><mo>âˆ‘</mo><mi>i</mi></munder><mi>ğ•€</mi><mo stretchy="false" form="prefix">{</mo><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><mo>&gt;</mo><mn>0</mn><mo stretchy="false" form="postfix">}</mo></mrow><annotation encoding="application/x-tex">
V_s \;=\; \frac{1}{I}\sum_{i} \mathbb{I}\{\lambda_{is} &gt; 0\}
</annotation></semantics></math></p></li>
<li><p>The total invasion fitness of invader <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> across sites is:Â Â <a id="def-Ii"></a> <math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>I</mi><mi>i</mi></msub><mspace width="0.278em"></mspace><mo>=</mo><mspace width="0.278em"></mspace><munder><mo>âˆ‘</mo><mi>s</mi></munder><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub></mrow><annotation encoding="application/x-tex">
I_i \;=\; \sum_{s} \lambda_{is}
</annotation></semantics></math></p></li>
</ul>
</li>
</ul>
<blockquote>
<p><strong>Note</strong>: Different summaries can be applied - for example, <a href="#def-Vs"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mi>s</mi></msub><annotation encoding="application/x-tex">V_s</annotation></semantics></math></a> can be the mean of <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> over species <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>, and/or <a href="#def-Ii"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mi>i</mi></msub><annotation encoding="application/x-tex">I_i</annotation></semantics></math></a> can be the mean over sites or a sum over positive values only, depending on management emphasis. Â </p>
</blockquote>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="id_2-overview-of-the-invasimap-r-package-conceptual-workflow">2. Overview of the <code>invasimap</code> R package conceptual workflow<a class="anchor" aria-label="anchor" href="#id_2-overview-of-the-invasimap-r-package-conceptual-workflow"></a>
</h2>
<p>This tutorial walks through the full <code>invasimap</code> workflow for quantifying, mapping, and interpreting <strong>invasion fitness</strong> <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> (<code>fitness$lambda[i, s]</code> in <code>lambda_mat</code>) in ecological communities, from raw data to actionable visualisations. The steps operationalise key ecological concepts: <strong>trait centrality</strong> (<code>tdp$scores$centrality</code>); <strong>trait dispersion</strong> (<code>tdp$metrics_df</code>); <strong>interaction strength</strong> <a href="#def-I"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mrow><mi>i</mi><mi>j</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">I_{ijs}</annotation></semantics></math></a> (<code>am$I_raw</code>); <strong>competition</strong> <a href="#def-alpha"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\alpha_{ij}</annotation></semantics></math></a> (<code>comp$a_ij</code>); <strong>environmental filtering</strong> <a href="#def-Ke"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>K</mi><mi>e</mi></msub><annotation encoding="application/x-tex">K_e</annotation></semantics></math></a> (<code>ek$K_env</code>); and summaries of <strong>invasion fitness</strong> <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> including site-level <strong>invasibility</strong> <a href="#def-Vs"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mi>s</mi></msub><annotation encoding="application/x-tex">V_s</annotation></semantics></math></a> and species-level <strong>invasiveness</strong> <a href="#def-Ii"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mi>i</mi></msub><annotation encoding="application/x-tex">I_i</annotation></semantics></math></a>.</p>
<p>All computations use site-level environmental data (<code>site_env</code>), species occurrence or abundance (<code>site_spp_pa</code> / <code>site_spp_ab</code>) at coordinates (<code>site_xy</code>), and species functional traits (<code>spp_trait</code>) to parameterise traitâ€“environmentâ€“interaction models consistent with invasion theory (Hui &amp; Richardson, 2017; Hui et al., 2016, 2021).</p>
<p>The workflow is modular, with each step implemented by targeted functions in <code>invasimap</code>:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Setup &amp; Dependencies</strong> Initialise the R environment, load packages, source helper functions, and set reproducibility controls.</p></li>
<li><p><strong>Data Preparation</strong> Collect, clean, and standardise trait datasets (<code>spp_trait</code>), including automated web-scraping with <strong><code><a href="reference/get_trait_data.html">get_trait_data()</a></code></strong>, merging with trait tables, and type conversion. Produces a standardised trait data frame suitable for downstream analyses.</p></li>
<li>
<p><strong>Functional Trait Space</strong> Characterise species in a multidimensional trait space with <strong><code><a href="reference/compute_trait_space.html">compute_trait_space()</a></code></strong>, calculating:</p>
<ul>
<li>
<strong>Trait centrality</strong> [link TBD] â€” <code>tdp$scores$centrality</code> via <strong><code><a href="reference/compute_trait_similarity.html">compute_trait_similarity()</a></code></strong>.</li>
<li>
<strong>Trait dispersion</strong> [link TBD] â€” <code>tdp$metrics_df</code> via <strong><code><a href="reference/compute_trait_dispersion.html">compute_trait_dispersion()</a></code></strong>.</li>
<li>Pairwise trait distances <a href="#def-dij"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>d</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">d_{ij}</annotation></semantics></math></a> stored as <code>trait_dist</code> or <code>comp$d_ij</code>.</li>
</ul>
</li>
<li>
<p><strong>Traitâ€“Environment Response</strong> Fit generalised linear mixed models linking traits and environmental covariates:</p>
<ul>
<li>Build GLMM formula with <strong><code><a href="reference/build_glmm_formula.html">build_glmm_formula()</a></code></strong>.</li>
<li>Simulate invader traits with <strong><code><a href="reference/simulate_invaders.html">simulate_invaders()</a></code></strong>.</li>
<li>Predict intrinsic growth rates <a href="#def-r"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>r</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">r_{is}</annotation></semantics></math></a> per site using <strong><code><a href="reference/predict_invader_response.html">predict_invader_response()</a></code></strong>, returning <code>fitness$r_mat</code>.</li>
</ul>
</li>
<li>
<p><strong>Interaction Strength</strong> Estimate pairwise biotic influence potentials:</p>
<ul>
<li>Compute <a href="#def-I"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mrow><mi>i</mi><mi>j</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">I_{ijs}</annotation></semantics></math></a> (<code>am$I_raw</code>) from <a href="#def-dij"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>d</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">d_{ij}</annotation></semantics></math></a> via <strong><code><a href="reference/compute_interaction_strength.html">compute_interaction_strength()</a></code></strong>.</li>
<li>Store general interaction distances <a href="#def-gall"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>g</mi><mrow><mi>i</mi><mi>j</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">a</mi><mi mathvariant="normal">l</mi><mi mathvariant="normal">l</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><annotation encoding="application/x-tex">g^{\mathrm{(all)}}_{ij}</annotation></semantics></math></a> (<code>cis$g_all</code>).</li>
<li>Extract resident abundance context <a href="#def-Nstar"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>N</mi><mrow><mi>j</mi><mi>s</mi></mrow><mo>*</mo></msubsup><annotation encoding="application/x-tex">N^{*}_{js}</annotation></semantics></math></a> (<code>cis$Nstar</code>).</li>
</ul>
</li>
<li><p><strong>Competition</strong> Transform distances into competition coefficients <a href="#def-alpha"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\alpha_{ij}</annotation></semantics></math></a> using the Gaussian trait kernel <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>K</mi><mi>t</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>d</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>;</mo><msub><mi>Ïƒ</mi><mi>t</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K_t(d_{ij};\sigma_t)</annotation></semantics></math> with bandwidth <a href="#def-sigmat"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Ïƒ</mi><mi>t</mi></msub><annotation encoding="application/x-tex">\sigma_t</annotation></semantics></math></a> (<code>comp$sigma_t</code>) via <strong><code><a href="reference/compute_competition_kernel.html">compute_competition_kernel()</a></code></strong>.</p></li>
<li>
<p><strong>Environmental Filtering</strong> Quantify match between species and site environments:</p>
<ul>
<li>Estimate environmental optima <a href="#def-theta"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î¸</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\theta_j</annotation></semantics></math></a> (<code>ek$env_opt</code>).</li>
<li>Compute mismatch <a href="#def-delta"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\Delta_{js}</annotation></semantics></math></a> (<code>ek$env_dist</code>) from site conditions <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ğ„</mi><mi>s</mi></msub><annotation encoding="application/x-tex">\mathbf{E}_s</annotation></semantics></math> (<code>site_env</code>).</li>
<li>Transform to environmental weights <a href="#def-Ke"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>K</mi><mi>e</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><mo>;</mo><msub><mi>Ïƒ</mi><mi>e</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K_e(\Delta_{js};\sigma_e)</annotation></semantics></math></a> using bandwidth <a href="#def-sigmae"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Ïƒ</mi><mi>e</mi></msub><annotation encoding="application/x-tex">\sigma_e</annotation></semantics></math></a> (<code>ek$sigma_e</code>) via <strong><code><a href="reference/compute_environment_kernel.html">compute_environment_kernel()</a></code></strong>.</li>
</ul>
</li>
<li>
<p><strong>Invasion Fitness</strong> Computes low-density per-capita growth rates <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> by integrating competition, environmental filtering, and intrinsic growth:</p>
<ul>
<li>
<p><strong>Matrix assembly</strong> â€” <strong><code><a href="reference/assemble_matrices.html">assemble_matrices()</a></code></strong> collates:</p>
<ul>
<li>Invader-resident competition coefficients <a href="#def-alpha"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\alpha_{ij}</annotation></semantics></math></a>
</li>
<li>Site-resident environmental weights <a href="#def-Ke"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>K</mi><mi>e</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><mo>;</mo><msub><mi>Ïƒ</mi><mi>e</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K_e(\Delta_{js};\sigma_e)</annotation></semantics></math></a>
</li>
<li>Resident equilibrium abundances <a href="#def-Nstar"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>N</mi><mrow><mi>j</mi><mi>s</mi></mrow><mo>*</mo></msubsup><annotation encoding="application/x-tex">N^{*}_{js}</annotation></semantics></math></a>
</li>
<li>Predicted intrinsic growth rates <a href="#def-r"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>r</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">r_{is}</annotation></semantics></math></a>
</li>
<li>Returns the impact tensor <a href="#def-I"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mrow><mi>i</mi><mi>j</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">I_{ijs}</annotation></semantics></math></a> and total penalties <a href="#def-Craw"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><annotation encoding="application/x-tex">C^{\mathrm{(raw)}}_{is}</annotation></semantics></math></a>
</li>
</ul>
</li>
<li><p><strong>Fitness calculation</strong> â€” <strong><code><a href="reference/compute_invasion_fitness.html">compute_invasion_fitness()</a></code></strong> applies: <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><mspace width="0.278em"></mspace><mo>=</mo><mspace width="0.278em"></mspace><msub><mi>r</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><mspace width="0.278em"></mspace><mo>âˆ’</mo><mspace width="0.278em"></mspace><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup></mrow><annotation encoding="application/x-tex">\lambda_{is} \;=\; r_{is} \;-\; C^{\mathrm{(raw)}}_{is}</annotation></semantics></math>, where the penalty <a href="#def-Craw"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><annotation encoding="application/x-tex">C^{\mathrm{(raw)}}_{is}</annotation></semantics></math></a> is the sum of all resident impacts <a href="#def-I"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mrow><mi>i</mi><mi>j</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">I_{ijs}</annotation></semantics></math></a> on invader <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.</p></li>
<li><p><strong>Variants</strong> â€” Optional penalty formulations include richness-scaled <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">r</mi><mi mathvariant="normal">i</mi><mi mathvariant="normal">c</mi><mi mathvariant="normal">h</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><mo stretchy="true" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">(C^{\mathrm{(rich)}}_{is})</annotation></semantics></math>, abundance-weighted <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">a</mi><mi mathvariant="normal">b</mi><mi mathvariant="normal">u</mi><mi mathvariant="normal">n</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><mo stretchy="true" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">(C^{\mathrm{(abun)}}_{is})</annotation></semantics></math>, and logistic-capped <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">l</mi><mi mathvariant="normal">o</mi><mi mathvariant="normal">g</mi><mi mathvariant="normal">i</mi><mi mathvariant="normal">s</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><mo stretchy="true" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">(C^{\mathrm{(logis)}}_{is})</annotation></semantics></math>.</p></li>
</ul>
</li>
<li>
<p><strong>Visualisation &amp; Interpretation</strong> Summarise <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math> into:</p>
<ul>
<li>
<strong>Invasibility</strong> <a href="#def-Vs"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mi>s</mi></msub><annotation encoding="application/x-tex">V_s</annotation></semantics></math></a> (<code>summary$Vs</code>): mean or proportion of positive <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math> over invaders at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.</li>
<li>
<strong>Invasiveness</strong> <a href="#def-Ii"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mi>i</mi></msub><annotation encoding="application/x-tex">I_i</annotation></semantics></math></a> (<code>summary$Ii</code>): mean or sum of <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math> over sites for species <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>.</li>
<li>Optional trait-level invasiveness summaries.</li>
</ul>
</li>
</ol>
<blockquote>
<p><strong>Note:</strong> Each module can be run independently if inputs are available, or sequentially to reproduce the full pipeline from data acquisition to invasion fitness mapping. Â </p>
</blockquote>
<hr>
</div>
</div>
<div class="section level1">
<h1 id="step-by-step-workflow">Step-by-step Workflow<a class="anchor" aria-label="anchor" href="#step-by-step-workflow"></a>
</h1>
<div class="section level3">
<h3 id="id_1-install-and-load-invasimap">1. Install and load <code>invasimap</code>
<a class="anchor" aria-label="anchor" href="#id_1-install-and-load-invasimap"></a>
</h3>
<p>Install and load the <code>invasimap</code> package from GitHub, ensuring all functions are available for use in the workflow.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># # install remotes if needed</span></span>
<span><span class="co"># install.packages("remotes")</span></span>
<span><span class="co"># remotes::install_github("macSands/invasimap")</span></span>
<span></span>
<span><span class="co"># Ensure the package is loaded when knitting</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/macSands/invasimap" class="external-link">invasimap</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">otherPkgs</span><span class="op">$</span><span class="va">invasimap</span><span class="op">$</span><span class="va">Version</span></span>
<span></span>
<span><span class="co"># Make sure all the functions are loaded</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/load_all.html" class="external-link">load_all</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># alternative during local development</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="id_2-load-other-r-libraries">2. Load other R libraries<a class="anchor" aria-label="anchor" href="#id_2-load-other-r-libraries"></a>
</h3>
<p>Load core libraries for spatial processing, biodiversity modelling, and visualization required across the <code>invasimap</code> analysis pipeline.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load essential packages</span></span>
<span><span class="co"># library(tidyverse)</span></span>
<span><span class="co"># --- Data Wrangling and Manipulation ---</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span> <span class="co"># Tidy data manipulation verbs (mutate, select, filter, etc.)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span> <span class="co"># Reshape data (wide â†” long, pivot functions)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span> <span class="co"># Modern lightweight data frames (tibble objects)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span> <span class="co"># Functional iteration (map(), etc.)</span></span>
<span></span>
<span><span class="co"># --- String and Factor Utilities ---</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span> <span class="co"># String pattern matching and manipulation (str_detect, etc.)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jacobkap/fastDummies" class="external-link">fastDummies</a></span><span class="op">)</span> <span class="co"># Quickly create dummy/one-hot variables for factors</span></span>
<span></span>
<span><span class="co"># --- Data Visualization ---</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span> <span class="co"># Grammar-of-graphics plotting</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span> <span class="co"># Colorblind-friendly palettes for ggplot2</span></span>
<span><span class="co"># library(lattice)     # Trellis (multi-panel) graphics</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.sthda.com/english/rpkgs/factoextra" class="external-link">factoextra</a></span><span class="op">)</span> <span class="co"># Visualize clustering and multivariate analyses, fviz_nbclust / silhouettes</span></span>
<span><span class="co"># library(RColorBrewer)</span></span>
<span></span>
<span><span class="co"># --- Spatial Data ---</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span> <span class="co"># Handling and plotting spatial vector data (simple features)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/" class="external-link">terra</a></span><span class="op">)</span> <span class="co"># Raster and spatial data operations</span></span>
<span></span>
<span><span class="co"># --- Statistical and Ecological Modelling ---</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/glmmTMB/glmmTMB" class="external-link">glmmTMB</a></span><span class="op">)</span> <span class="co"># Fit GLMMs (Generalized Linear Mixed Models), e.g., Tweedie, NB, Poisson</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.stats.ox.ac.uk/pub/MASS4/" class="external-link">MASS</a></span><span class="op">)</span> <span class="co"># Statistical functions and kernel density estimation (kde2d, etc.)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://svn.r-project.org/R-packages/trunk/cluster/" class="external-link">cluster</a></span><span class="op">)</span> <span class="co"># Clustering algorithms, Gower distance, diagnostics</span></span>
<span><span class="co"># library(vegan)       # Community ecology, ordination (PCoA, diversity metrics)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://davidcsterratt.github.io/geometry/" class="external-link">geometry</a></span><span class="op">)</span> <span class="co"># Convex hulls, volumes, and related geometry calculations</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ClustGeo</span><span class="op">)</span> <span class="co"># for spatially constrained clustering</span></span>
<span></span>
<span><span class="co"># --- Model Performance and Diagnostics ---</span></span>
<span><span class="co"># library(performance) # Model checking, diagnostics, and performance metrics</span></span>
<span><span class="co"># options(warn = -1)</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="id_3-data-access-and-preparation-using-dissmapr">3. Data access and preparation using <code>dissmapr</code>
<a class="anchor" aria-label="anchor" href="#id_3-data-access-and-preparation-using-dissmapr"></a>
</h3>
<p>To acquire and prepare species occurrence data for biodiversity modelling using the <code>dissmapr</code> package, a series of modular functions streamline the workflow from raw observations to spatially aligned environmental predictors.</p>
<div class="section level4">
<h4 id="id_31-install-dissmapr">3.1. Install <code>dissmapr</code>
<a class="anchor" aria-label="anchor" href="#id_31-install-dissmapr"></a>
</h4>
<p>Install and load the <code>dissmapr</code> package from GitHub, ensuring all functions are available for use in the workflow.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># # install remotes if needed</span></span>
<span><span class="co"># install.packages("remotes")</span></span>
<span><span class="co"># remotes::install_github("macSands/dissmapr")</span></span>
<span></span>
<span><span class="co"># Ensure the package is loaded</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">dissmapr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">otherPkgs</span><span class="op">$</span><span class="va">dissmapr</span><span class="op">$</span><span class="va">Version</span></span>
<span><span class="co">#&gt; [1] "0.1.0"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="id_32-import-and-harmonise-biodiversity-occurrence-data">3.2. Import and harmonise biodiversity-occurrence data<a class="anchor" aria-label="anchor" href="#id_32-import-and-harmonise-biodiversity-occurrence-data"></a>
</h4>
<p>The process begins with <a href="https://macsands.github.io/dissmapr/reference/get_occurrence_data.html" class="external-link"><code>dissmapr::get_occurrence_data()</code></a>, which imports biodiversity records, such as a GBIF butterfly dataset for South Africa, and harmonizes them into standardised formats. Input sources can include local CSV files, URLs, or zipped GBIF downloads. The function filters data by taxon and region, returning both raw records and site-by-species matrices in presence-absence or abundance form.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use local GBIF data</span></span>
<span><span class="va">bfly_data</span> <span class="op">&lt;-</span> <span class="fu">dissmapr</span><span class="fu">::</span><span class="fu">get_occurrence_data</span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"gbif_butterflies.csv"</span>, package <span class="op">=</span> <span class="st">"invasimap"</span><span class="op">)</span>,</span>
<span>  source_type <span class="op">=</span> <span class="st">"local_csv"</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">"\t"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check results but only a subset of columns to fit in console</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">bfly_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 81825    52</span></span>
<span><span class="co"># str(bfly_data[,c(51,52,22,23,1,14,16,17,30)])</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bfly_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">51</span>, <span class="fl">52</span>, <span class="fl">22</span>, <span class="fl">23</span>, <span class="fl">1</span>, <span class="fl">14</span>, <span class="fl">16</span>, <span class="fl">17</span>, <span class="fl">30</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;   site_id pa         y        x    gbifID             verbatimScientificName</span></span>
<span><span class="co">#&gt; 1       1  1 -34.42086 19.24410 923051749                   Pieris brassicae</span></span>
<span><span class="co">#&gt; 2       2  1 -33.96044 18.75564 922985630                   Pieris brassicae</span></span>
<span><span class="co">#&gt; 3       3  1 -33.91651 18.40321 922619348 Papilio demodocus subsp. demodocus</span></span>
<span><span class="co">#&gt; 4       1  1 -34.42086 19.24410 922426210 Mylothris agathina subsp. agathina</span></span>
<span><span class="co">#&gt; 5       4  1 -34.35024 18.47488 921650584                  Eutricha capensis</span></span>
<span><span class="co">#&gt; 6       5  1 -33.58570 25.65097 921485695            Drepanogynis bifasciata</span></span>
<span><span class="co">#&gt;   countryCode                                          locality</span></span>
<span><span class="co">#&gt; 1          ZA                                          Hermanus</span></span>
<span><span class="co">#&gt; 2          ZA                                   Polkadraai Road</span></span>
<span><span class="co">#&gt; 3          ZA                                       Signal Hill</span></span>
<span><span class="co">#&gt; 4          ZA                                          Hermanus</span></span>
<span><span class="co">#&gt; 5          ZA Cape of Good Hope / Cape Point Area, South Africa</span></span>
<span><span class="co">#&gt; 6          ZA                             Kudu Ridge Game Lodge</span></span>
<span><span class="co">#&gt;          eventDate</span></span>
<span><span class="co">#&gt; 1 2012-10-13T00:00</span></span>
<span><span class="co">#&gt; 2 2012-11-01T00:00</span></span>
<span><span class="co">#&gt; 3 2012-10-31T00:00</span></span>
<span><span class="co">#&gt; 4 2012-10-13T00:00</span></span>
<span><span class="co">#&gt; 5 2012-10-30T00:00</span></span>
<span><span class="co">#&gt; 6 2012-10-23T00:00</span></span>
<span></span>
<span><span class="co"># Use local data loaded into the environment as a data.frame</span></span>
<span><span class="co"># local_df = read.csv(system.file("extdata", "site_species.csv", package = "dissmapr")</span></span>
<span><span class="co"># head(local_df)</span></span>
<span><span class="co"># bfly_data = get_occurrence_data(</span></span>
<span><span class="co">#   data = local_df,</span></span>
<span><span class="co">#   source_type = 'data_frame')</span></span>
<span></span>
<span><span class="co"># # Use local .csv file in `invasimap` package</span></span>
<span><span class="co"># bfly_data = get_occurrence_data(</span></span>
<span><span class="co">#   data = system.file("extdata", "site_species.csv", package = "invasimap"),</span></span>
<span><span class="co">#   source_type = 'local_csv')</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># # Check results but only a subset of columns to fit in console</span></span>
<span><span class="co"># dim(bfly_data)</span></span>
<span><span class="co"># # str(bfly_data)</span></span>
<span><span class="co"># head(bfly_data)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="id_33-format-biodiversity-records-to-longwide-formats">3.3. Format biodiversity records to long/wide formats<a class="anchor" aria-label="anchor" href="#id_33-format-biodiversity-records-to-longwide-formats"></a>
</h4>
<p>Next, <a href="https://macsands.github.io/dissmapr/reference/format_df.html" class="external-link"><code>dissmapr::format_df()</code></a> restructures the raw records into tidy long and wide formats. This assigns unique site IDs, extracts key fields (coordinates, species names, observation values), and prepares two main outputs: <code>site_obs</code> (long format for mapping) and <code>site_spp</code> (wide format for species-level analysis).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Continue from GBIF data</span></span>
<span><span class="va">bfly_result</span> <span class="op">&lt;-</span> <span class="fu">dissmapr</span><span class="fu">::</span><span class="fu">format_df</span><span class="op">(</span></span>
<span>  data        <span class="op">=</span> <span class="va">bfly_data</span>, <span class="co"># A `data.frame` of biodiversity records</span></span>
<span>  species_col <span class="op">=</span> <span class="st">"verbatimScientificName"</span>, <span class="co"># Name of species column (required for `"long"`)</span></span>
<span>  value_col   <span class="op">=</span> <span class="st">"pa"</span>, <span class="co"># Name of value column (e.g. presence/abundance; for `"long"`)</span></span>
<span>  extra_cols  <span class="op">=</span> <span class="cn">NULL</span>, <span class="co"># Character vector of other columns to keep</span></span>
<span>  format      <span class="op">=</span> <span class="st">"long"</span> <span class="co"># Either`"long"` or `"wide"`</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># # Continue using local data</span></span>
<span><span class="co"># bfly_result = dissmapr::format_df(</span></span>
<span><span class="co">#   data        = bfly_data, # A `data.frame` of biodiversity records</span></span>
<span><span class="co">#   species_col = 'sp_name', # Name of species column (required for `"long"`)</span></span>
<span><span class="co">#   value_col   = 'count' # Name of value column (e.g. presence/abundance; for `"long"`)</span></span>
<span><span class="co">#   )</span></span>
<span></span>
<span><span class="co"># Check `bfly_result` structure</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">bfly_result</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ site_obs:'data.frame':    79953 obs. of  5 variables:</span></span>
<span><span class="co">#&gt;  $ site_spp: tibble [56,090 Ã— 2,871] (S3: tbl_df/tbl/data.frame)</span></span>
<span></span>
<span><span class="co"># Optional: Create new objects from list items</span></span>
<span><span class="va">site_obs</span> <span class="op">&lt;-</span> <span class="va">bfly_result</span><span class="op">$</span><span class="va">site_obs</span></span>
<span><span class="va">site_spp</span> <span class="op">&lt;-</span> <span class="va">bfly_result</span><span class="op">$</span><span class="va">site_spp</span></span>
<span></span>
<span><span class="co"># Check results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">site_obs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 79953     5</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">site_obs</span><span class="op">)</span></span>
<span><span class="co">#&gt;   site_id        x         y                            species value</span></span>
<span><span class="co">#&gt; 1       1 19.24410 -34.42086                   Pieris brassicae     1</span></span>
<span><span class="co">#&gt; 2       2 18.75564 -33.96044                   Pieris brassicae     1</span></span>
<span><span class="co">#&gt; 3       3 18.40321 -33.91651 Papilio demodocus subsp. demodocus     1</span></span>
<span><span class="co">#&gt; 4       1 19.24410 -34.42086 Mylothris agathina subsp. agathina     1</span></span>
<span><span class="co">#&gt; 5       4 18.47488 -34.35024                  Eutricha capensis     1</span></span>
<span><span class="co">#&gt; 6       5 25.65097 -33.58570            Drepanogynis bifasciata     1</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">site_spp</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 56090  2871</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">site_spp</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 6</span></span>
<span><span class="co">#&gt;   site_id     x     y `Mylothris agathina subsp. agathina` `Pieris brassicae`</span></span>
<span><span class="co">#&gt;     &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt;                                &lt;dbl&gt;              &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1       1  19.2 -34.4                                    1                  1</span></span>
<span><span class="co">#&gt; 2       2  18.8 -34.0                                    0                  1</span></span>
<span><span class="co">#&gt; 3       3  18.4 -33.9                                    0                  0</span></span>
<span><span class="co">#&gt; 4       4  18.5 -34.4                                    0                  0</span></span>
<span><span class="co">#&gt; 5       5  25.7 -33.6                                    0                  0</span></span>
<span><span class="co">#&gt; 6       6  22.2 -33.6                                    0                  0</span></span>
<span><span class="co">#&gt; # â„¹ 1 more variable: `Tarucus thespis` &lt;dbl&gt;</span></span>
<span></span>
<span><span class="co">#### Get parameters from processed data to use later</span></span>
<span><span class="co"># Number of species</span></span>
<span><span class="op">(</span><span class="va">n_sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">site_spp</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">-</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2868</span></span>
<span></span>
<span><span class="co"># Species names</span></span>
<span><span class="va">sp_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">site_spp</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">sp_cols</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;  [1] "Mylothris agathina subsp. agathina" "Pieris brassicae"                  </span></span>
<span><span class="co">#&gt;  [3] "Tarucus thespis"                    "Acraea horta"                      </span></span>
<span><span class="co">#&gt;  [5] "Danaus chrysippus"                  "Papilio demodocus subsp. demodocus"</span></span>
<span><span class="co">#&gt;  [7] "Eutricha capensis"                  "Mesocelis monticola"               </span></span>
<span><span class="co">#&gt;  [9] "Vanessa cardui"                     "Cuneisigna obstans"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="id_34-generate-spatial-grid-and-gridded-summaries">3.4. Generate spatial grid and gridded summaries<a class="anchor" aria-label="anchor" href="#id_34-generate-spatial-grid-and-gridded-summaries"></a>
</h4>
<p>To integrate the data spatially, <a href="https://macsands.github.io/dissmapr/reference/generate_grid.html" class="external-link"><code>dissmapr::generate_grid()</code></a> overlays a user-defined spatial lattice (e.g.Â 0.5Â° grid), aggregates biodiversity observations per grid cell, and computes standardised metrics such as species richness and observation effort. Outputs include gridded species matrices (<code>grid_spp</code>, <code>grid_spp_pa</code>), a spatial polygon (<code>grid_sf</code>), and raster layers (<code>grid_r</code>), enabling downstream spatial modelling.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Load the national boundary</span></span>
<span><span class="va">rsa</span> <span class="op">&lt;-</span> <span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_read.html" class="external-link">st_read</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rsa.shp"</span>, package <span class="op">=</span> <span class="st">"invasimap"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Reading layer `rsa' from data source </span></span>
<span><span class="co">#&gt;   `D:\Methods\R\myR_Packages\cleanVersions\invasimap\inst\extdata\rsa.shp' </span></span>
<span><span class="co">#&gt;   using driver `ESRI Shapefile'</span></span>
<span><span class="co">#&gt; Simple feature collection with 11 features and 8 fields</span></span>
<span><span class="co">#&gt; Geometry type: MULTIPOLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 16.45189 ymin: -34.83417 xmax: 32.94498 ymax: -22.12503</span></span>
<span><span class="co">#&gt; Geodetic CRS:  WGS 84</span></span>
<span></span>
<span><span class="co"># 2. Choose a working resolution</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fl">0.5</span> <span class="co"># decimal degreesÂ° (â‰ˆ 55 km at the equator)</span></span>
<span></span>
<span><span class="co"># 3. Convert the AoI to a 'terra' vector</span></span>
<span><span class="va">rsa_vect</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/vect.html" class="external-link">vect</a></span><span class="op">(</span><span class="va">rsa</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. Initialise a blank raster template</span></span>
<span><span class="va">grid</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="va">rsa_vect</span>, resolution <span class="op">=</span> <span class="va">res</span>, crs <span class="op">=</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/crs.html" class="external-link">crs</a></span><span class="op">(</span><span class="va">rsa_vect</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 5. Populate the raster with placeholder values</span></span>
<span><span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/values.html" class="external-link">values</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="co"># 6. Clip the raster to the AoI</span></span>
<span><span class="va">grid_masked</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/mask.html" class="external-link">mask</a></span><span class="op">(</span><span class="va">grid</span>, <span class="va">rsa_vect</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 7. Generate a 0.5Â° grid summary for the point dataset `site_spp`</span></span>
<span><span class="va">grid_list</span> <span class="op">&lt;-</span> <span class="fu">dissmapr</span><span class="fu">::</span><span class="fu">generate_grid</span><span class="op">(</span></span>
<span>  data          <span class="op">=</span> <span class="va">site_spp</span>, <span class="co"># point data with x/y + species columns</span></span>
<span>  x_col         <span class="op">=</span> <span class="st">"x"</span>, <span class="co"># longitude column</span></span>
<span>  y_col         <span class="op">=</span> <span class="st">"y"</span>, <span class="co"># latitude  column</span></span>
<span>  grid_size     <span class="op">=</span> <span class="fl">0.5</span>, <span class="co"># cell size in degrees</span></span>
<span>  sum_cols      <span class="op">=</span> <span class="fl">4</span><span class="op">:</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">site_spp</span><span class="op">)</span>, <span class="co"># columns to aggregate * could also use `names(site_spp)[4:ncol(site_spp)]`</span></span>
<span>  crs_epsg      <span class="op">=</span> <span class="fl">4326</span> <span class="co"># WGS84</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect the returned list</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">grid_list</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 4</span></span>
<span><span class="co">#&gt;  $ grid_r     :S4 class 'SpatRaster' [package "terra"]</span></span>
<span><span class="co">#&gt;  $ grid_sf    :Classes 'sf' and 'data.frame':    1110 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;   ..- attr(*, "sf_column")= chr "geometry"</span></span>
<span><span class="co">#&gt;   ..- attr(*, "agr")= Factor w/ 3 levels "constant","aggregate",..: NA NA NA NA NA NA NA</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "names")= chr [1:7] "centroid_lon" "centroid_lat" "grid_id" "mapsheet" ...</span></span>
<span><span class="co">#&gt;  $ grid_spp   : tibble [415 Ã— 2,874] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;  $ grid_spp_pa: tibble [415 Ã— 2,874] (S3: tbl_df/tbl/data.frame)</span></span>
<span></span>
<span><span class="co"># (Optional) Promote list items to named objects</span></span>
<span><span class="va">grid_r</span> <span class="op">&lt;-</span> <span class="va">grid_list</span><span class="op">$</span><span class="va">grid_r</span><span class="op">$</span><span class="va">grid_id</span> <span class="co"># raster</span></span>
<span><span class="va">grid_sf</span> <span class="op">&lt;-</span> <span class="va">grid_list</span><span class="op">$</span><span class="va">grid_sf</span> <span class="co"># polygons for mapping or joins</span></span>
<span><span class="va">grid_spp</span> <span class="op">&lt;-</span> <span class="va">grid_list</span><span class="op">$</span><span class="va">grid_spp</span> <span class="co"># tabular summary per cell</span></span>
<span><span class="va">grid_spp_pa</span> <span class="op">&lt;-</span> <span class="va">grid_list</span><span class="op">$</span><span class="va">grid_spp_pa</span> <span class="co"># presence/absence summary</span></span>
<span></span>
<span><span class="co"># Quick checks</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">grid_sf</span><span class="op">)</span> <span class="co"># ; head(grid_sf)</span></span>
<span><span class="co">#&gt; [1] 1110    8</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">grid_spp</span><span class="op">)</span> <span class="co"># ; head(grid_spp[, 1:8])</span></span>
<span><span class="co">#&gt; [1]  415 2874</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">grid_spp_pa</span><span class="op">)</span> <span class="co"># ; head(grid_spp_pa[, 1:8])</span></span>
<span><span class="co">#&gt; [1]  415 2874</span></span>
<span></span>
<span><span class="co"># 1. Extract &amp; stretch the layers</span></span>
<span><span class="va">effRich_r</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">grid_list</span><span class="op">$</span><span class="va">grid_r</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"obs_sum"</span>, <span class="st">"spp_rich"</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2. Open a 1Ã—2 layout and plot each layer + outline</span></span>
<span><span class="va">old_par</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span></span>
<span>  mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, <span class="co"># multiâ€figure by row: 1 row and 2 columns</span></span>
<span>  mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="co"># margins sizes: bottom (1 lines)|left (1)|top (1)|right (2)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">effRich_r</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    col <span class="op">=</span> <span class="fu">viridisLite</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">turbo</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>    colNA <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    axes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"Sampling effort (âˆšobs count)"</span>,</span>
<span>      <span class="st">"Species richness (âˆšunique count)"</span></span>
<span>    <span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,</span>
<span>    cex.main <span class="op">=</span> <span class="fl">0.8</span></span>
<span>  <span class="op">)</span> <span class="co"># â† smaller title)</span></span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/vect.html" class="external-link">vect</a></span><span class="op">(</span><span class="va">rsa</span><span class="op">)</span>, add <span class="op">=</span> <span class="cn">TRUE</span>, border <span class="op">=</span> <span class="st">"black"</span>, lwd <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="reference/figures/README-grid-1.png" width="100%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span><span class="va">old_par</span><span class="op">)</span> <span class="co"># reset plotting parameters</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="id_35-retrieve-crop-resample-and-link-environmental-rasters-to-sampling-sites">3.5. Retrieve, crop, resample, and link environmental rasters to sampling sites<a class="anchor" aria-label="anchor" href="#id_35-retrieve-crop-resample-and-link-environmental-rasters-to-sampling-sites"></a>
</h4>
<p>Environmental predictors are appended using <a href="https://macsands.github.io/dissmapr/reference/get_enviro_data.html" class="external-link"><code>dissmapr::get_enviro_data()</code></a>, which buffers the grid, downloads raster data (e.g.Â WorldClim bioclimatic variables), resamples it, and links values to grid-cell centroids. This produces both a site-by-environment data frame (<code>env_df</code>) and a SpatRaster object (<code>env_r</code>), aligning biological and environmental data.</p>
<p>Begin by reading in a predefined target species list, then filter a site-by-species dataset (<code>grid_spp</code>) to retain only relevant species observations, and reshape the data for further analysis. This produces both a filtered long-format dataset (<code>grid_obs</code>) and a cleaned wide-format site-by-species matrix (<code>grid_spp</code>).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in target species list</span></span>
<span><span class="va">species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>  <span class="st">"rsa_butterfly_species_names_n27_100plus.csv"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"invasimap"</span></span>
<span><span class="op">)</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">$</span><span class="va">species</span></span>
<span></span>
<span><span class="co"># Filter `grid_spp` and convert to long-format</span></span>
<span><span class="va">grid_obs</span> <span class="op">&lt;-</span> <span class="va">grid_spp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">mapsheet</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># Drop mapsheet metadata</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span></span>
<span>    cols <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">grid_id</span>, <span class="va">centroid_lon</span>, <span class="va">centroid_lat</span>, <span class="va">obs_sum</span>, <span class="va">spp_rich</span><span class="op">)</span>, <span class="co"># Keep core metadata columns only</span></span>
<span>    names_to <span class="op">=</span> <span class="st">"species"</span>,</span>
<span>    values_to <span class="op">=</span> <span class="st">"count"</span>,</span>
<span>    values_drop_na <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>    <span class="co"># obs_sum &gt; 100,                                   # Only high-observation sites</span></span>
<span>    <span class="va">count</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="co"># Remove absent species</span></span>
<span>    <span class="va">species</span> <span class="op"><a href="https://rspatial.github.io/terra/reference/match.html" class="external-link">%in%</a></span> <span class="op">!</span><span class="op">!</span><span class="va">species</span> <span class="co"># Keep only target species</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span></span>
<span>    site_id <span class="op">=</span> <span class="va">grid_id</span>, <span class="co"># Change 'grid_id' to 'site_id'</span></span>
<span>    x <span class="op">=</span> <span class="va">centroid_lon</span>, <span class="co"># Change 'centroid_lon' to 'x'</span></span>
<span>    y <span class="op">=</span> <span class="va">centroid_lat</span> <span class="co"># Change 'centroid_lat' to 'y'</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/relocate.html" class="external-link">relocate</a></span><span class="op">(</span><span class="va">site_id</span>, <span class="va">x</span>, <span class="va">y</span>, <span class="va">obs_sum</span>, <span class="va">spp_rich</span>, <span class="va">species</span>, <span class="va">count</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">grid_obs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1737    7</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">grid_obs</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 7</span></span>
<span><span class="co">#&gt;   site_id     x     y obs_sum spp_rich species                    count</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;                      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 1027     29.2 -22.3      41       31 Utetheisa pulchella            1</span></span>
<span><span class="co">#&gt; 2 1029     30.3 -22.3       7        7 Danaus chrysippus orientis     1</span></span>
<span><span class="co">#&gt; 3 1029     30.3 -22.3       7        7 Telchinia serena               1</span></span>
<span><span class="co">#&gt; 4 1031     31.3 -22.3     107       76 Vanessa cardui                 1</span></span>
<span><span class="co">#&gt; 5 1031     31.3 -22.3     107       76 Utetheisa pulchella            2</span></span>
<span><span class="co">#&gt; 6 1031     31.3 -22.3     107       76 Hypolimnas misippus            2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">grid_obs</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 27</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">grid_obs</span><span class="op">$</span><span class="va">site_id</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 314</span></span>
<span></span>
<span><span class="co"># Reshape site-by-species matrix to wide format and clean</span></span>
<span><span class="va">grid_spp</span> <span class="op">&lt;-</span> <span class="va">grid_obs</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span></span>
<span>    names_from <span class="op">=</span> <span class="va">species</span>,</span>
<span>    values_from <span class="op">=</span> <span class="va">count</span>,</span>
<span>    values_fill <span class="op">=</span> <span class="fl">0</span> <span class="co"># Replace missing counts with 0</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">grid_spp</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 314  32</span></span>
<span><span class="co"># head(grid_spp)</span></span></code></pre></div>
<p>Then proceed to retrieve and process environmental data using <a href="https://macsands.github.io/dissmapr/reference/get_enviro_data.html" class="external-link"><code>dissmapr::get_enviro_data()</code></a>. In the example below, 19 bioclimatic variables are downloaded from WorldClim v2.1 (â‰ˆ10 km resolution) for all site centroids in the <code>grid_spp</code> dataset. It performs the following steps:</p>
<ol style="list-style-type: decimal">
<li>Retrieves WorldClim â€œbioâ€ variables via the <code>geodata</code> interface.</li>
<li>Buffers the area of interest (AOI) by 10 km.</li>
<li>Retains site-level metadata (<code>obs_sum</code>, <code>spp_rich</code>) and excludes species columns.</li>
</ol>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Retrieve 19 bioclim layers (â‰ˆ10-km, WorldClim v2.1) for all grid centroids</span></span>
<span><span class="va">data_path</span> <span class="op">&lt;-</span> <span class="st">"inst/extdata"</span> <span class="co"># cache folder for rasters</span></span>
<span><span class="va">enviro_list</span> <span class="op">&lt;-</span> <span class="fu">dissmapr</span><span class="fu">::</span><span class="fu">get_enviro_data</span><span class="op">(</span></span>
<span>  data       <span class="op">=</span> <span class="va">grid_spp</span>, <span class="co"># centroids + obs_sum + spp_rich</span></span>
<span>  buffer_km  <span class="op">=</span> <span class="fl">10</span>, <span class="co"># pad the AOI slightly</span></span>
<span>  source     <span class="op">=</span> <span class="st">"geodata"</span>, <span class="co"># WorldClim/SoilGrids interface</span></span>
<span>  var        <span class="op">=</span> <span class="st">"bio"</span>, <span class="co"># bioclim variable set</span></span>
<span>  res        <span class="op">=</span> <span class="fl">5</span>, <span class="co"># 5-arc-min â‰ˆ 10 km</span></span>
<span>  grid_r     <span class="op">=</span> <span class="va">grid_r</span>, <span class="co"># To set resampling resolution, if necessary</span></span>
<span>  path       <span class="op">=</span> <span class="va">data_path</span>,</span>
<span>  sp_cols    <span class="op">=</span> <span class="fl">7</span><span class="op">:</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">grid_spp</span><span class="op">)</span>, <span class="co"># ignore species columns</span></span>
<span>  ext_cols   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"obs_sum"</span>, <span class="st">"spp_rich"</span><span class="op">)</span> <span class="co"># carry effort &amp; richness through</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Quick checks</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">enviro_list</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 3</span></span>
<span><span class="co">#&gt;  $ env_rast:S4 class 'SpatRaster' [package "terra"]</span></span>
<span><span class="co">#&gt;  $ sites_sf: sf [314 Ã— 2] (S3: sf/tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   ..- attr(*, "sf_column")= chr "geometry"</span></span>
<span><span class="co">#&gt;   ..- attr(*, "agr")= Factor w/ 3 levels "constant","aggregate",..: NA</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "names")= chr "site_id"</span></span>
<span><span class="co">#&gt;  $ env_df  : tibble [314 Ã— 24] (S3: tbl_df/tbl/data.frame)</span></span>
<span></span>
<span><span class="co"># (Optional) Assign concise layer names for readability</span></span>
<span><span class="co"># Find names here https://www.worldclim.org/data/bioclim.html</span></span>
<span><span class="va">names_env</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"temp_mean"</span>, <span class="st">"mdr"</span>, <span class="st">"iso"</span>, <span class="st">"temp_sea"</span>, <span class="st">"temp_max"</span>, <span class="st">"temp_min"</span>,</span>
<span>  <span class="st">"temp_range"</span>, <span class="st">"temp_wetQ"</span>, <span class="st">"temp_dryQ"</span>, <span class="st">"temp_warmQ"</span>,</span>
<span>  <span class="st">"temp_coldQ"</span>, <span class="st">"rain_mean"</span>, <span class="st">"rain_wet"</span>, <span class="st">"rain_dry"</span>,</span>
<span>  <span class="st">"rain_sea"</span>, <span class="st">"rain_wetQ"</span>, <span class="st">"rain_dryQ"</span>, <span class="st">"rain_warmQ"</span>, <span class="st">"rain_coldQ"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">enviro_list</span><span class="op">$</span><span class="va">env_rast</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">names_env</span></span>
<span></span>
<span><span class="co"># (Optional) Promote frequently-used objects</span></span>
<span><span class="va">env_r</span> <span class="op">&lt;-</span> <span class="va">enviro_list</span><span class="op">$</span><span class="va">env_rast</span> <span class="co"># cropped climate stack</span></span>
<span><span class="va">env_df</span> <span class="op">&lt;-</span> <span class="va">enviro_list</span><span class="op">$</span><span class="va">env_df</span> <span class="co"># site Ã— environment data-frame</span></span>
<span></span>
<span><span class="co"># Quick checks</span></span>
<span><span class="va">env_r</span></span>
<span><span class="co">#&gt; class       : SpatRaster </span></span>
<span><span class="co">#&gt; size        : 30, 37, 19  (nrow, ncol, nlyr)</span></span>
<span><span class="co">#&gt; resolution  : 0.5, 0.5  (x, y)</span></span>
<span><span class="co">#&gt; extent      : 15.5, 34, -36, -21  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#&gt; coord. ref. : lon/lat WGS 84 (EPSG:4326) </span></span>
<span><span class="co">#&gt; source(s)   : memory</span></span>
<span><span class="co">#&gt; names       : temp_mean,       mdr,      iso, temp_sea, temp_max,  temp_min, ... </span></span>
<span><span class="co">#&gt; min values  :  9.779773,  8.977007, 47.10606, 228.9986, 19.92147, -4.110302, ... </span></span>
<span><span class="co">#&gt; max values  : 24.406433, 18.352308, 64.92966, 653.4167, 36.19497, 12.005042, ...</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">env_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 314  24</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">env_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 24</span></span>
<span><span class="co">#&gt;   site_id     x     y bio01 bio02 bio03 bio04 bio05 bio06 bio07 bio08 bio09</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 1027     29.2 -22.3  21.8 14.5   55.1  430.  32.6  6.30  26.3  26.2  15.9</span></span>
<span><span class="co">#&gt; 2 1029     30.3 -22.3  22.8 13.9   58.0  359.  32.7  8.79  23.9  26.5  17.8</span></span>
<span><span class="co">#&gt; 3 1031     31.3 -22.3  24.2 14.2   61.3  326.  34.2 10.9   23.2  27.5  19.7</span></span>
<span><span class="co">#&gt; 4 117      18.2 -34.3  20.2 11.8   56.8  317.  29.8  9.29  20.5  19.9  19.8</span></span>
<span><span class="co">#&gt; 5 118      18.7 -34.3  16.2  9.28  52.3  309.  25.4  7.65  17.7  12.4  19.8</span></span>
<span><span class="co">#&gt; 6 119      19.3 -34.3  15.8 10.2   53.5  321.  25.8  6.67  19.2  11.8  19.6</span></span>
<span><span class="co">#&gt; # â„¹ 12 more variables: bio10 &lt;dbl&gt;, bio11 &lt;dbl&gt;, bio12 &lt;dbl&gt;, bio13 &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   bio14 &lt;dbl&gt;, bio15 &lt;dbl&gt;, bio16 &lt;dbl&gt;, bio17 &lt;dbl&gt;, bio18 &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   bio19 &lt;dbl&gt;, obs_sum &lt;dbl&gt;, spp_rich &lt;dbl&gt;</span></span>
<span></span>
<span><span class="co"># Build the final site Ã— environment table</span></span>
<span><span class="va">grid_env</span> <span class="op">&lt;-</span> <span class="va">env_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>    <span class="va">site_id</span>, <span class="va">x</span>, <span class="va">y</span>,</span>
<span>    <span class="va">obs_sum</span>, <span class="va">spp_rich</span>, <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span></span>
<span>    .cols <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">site_id</span>, <span class="va">x</span>, <span class="va">y</span>, <span class="va">obs_sum</span>, <span class="va">spp_rich</span><span class="op">)</span>, <span class="co"># all other columns</span></span>
<span>    .fns <span class="op">=</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span>, <span class="co"># Scale bio</span></span>
<span>    .names <span class="op">=</span> <span class="st">"{.col}"</span> <span class="co"># keep same names</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">grid_env</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; tibble [314 Ã— 24] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">grid_env</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 24</span></span>
<span><span class="co">#&gt;   site_id     x     y obs_sum spp_rich  bio01   bio02  bio03  bio04  bio05 bio06</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 1027     29.2 -22.3      41       31  1.75   0.274  -0.309  0.272  1.24  0.605</span></span>
<span><span class="co">#&gt; 2 1029     30.3 -22.3       7        7  2.20  -0.0315  0.616 -0.450  1.27  1.28 </span></span>
<span><span class="co">#&gt; 3 1031     31.3 -22.3     107       76  2.78   0.150   1.68  -0.792  1.79  1.87 </span></span>
<span><span class="co">#&gt; 4 117      18.2 -34.3    4246      231  1.08  -1.05    0.236 -0.883  0.241 1.42 </span></span>
<span><span class="co">#&gt; 5 118      18.7 -34.3    2202      215 -0.628 -2.25   -1.21  -0.975 -1.30  0.973</span></span>
<span><span class="co">#&gt; 6 119      19.3 -34.3     989      173 -0.799 -1.79   -0.838 -0.842 -1.15  0.706</span></span>
<span><span class="co">#&gt; # â„¹ 13 more variables: bio07 &lt;dbl&gt;, bio08 &lt;dbl&gt;, bio09 &lt;dbl&gt;, bio10 &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   bio11 &lt;dbl&gt;, bio12 &lt;dbl&gt;, bio13 &lt;dbl&gt;, bio14 &lt;dbl&gt;, bio15 &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   bio16 &lt;dbl&gt;, bio17 &lt;dbl&gt;, bio18 &lt;dbl&gt;, bio19 &lt;dbl&gt;</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="id_36-remove-highly-correlated-predictors-optional">3.6. Remove highly correlated predictors (optional)<a class="anchor" aria-label="anchor" href="#id_36-remove-highly-correlated-predictors-optional"></a>
</h4>
<p>Finally, <a href="https://macsands.github.io/dissmapr/reference/rm_correlated.html" class="external-link"><code>dissmapr::rm_correlated()</code></a> optionally reduces multicollinearity by filtering out highly correlated predictors based on a threshold (e.g.Â r &gt; 0.70), improving model stability and interpretability. Together, these functions provide a reproducible and scalable pipeline for preparing ecological datasets for spatial analysis.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># # (Optional) Rename BIO</span></span>
<span><span class="co"># names(env_df) = c("grid_id", "centroid_lon", "centroid_lat", names_env, "obs_sum", "spp_rich")</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># # Run the filter and compare dimensions</span></span>
<span><span class="co"># # Filter environmental predictors for |r| &gt; 0.70</span></span>
<span><span class="co"># env_vars_reduced = dissmapr::rm_correlated(</span></span>
<span><span class="co">#   data       = env_df[, 4:23],  # drop ID + coord columns</span></span>
<span><span class="co">#   cols       = NULL,                  # infer all numeric cols</span></span>
<span><span class="co">#   threshold  = 0.70,</span></span>
<span><span class="co">#   plot       = TRUE                   # show heat-map of retained vars</span></span>
<span><span class="co"># )</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># # Before vs after</span></span>
<span><span class="co"># c(original = ncol(env_df[, c(4, 6:24)]),</span></span>
<span><span class="co">#   reduced  = ncol(env_vars_reduced))</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="id_4-data-access-and-preparation-using-invasimap">4. Data access and preparation using <code>invasimap</code>
<a class="anchor" aria-label="anchor" href="#id_4-data-access-and-preparation-using-invasimap"></a>
</h3>
<div class="section level4">
<h4 id="id_41-retrieve-and-link-trait-and-metadata-for-each-species">4.1. Retrieve and link trait and metadata for each species<a class="anchor" aria-label="anchor" href="#id_41-retrieve-and-link-trait-and-metadata-for-each-species"></a>
</h4>
<p>This utility provides an automated pipeline for extracting and joining both biological trait data and rich metadata for any focal species. The function integrates several steps:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Trait Table Lookup</strong>: Retrieves speciesâ€™ trait data from a local trait table (CSV) or a <a href="https://www.try-db.org/TryWeb/Home.php" class="external-link">TRY</a>-style database, using fuzzy matching to ensure robust linkage even when there are minor naming inconsistencies.</li>
<li>
<strong>Wikipedia Metadata Scraping</strong>: Optionally augments each species entry with a taxonomic summary, higher taxonomy, and representative images scraped directly from <a href="https://www.wikipedia.org/" class="external-link">Wikipedia</a>.</li>
<li>
<strong>Image-based Color Palette Extraction</strong>: If enabled, downloads and processes public domain images to extract the most frequent colors, optionally removing green/white backgrounds to focus on diagnostic features.</li>
<li>
<strong>Flexible Output</strong>: Returns a single-row tibble with the species name, trait data, taxonomic metadata, image URL, and color palette - all harmonized for downstream analyses or visualization.</li>
</ol>
<p>This function greatly simplifies the assembly of a unified species-trait-metadata table, which is essential for trait-based community ecology, macroecology, and biodiversity informatics projects.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># # Local trait data.frame version 1</span></span>
<span><span class="co"># btfly_traits1 = read.csv(system.file("extdata", "Middleton_etal_2020_traits.csv", package = "invasimap"))</span></span>
<span><span class="co"># str(btfly_traits1)</span></span>
<span><span class="co"># length(unique(btfly_traits1$Species))</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># # Github trait data.frame</span></span>
<span><span class="co"># git_url = "https://raw.githubusercontent.com/RiesLabGU/LepTraits/main/consensus/consensus.csv"</span></span>
<span><span class="co"># # Make sure inst/extdata exists then define destination</span></span>
<span><span class="co"># dir.create("inst/extdata", recursive = TRUE, showWarnings = FALSE)</span></span>
<span><span class="co"># destfile = file.path("inst", "extdata", "consensus.csv")</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># # Download the raw CSV</span></span>
<span><span class="co"># download.file(</span></span>
<span><span class="co">#   url = git_url,</span></span>
<span><span class="co">#   destfile = destfile,</span></span>
<span><span class="co">#   mode = "wb"    # important on Windows</span></span>
<span><span class="co"># )</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># # 4. Read it from disk</span></span>
<span><span class="co"># btfly_traits2 = read.csv(destfile, stringsAsFactors = FALSE)</span></span>
<span><span class="co"># str(btfly_traits2)</span></span>
<span><span class="co"># length(unique(btfly_traits2$Species))</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># # Retrieve and join trait/metadata for all species in the observation set</span></span>
<span><span class="co"># spp_traits = purrr::map_dfr(</span></span>
<span><span class="co">#   unique(grid_obs$species),</span></span>
<span><span class="co">#   ~get_trait_data(</span></span>
<span><span class="co">#     species = .x,</span></span>
<span><span class="co">#     remove_bg = FALSE,</span></span>
<span><span class="co">#     n_palette = 5,</span></span>
<span><span class="co">#     preview = FALSE,</span></span>
<span><span class="co">#     save_folder = NULL,</span></span>
<span><span class="co">#     do_summary = TRUE,</span></span>
<span><span class="co">#     do_taxonomy = TRUE,</span></span>
<span><span class="co">#     do_image = TRUE,</span></span>
<span><span class="co">#     do_palette = TRUE,</span></span>
<span><span class="co">#     use_try = FALSE,</span></span>
<span><span class="co">#     try_data = NULL,</span></span>
<span><span class="co">#     # local_trait_df = btfly_traits1,</span></span>
<span><span class="co">#     local_trait_df = btfly_traits2,</span></span>
<span><span class="co">#     local_species_col = 'Species',</span></span>
<span><span class="co">#     # github_url = git_url,</span></span>
<span><span class="co">#     max_dist = 1</span></span>
<span><span class="co">#   )</span></span>
<span><span class="co"># )</span></span>
<span></span>
<span><span class="co"># Local trait data.frame version 2</span></span>
<span><span class="va">btfly_traits3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"species_traits.csv"</span>, package <span class="op">=</span> <span class="st">"invasimap"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># btfly_traits3 = read.csv(system.file("extdata", "species_traits_sim.csv", package = "invasimap"))</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">btfly_traits3</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    27 obs. of  21 variables:</span></span>
<span><span class="co">#&gt;  $ species     : chr  "Acraea horta" "Amata cerbera" "Bicyclus safitza safitza" "Cacyreus lingeus" ...</span></span>
<span><span class="co">#&gt;  $ trait_cont1 : num  0.83 0.874 -0.428 0.661 0.283 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont2 : num  0.811 -0.106 0.672 0.475 0.622 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont3 : num  -0.922 0.498 0.355 -0.657 -0.478 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont4 : num  -0.684 -0.282 0.291 0.552 0.127 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont5 : num  0.0715 -0.9955 0.2179 0.6736 0.503 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont6 : num  0.16 0.643 -0.773 0.529 0.247 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont7 : num  0.2035 -0.606 0.0705 -0.6409 -0.0962 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont8 : num  -0.425 -0.611 0.568 -0.742 -0.742 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont9 : num  0.1493 -0.2933 0.0949 0.7854 -0.02 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont10: num  -0.5772 0.0992 -0.036 -0.6811 -0.7008 ...</span></span>
<span><span class="co">#&gt;  $ trait_cat11 : chr  "wetland" "forest" "wetland" "wetland" ...</span></span>
<span><span class="co">#&gt;  $ trait_cat12 : chr  "diurnal" "nocturnal" "diurnal" "nocturnal" ...</span></span>
<span><span class="co">#&gt;  $ trait_cat13 : chr  "bivoltine" "multivoltine" "univoltine" "multivoltine" ...</span></span>
<span><span class="co">#&gt;  $ trait_cat14 : chr  "detritivore" "detritivore" "generalist" "nectarivore" ...</span></span>
<span><span class="co">#&gt;  $ trait_cat15 : chr  "migratory" "resident" "resident" "migratory" ...</span></span>
<span><span class="co">#&gt;  $ trait_ord16 : int  4 1 4 3 4 1 1 4 1 1 ...</span></span>
<span><span class="co">#&gt;  $ trait_ord17 : int  1 4 4 3 2 4 3 5 4 3 ...</span></span>
<span><span class="co">#&gt;  $ trait_bin18 : int  1 1 1 0 1 1 1 1 0 0 ...</span></span>
<span><span class="co">#&gt;  $ trait_bin19 : int  1 0 1 0 0 1 1 1 0 1 ...</span></span>
<span><span class="co">#&gt;  $ trait_ord20 : chr  "medium" "large" "medium" "medium" ...</span></span>
<span><span class="co"># length(unique(btfly_traits3$species))</span></span>
<span></span>
<span><span class="co"># Retrieve and join trait/metadata for all species in the observation set</span></span>
<span><span class="va">spp_traits</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map_dfr.html" class="external-link">map_dfr</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">grid_obs</span><span class="op">$</span><span class="va">species</span><span class="op">)</span>,</span>
<span>  <span class="op">~</span> <span class="fu"><a href="reference/get_trait_data.html">get_trait_data</a></span><span class="op">(</span></span>
<span>    species <span class="op">=</span> <span class="va">.x</span>,</span>
<span>    n_palette <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    preview <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    do_summary <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    do_taxonomy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    do_image <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    do_palette <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    local_trait_df <span class="op">=</span> <span class="va">btfly_traits3</span>,</span>
<span>    local_species_col <span class="op">=</span> <span class="st">"species"</span>,</span>
<span>    max_dist <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># The final output combines trait data, taxonomic info, Wikipedia summary, images, and color palette for each species.</span></span>
<span><span class="co"># This integrated dataset supports multi-faceted biodiversity, trait, and visualization analyses.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">spp_traits</span><span class="op">)</span></span>
<span><span class="co">#&gt; tibble [27 Ã— 29] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;  $ species     : chr [1:27] "Utetheisa pulchella" "Danaus chrysippus orientis" "Telchinia serena" "Vanessa cardui" ...</span></span>
<span><span class="co">#&gt;  $ summary     : chr [1:27] "Utetheisa pulchella, the crimson-speckled flunkey, crimson-speckled footman, or crimson-speckled moth, is a mot"| __truncated__ NA "Acraea serena, the dancing acraea, is a butterfly of the family Nymphalidae. It is found throughout Africa sout"| __truncated__ "Vanessa cardui is the most widespread of all butterfly species. It is commonly called the painted lady, or form"| __truncated__ ...</span></span>
<span><span class="co">#&gt;  $ Kingdom     : Named chr [1:27] "Animalia" NA "Animalia" "Animalia" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:27] "Kingdom" "Kingdom" "Kingdom" "Kingdom" ...</span></span>
<span><span class="co">#&gt;  $ Phylum      : Named chr [1:27] "Arthropoda" NA "Arthropoda" "Arthropoda" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:27] "Phylum" "Phylum" "Phylum" "Phylum" ...</span></span>
<span><span class="co">#&gt;  $ Class       : Named chr [1:27] "Insecta" NA "Insecta" "Insecta" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:27] "Class" "Class" "Class" "Class" ...</span></span>
<span><span class="co">#&gt;  $ Order       : Named chr [1:27] "Lepidoptera" NA "Lepidoptera" "Lepidoptera" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:27] "Order" "Order" "Order" "Order" ...</span></span>
<span><span class="co">#&gt;  $ Family      : Named chr [1:27] "Erebidae" NA "Nymphalidae" "Nymphalidae" ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:27] "Family" "Family" "Family" "Family" ...</span></span>
<span><span class="co">#&gt;  $ img_url     : chr [1:27] "https://upload.wikimedia.org/wikipedia/commons/thumb/a/a6/Arctiidae_-_Utetheisa_pulchella.JPG/250px-Arctiidae_-"| __truncated__ NA "https://upload.wikimedia.org/wikipedia/commons/thumb/2/2a/Dancing_acraea_%28Acraea_serena%29_underside_Maputo.j"| __truncated__ "https://upload.wikimedia.org/wikipedia/commons/thumb/c/c8/0_Belle-dame_%28Vanessa_cardui%29_-_Echinacea_purpure"| __truncated__ ...</span></span>
<span><span class="co">#&gt;  $ palette     : chr [1:27] "#535509, #A59A47, #8C8012, #1D220C, #CCBF98" NA "#9C9C6D, #86885D, #E8E6CE, #534832, #B4862D" "#CC8242, #EA8FD3, #3A311F, #C65D9B, #6A5C42" ...</span></span>
<span><span class="co">#&gt;  $ trait_cont1 : num [1:27] 0.0284 0.0382 -0.8351 -0.2196 0.314 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont2 : num [1:27] -0.203 -0.224 -0.307 0.569 0.666 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont3 : num [1:27] -0.9969 0.0288 0.0288 0.1632 0.5191 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont4 : num [1:27] 0.48 -0.533 0.925 0.466 -0.39 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont5 : num [1:27] 0.8748 -0.0945 0.263 0.701 -0.9972 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont6 : num [1:27] 0.87 -0.703 0.36 0.101 0.559 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont7 : num [1:27] 0.586 -0.366 0.836 -0.733 0.459 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont8 : num [1:27] -0.0974 -0.8555 -0.8781 0.6775 -0.7754 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont9 : num [1:27] 0.123 -0.657 -0.865 -0.859 -0.373 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont10: num [1:27] 0.85009 -0.00145 -0.5899 0.77351 -0.62313 ...</span></span>
<span><span class="co">#&gt;  $ trait_cat11 : chr [1:27] "wetland" "grassland" "forest" "forest" ...</span></span>
<span><span class="co">#&gt;  $ trait_cat12 : chr [1:27] "diurnal" "diurnal" "nocturnal" "diurnal" ...</span></span>
<span><span class="co">#&gt;  $ trait_cat13 : chr [1:27] "multivoltine" "univoltine" "multivoltine" "univoltine" ...</span></span>
<span><span class="co">#&gt;  $ trait_cat14 : chr [1:27] "detritivore" "detritivore" "generalist" "generalist" ...</span></span>
<span><span class="co">#&gt;  $ trait_cat15 : chr [1:27] "migratory" "migratory" "migratory" "resident" ...</span></span>
<span><span class="co">#&gt;  $ trait_ord16 : int [1:27] 3 1 2 3 1 1 4 4 2 2 ...</span></span>
<span><span class="co">#&gt;  $ trait_ord17 : int [1:27] 4 4 5 3 4 2 1 5 2 5 ...</span></span>
<span><span class="co">#&gt;  $ trait_bin18 : int [1:27] 1 1 1 0 0 0 1 1 0 0 ...</span></span>
<span><span class="co">#&gt;  $ trait_bin19 : int [1:27] 1 1 1 0 0 0 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;  $ trait_ord20 : chr [1:27] "small" "medium" "large" "large" ...</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">spp_traits</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 29</span></span>
<span><span class="co">#&gt;   species  summary Kingdom Phylum Class Order Family img_url palette trait_cont1</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;    &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;         &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 Utetheiâ€¦ Utetheâ€¦ Animalâ€¦ Arthrâ€¦ Inseâ€¦ Lepiâ€¦ Erebiâ€¦ https:â€¦ #53550â€¦      0.0284</span></span>
<span><span class="co">#&gt; 2 Danaus â€¦ &lt;NA&gt;    &lt;NA&gt;    &lt;NA&gt;   &lt;NA&gt;  &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt;    &lt;NA&gt;         0.0382</span></span>
<span><span class="co">#&gt; 3 Telchinâ€¦ Acraeaâ€¦ Animalâ€¦ Arthrâ€¦ Inseâ€¦ Lepiâ€¦ Nymphâ€¦ https:â€¦ #9C9C6â€¦     -0.835 </span></span>
<span><span class="co">#&gt; 4 Vanessaâ€¦ Vanessâ€¦ Animalâ€¦ Arthrâ€¦ Inseâ€¦ Lepiâ€¦ Nymphâ€¦ https:â€¦ #CC824â€¦     -0.220 </span></span>
<span><span class="co">#&gt; 5 Hypolimâ€¦ Hypoliâ€¦ Animalâ€¦ Arthrâ€¦ Inseâ€¦ Lepiâ€¦ Nymphâ€¦ https:â€¦ #B2A79â€¦      0.314 </span></span>
<span><span class="co">#&gt; 6 Pieris â€¦ Pierisâ€¦ Animalâ€¦ Arthrâ€¦ Inseâ€¦ Lepiâ€¦ Pieriâ€¦ https:â€¦ #515A2â€¦     -0.765 </span></span>
<span><span class="co">#&gt; # â„¹ 19 more variables: trait_cont2 &lt;dbl&gt;, trait_cont3 &lt;dbl&gt;, trait_cont4 &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   trait_cont5 &lt;dbl&gt;, trait_cont6 &lt;dbl&gt;, trait_cont7 &lt;dbl&gt;, trait_cont8 &lt;dbl&gt;,</span></span>
<span><span class="co">#&gt; #   trait_cont9 &lt;dbl&gt;, trait_cont10 &lt;dbl&gt;, trait_cat11 &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   trait_cat12 &lt;chr&gt;, trait_cat13 &lt;chr&gt;, trait_cat14 &lt;chr&gt;, trait_cat15 &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   trait_ord16 &lt;int&gt;, trait_ord17 &lt;int&gt;, trait_bin18 &lt;int&gt;, trait_bin19 &lt;int&gt;,</span></span>
<span><span class="co">#&gt; #   trait_ord20 &lt;chr&gt;</span></span>
<span></span>
<span><span class="co"># Count how many nonâ€NA IDs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">btfly_traits3</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 27</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">grid_obs</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 27</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">spp_traits</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 27</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="id_42-alternatively-load-local-combined-site-environment-and-trait-data">4.2. Alternatively, load local combined site, environment, and trait data<a class="anchor" aria-label="anchor" href="#id_42-alternatively-load-local-combined-site-environment-and-trait-data"></a>
</h4>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read GBIF species occurrence with simulated traits and enviro data (one row per site-species combination)</span></span>
<span><span class="va">site_env_spp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"site_env_spp_simulated.csv"</span>, package <span class="op">=</span> <span class="st">"invasimap"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># site_env_spp = read.csv(system.file("extdata", "site_env_spp_trt_sim.csv", package = "invasimap"))</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">site_env_spp</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 11205    36</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">site_env_spp</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    11205 obs. of  36 variables:</span></span>
<span><span class="co">#&gt;  $ site_id     : int  1026 1026 1026 1026 1026 1026 1026 1026 1026 1026 ...</span></span>
<span><span class="co">#&gt;  $ x           : num  28.8 28.8 28.8 28.8 28.8 ...</span></span>
<span><span class="co">#&gt;  $ y           : num  -22.3 -22.3 -22.3 -22.3 -22.3 ...</span></span>
<span><span class="co">#&gt;  $ species     : chr  "Acraea horta" "Amata cerbera" "Bicyclus safitza safitza" "Cacyreus lingeus" ...</span></span>
<span><span class="co">#&gt;  $ count       : int  10 0 0 0 9 8 8 3 19 0 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont1 : num  0.83 0.874 -0.428 0.661 0.283 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont2 : num  0.811 -0.106 0.672 0.475 0.622 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont3 : num  -0.922 0.498 0.355 -0.657 -0.478 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont4 : num  -0.684 -0.282 0.291 0.552 0.127 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont5 : num  0.0715 -0.9955 0.2179 0.6736 0.503 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont6 : num  0.16 0.643 -0.773 0.529 0.247 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont7 : num  0.2035 -0.606 0.0705 -0.6409 -0.0962 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont8 : num  -0.425 -0.611 0.568 -0.742 -0.742 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont9 : num  0.1493 -0.2933 0.0949 0.7854 -0.02 ...</span></span>
<span><span class="co">#&gt;  $ trait_cont10: num  -0.5772 0.0992 -0.036 -0.6811 -0.7008 ...</span></span>
<span><span class="co">#&gt;  $ trait_cat11 : chr  "wetland" "forest" "wetland" "wetland" ...</span></span>
<span><span class="co">#&gt;  $ trait_cat12 : chr  "diurnal" "nocturnal" "diurnal" "nocturnal" ...</span></span>
<span><span class="co">#&gt;  $ trait_cat13 : chr  "bivoltine" "multivoltine" "univoltine" "multivoltine" ...</span></span>
<span><span class="co">#&gt;  $ trait_cat14 : chr  "detritivore" "detritivore" "generalist" "nectarivore" ...</span></span>
<span><span class="co">#&gt;  $ trait_cat15 : chr  "migratory" "resident" "resident" "migratory" ...</span></span>
<span><span class="co">#&gt;  $ trait_ord16 : int  4 1 4 3 4 1 1 4 1 1 ...</span></span>
<span><span class="co">#&gt;  $ trait_ord17 : int  1 4 4 3 2 4 3 5 4 3 ...</span></span>
<span><span class="co">#&gt;  $ trait_bin18 : int  1 1 1 0 1 1 1 1 0 0 ...</span></span>
<span><span class="co">#&gt;  $ trait_bin19 : int  1 0 1 0 0 1 1 1 0 1 ...</span></span>
<span><span class="co">#&gt;  $ trait_ord20 : chr  "medium" "large" "medium" "medium" ...</span></span>
<span><span class="co">#&gt;  $ env1        : num  2.2 2.2 2.2 2.2 2.2 ...</span></span>
<span><span class="co">#&gt;  $ env2        : num  0.647 0.647 0.647 0.647 0.647 ...</span></span>
<span><span class="co">#&gt;  $ env3        : num  -0.491 -0.491 -0.491 -0.491 -0.491 ...</span></span>
<span><span class="co">#&gt;  $ env4        : num  -0.793 -0.793 -0.793 -0.793 -0.793 ...</span></span>
<span><span class="co">#&gt;  $ env5        : num  0.822 0.822 0.822 0.822 0.822 ...</span></span>
<span><span class="co">#&gt;  $ env6        : num  1.55 1.55 1.55 1.55 1.55 ...</span></span>
<span><span class="co">#&gt;  $ env7        : num  0.419 0.419 0.419 0.419 0.419 ...</span></span>
<span><span class="co">#&gt;  $ env8        : num  -1.05 -1.05 -1.05 -1.05 -1.05 ...</span></span>
<span><span class="co">#&gt;  $ env9        : num  -0.0537 -0.0537 -0.0537 -0.0537 -0.0537 ...</span></span>
<span><span class="co">#&gt;  $ env10       : num  0.933 0.933 0.933 0.933 0.933 ...</span></span>
<span><span class="co">#&gt;  $ cat11_num   : int  -1 1 -1 -1 1 0 1 -1 1 0 ...</span></span>
<span></span>
<span><span class="co"># Check the results</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">grid_obs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "site_id"  "x"        "y"        "obs_sum"  "spp_rich" "species"  "count"</span></span>
<span><span class="co"># names(grid_env)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="id_5-model-inputs">5. Model Inputs<a class="anchor" aria-label="anchor" href="#id_5-model-inputs"></a>
</h3>
<p>Shape your data so every row is â€œone species at one site,â€ with that speciesâ€™ traits and that siteâ€™s environment.</p>
<div class="section level4">
<h4 id="id_51-format-site-locations">5.1. Format <strong>site-locations</strong>
<a class="anchor" aria-label="anchor" href="#id_51-format-site-locations"></a>
</h4>
<p>This section isolates the unique spatial coordinates for each sampling site. The resulting table (<code>site_xy</code>) will be used for spatial mapping, distance calculations, and for merging environmental and biodiversity metrics with precise locations.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create site coordinate table i.e. # Unique site coordinates</span></span>
<span><span class="va">site_xy</span> <span class="op">&lt;-</span> <span class="va">site_env_spp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">site_id</span>, <span class="va">x</span>, <span class="va">y</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>.site_id_rn <span class="op">=</span> <span class="va">site_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">".site_id_rn"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">site_xy</span><span class="op">)</span></span>
<span><span class="co">#&gt;      site_id     x         y</span></span>
<span><span class="co">#&gt; 1026    1026 28.75 -22.25004</span></span>
<span><span class="co">#&gt; 1027    1027 29.25 -22.25004</span></span>
<span><span class="co">#&gt; 1028    1028 29.75 -22.25004</span></span>
<span><span class="co">#&gt; 1029    1029 30.25 -22.25004</span></span>
<span><span class="co">#&gt; 1030    1030 30.75 -22.25004</span></span>
<span><span class="co">#&gt; 1031    1031 31.25 -22.25004</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="id_52-format-site-environment-variables">5.2. Format <strong>site-environment</strong> variables<a class="anchor" aria-label="anchor" href="#id_52-format-site-environment-variables"></a>
</h4>
<p>Here, we extract a site-by-environment matrix containing the values of all measured environmental covariates at each sampling site. This matrix (<code>site_env</code>) enables analyses of environmental gradients, spatial drivers of community composition, and covariate modeling.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Site-by-environment matrix</span></span>
<span><span class="va">site_env</span> <span class="op">&lt;-</span> <span class="va">site_env_spp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>    <span class="va">site_id</span>, <span class="va">x</span>, <span class="va">y</span>,</span>
<span>    <span class="va">env1</span><span class="op">:</span><span class="va">env10</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>site_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">site_id</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># ensure character</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>.site_id_rn <span class="op">=</span> <span class="va">site_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">".site_id_rn"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">site_env</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 415  13</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">site_env</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;      site_id     x         y     env1      env2       env3</span></span>
<span><span class="co">#&gt; 1026    1026 28.75 -22.25004 2.203029 0.6471631 -0.4910981</span></span>
<span><span class="co">#&gt; 1027    1027 29.25 -22.25004 2.086006 1.4025519 -0.4471106</span></span>
<span><span class="co">#&gt; 1028    1028 29.75 -22.25004 2.233508 0.8008731 -0.5405243</span></span>
<span><span class="co">#&gt; 1029    1029 30.25 -22.25004 2.333375 1.1607272 -0.4405388</span></span>
<span><span class="co">#&gt; 1030    1030 30.75 -22.25004 2.153073 1.2747649 -0.2945477</span></span>
<span><span class="co">#&gt; 1031    1031 31.25 -22.25004 2.046307 1.4773531 -0.4125693</span></span>
<span></span>
<span><span class="co"># site_env = grid_env %&gt;%</span></span>
<span><span class="co">#   dplyr::select(site_id, x, y,</span></span>
<span><span class="co">#                 obs_sum, spp_rich,</span></span>
<span><span class="co">#                 bio01:bio19) %&gt;%</span></span>
<span><span class="co">#   distinct() %&gt;%</span></span>
<span><span class="co">#   mutate(.site_id_rn = site_id) %&gt;%</span></span>
<span><span class="co">#   column_to_rownames(var = ".site_id_rn")</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># dim(site_env)</span></span>
<span><span class="co"># head(site_env[1:6,1:6])</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="id_53-format-site-species-abundances-and-presence-absence">5.3. Format <strong>site-species</strong> abundances and presence-absence<a class="anchor" aria-label="anchor" href="#id_53-format-site-species-abundances-and-presence-absence"></a>
</h4>
<p>This section generates two site-by-species matrices: one containing abundances (<code>site_spp_ab</code>), and one indicating presence-absence (<code>site_spp_pa</code>). These matrices are fundamental for calculating community diversity, richness, and for modeling occupancy and abundance patterns.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Site-by-species abundance matrix (wide format)</span></span>
<span><span class="co"># site_spp_ab = grid_obs %&gt;%</span></span>
<span><span class="va">site_spp_ab</span> <span class="op">&lt;-</span> <span class="va">site_env_spp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co">#</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">site_id</span>, <span class="va">x</span>, <span class="va">y</span>, <span class="va">species</span>, <span class="va">count</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span></span>
<span>    names_from  <span class="op">=</span> <span class="va">species</span>,</span>
<span>    values_from <span class="op">=</span> <span class="va">count</span>,</span>
<span>    values_fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>count <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>.site_id_rn <span class="op">=</span> <span class="va">site_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">".site_id_rn"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">site_spp_ab</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 415  30</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">site_spp_ab</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;      site_id     x         y Acraea horta Amata cerbera</span></span>
<span><span class="co">#&gt; 1026    1026 28.75 -22.25004           10             0</span></span>
<span><span class="co">#&gt; 1027    1027 29.25 -22.25004            0             7</span></span>
<span><span class="co">#&gt; 1028    1028 29.75 -22.25004            0             0</span></span>
<span><span class="co">#&gt; 1029    1029 30.25 -22.25004            0            31</span></span>
<span><span class="co">#&gt; 1030    1030 30.75 -22.25004            0            12</span></span>
<span><span class="co">#&gt; 1031    1031 31.25 -22.25004            0             7</span></span>
<span><span class="co">#&gt;      Bicyclus safitza safitza</span></span>
<span><span class="co">#&gt; 1026                        0</span></span>
<span><span class="co">#&gt; 1027                        0</span></span>
<span><span class="co">#&gt; 1028                        0</span></span>
<span><span class="co">#&gt; 1029                        0</span></span>
<span><span class="co">#&gt; 1030                        3</span></span>
<span><span class="co">#&gt; 1031                        0</span></span>
<span></span>
<span><span class="co"># Site-by-species presence/absence matrix (wide format)</span></span>
<span><span class="co"># site_spp_pa = grid_obs %&gt;%</span></span>
<span><span class="va">site_spp_pa</span> <span class="op">&lt;-</span> <span class="va">site_env_spp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>pa <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="va">count</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">site_id</span>, <span class="va">x</span>, <span class="va">y</span>, <span class="va">species</span>, <span class="va">pa</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span></span>
<span>    names_from  <span class="op">=</span> <span class="va">species</span>,</span>
<span>    values_from <span class="op">=</span> <span class="va">pa</span>,</span>
<span>    values_fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>pa <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>.site_id_rn <span class="op">=</span> <span class="va">site_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">".site_id_rn"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">site_spp_pa</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 415  30</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">site_spp_pa</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;      site_id     x         y Acraea horta Amata cerbera</span></span>
<span><span class="co">#&gt; 1026    1026 28.75 -22.25004            1             0</span></span>
<span><span class="co">#&gt; 1027    1027 29.25 -22.25004            0             1</span></span>
<span><span class="co">#&gt; 1028    1028 29.75 -22.25004            0             0</span></span>
<span><span class="co">#&gt; 1029    1029 30.25 -22.25004            0             1</span></span>
<span><span class="co">#&gt; 1030    1030 30.75 -22.25004            0             1</span></span>
<span><span class="co">#&gt; 1031    1031 31.25 -22.25004            0             1</span></span>
<span><span class="co">#&gt;      Bicyclus safitza safitza</span></span>
<span><span class="co">#&gt; 1026                        0</span></span>
<span><span class="co">#&gt; 1027                        0</span></span>
<span><span class="co">#&gt; 1028                        0</span></span>
<span><span class="co">#&gt; 1029                        0</span></span>
<span><span class="co">#&gt; 1030                        1</span></span>
<span><span class="co">#&gt; 1031                        0</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="id_54-format-species-trait-values">5.4. Format <strong>species-trait</strong> values<a class="anchor" aria-label="anchor" href="#id_54-format-species-trait-values"></a>
</h4>
<p>Here we build the species-by-trait matrix (<code>spp_trait</code>), including all measured continuous, categorical, and ordinal traits for each species. This structure is central for trait-based analyses of community assembly, functional diversity, and invasion processes.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Species-by-trait matrix (wide)</span></span>
<span><span class="co"># Extract and process continuous, categorical, and ordinal trait data</span></span>
<span><span class="va">spp_trait</span> <span class="op">&lt;-</span> <span class="va">spp_traits</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># site_env_spp</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>    <span class="va">species</span>, <span class="va">trait_cont1</span><span class="op">:</span><span class="va">trait_cont10</span>,</span>
<span>    <span class="va">trait_cat11</span><span class="op">:</span><span class="va">trait_cat15</span>,</span>
<span>    <span class="va">trait_ord16</span><span class="op">:</span><span class="va">trait_ord20</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>.species_rn <span class="op">=</span> <span class="va">species</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">".species_rn"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html" class="external-link">where</a></span><span class="op">(</span><span class="va">is.character</span><span class="op">)</span>, <span class="va">as.factor</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># spp_trait = spp_traits %&gt;% # site_env_spp</span></span>
<span><span class="co">#   dplyr::select(species, trt1:trt20) %&gt;%</span></span>
<span><span class="co">#   distinct() %&gt;%</span></span>
<span><span class="co">#   mutate(.species_rn = species) %&gt;%</span></span>
<span><span class="co">#   column_to_rownames(var = ".species_rn") %&gt;%</span></span>
<span><span class="co">#   mutate(across(where(is.character), as.factor))</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">spp_trait</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 27 21</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">spp_trait</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                               species trait_cont1 trait_cont2</span></span>
<span><span class="co">#&gt; Utetheisa pulchella               Utetheisa pulchella  0.02842357  -0.2030292</span></span>
<span><span class="co">#&gt; Danaus chrysippus orientis Danaus chrysippus orientis  0.03819190  -0.2237834</span></span>
<span><span class="co">#&gt; Telchinia serena                     Telchinia serena -0.83512488  -0.3065035</span></span>
<span><span class="co">#&gt; Vanessa cardui                         Vanessa cardui -0.21959307   0.5693856</span></span>
<span><span class="co">#&gt; Hypolimnas misippus               Hypolimnas misippus  0.31398458   0.6658322</span></span>
<span><span class="co">#&gt; Pieris brassicae                     Pieris brassicae -0.76502528  -0.1364975</span></span>
<span><span class="co">#&gt;                            trait_cont3 trait_cont4 trait_cont5</span></span>
<span><span class="co">#&gt; Utetheisa pulchella        -0.99685889   0.4797106  0.87477170</span></span>
<span><span class="co">#&gt; Danaus chrysippus orientis  0.02882587  -0.5325932 -0.09453685</span></span>
<span><span class="co">#&gt; Telchinia serena            0.02881542   0.9252160  0.26301460</span></span>
<span><span class="co">#&gt; Vanessa cardui              0.16320801   0.4664918  0.70096550</span></span>
<span><span class="co">#&gt; Hypolimnas misippus         0.51908854  -0.3895633 -0.99723831</span></span>
<span><span class="co">#&gt; Pieris brassicae           -0.71904181   0.4879493 -0.21005391</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="id_6-data-summaries-and-visualisation">6. Data summaries and visualisation<a class="anchor" aria-label="anchor" href="#id_6-data-summaries-and-visualisation"></a>
</h3>
<div class="section level4">
<h4 id="id_61-summarise-site-level-diversity">6.1. Summarise site-level diversity<a class="anchor" aria-label="anchor" href="#id_61-summarise-site-level-diversity"></a>
</h4>
<p>This section quantifies and visualizes site-level biodiversity, focusing on local species richness and abundance. Calculating these metrics is essential for mapping alpha diversity, assessing community structure, and identifying spatial patterns of biodiversity hotspots and low-diversity areas across the study landscape.</p>
<ul>
<li>
<strong>Species richness</strong> (spp_rich<sub>s</sub>): the number of species present (non-zero counts) at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.</li>
<li>
<strong>Total abundance</strong> (obs_sum<sub>s</sub>): the sum of all individual counts across species at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math> (a proxy for sampling effort).</li>
<li>
<strong>Mean abundance per species</strong> (obs_mean<sub>s</sub>): total abundance at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math> divided by the number of species columns (N); effectively the average count per species regardless of whether it is present.</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate site-level diversity metrics from the species-by-abundance matrix:</span></span>
<span><span class="va">spp_rich_obs</span> <span class="op">&lt;-</span> <span class="va">site_spp_ab</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    spp_rich <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.</span>, <span class="op">-</span><span class="va">site_id</span>, <span class="op">-</span><span class="va">x</span>, <span class="op">-</span><span class="va">y</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>, <span class="co"># Species richness: number of species present</span></span>
<span>    obs_sum <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.</span>, <span class="op">-</span><span class="va">site_id</span>, <span class="op">-</span><span class="va">x</span>, <span class="op">-</span><span class="va">y</span><span class="op">)</span><span class="op">)</span>, <span class="co"># Total abundance: sum of all individuals</span></span>
<span>    obs_mean <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/rowSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.</span>, <span class="op">-</span><span class="va">site_id</span>, <span class="op">-</span><span class="va">x</span>, <span class="op">-</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="co"># Mean abundance per species</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># Keep summary metrics and site coordinates</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">site_id</span>, <span class="va">x</span>, <span class="va">y</span>, <span class="va">spp_rich</span>, <span class="va">obs_sum</span>, <span class="va">obs_mean</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>site_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">site_id</span><span class="op">)</span><span class="op">)</span> <span class="co"># Ensure site_id` is a</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">spp_rich_obs</span><span class="op">)</span></span>
<span><span class="co">#&gt;      site_id     x         y spp_rich obs_sum obs_mean</span></span>
<span><span class="co">#&gt; 1026    1026 28.75 -22.25004       17     172 6.370370</span></span>
<span><span class="co">#&gt; 1027    1027 29.25 -22.25004        9      92 3.407407</span></span>
<span><span class="co">#&gt; 1028    1028 29.75 -22.25004       11     131 4.851852</span></span>
<span><span class="co">#&gt; 1029    1029 30.25 -22.25004       12     129 4.777778</span></span>
<span><span class="co">#&gt; 1030    1030 30.75 -22.25004       13     136 5.037037</span></span>
<span><span class="co">#&gt; 1031    1031 31.25 -22.25004       13      99 3.666667</span></span>
<span></span>
<span><span class="co"># Define a custom color palette for richness mapping (blue = low, dark red = high)</span></span>
<span><span class="va">col_pal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"green"</span>, <span class="st">"yellow"</span>, <span class="st">"orange"</span>, <span class="st">"red"</span>, <span class="st">"darkred"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize spatial distribution of site-level species richness</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">spp_rich_obs</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">spp_rich</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># Use custom color gradient, reversed so high richness is warm/dark, low is cool/blue</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/flip.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu">col_pal</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"âˆš(Richness)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">spp_rich</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"grey80"</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> <span class="co"># Overlay actual richness values</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">rsa</span>, inherit.aes <span class="op">=</span> <span class="cn">FALSE</span>, fill <span class="op">=</span> <span class="cn">NA</span>, color <span class="op">=</span> <span class="st">"black"</span>, size <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op">+</span> <span class="co"># Plot boundary</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"Longitude"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Latitude"</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"Spatial Distribution of Species Richness"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-site-richness-1.png" width="100%" style="display: block; margin: auto;"></p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="id_7-functional-trait-space">7. Functional Trait Space<a class="anchor" aria-label="anchor" href="#id_7-functional-trait-space"></a>
</h2>
<div class="section level3">
<h3 id="id_71-basic-trait-similarity">7.1. Basic trait similarity<a class="anchor" aria-label="anchor" href="#id_71-basic-trait-similarity"></a>
</h3>
<p>To diagnose which functional dimensions are more conserved versus variable across the metacommunity, we compute <strong>trait-level similarity</strong> for each trait across all species. This allows identification of traits that might constrain or facilitate invasion and coexistence (e.g., highly conserved traits might reflect strong filtering, while highly variable traits may be axes of ecological opportunity).</p>
<p>We use the <code><a href="reference/compute_trait_similarity.html">compute_trait_similarity()</a></code> function, which calculates similarity as follows:</p>
<ul>
<li>
<strong>Numeric traits</strong>: Scaled to [0,1], pairwise Euclidean distances are computed, and similarity is 1 - mean(distance). If all values are identical or only one value is present, similarity is 100%.</li>
<li>
<strong>Categorical traits</strong>: Similarity is the proportion of all possible species pairs that share the same category (level).</li>
</ul>
<p>The output is a table of percent similarity for each trait, allowing direct comparison of conservation vs.Â lability across traits.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute Trait Similarity for Numeric and Categorical Variables</span></span>
<span><span class="va">df_traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/compute_trait_similarity.html">compute_trait_similarity</a></span><span class="op">(</span><span class="va">spp_trait</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">df_traits</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 Ã— 2</span></span>
<span><span class="co">#&gt;   Trait       Similarity</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;            &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 trait_cont1       61.8</span></span>
<span><span class="co">#&gt; 2 trait_cont2       63.9</span></span>
<span><span class="co">#&gt; 3 trait_cont3       65.1</span></span>
<span><span class="co">#&gt; 4 trait_cont4       64.1</span></span>
<span><span class="co">#&gt; 5 trait_cont5       69.5</span></span>
<span><span class="co">#&gt; 6 trait_cont6       62.5</span></span>
<span></span>
<span><span class="co"># Barplot: trait-level similarity (percent identity or scaled distance)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">df_traits</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">Trait</span>, <span class="va">Similarity</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">Similarity</span>, fill <span class="op">=</span> <span class="va">Similarity</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span>show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"inferno"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># ramp color scale</span></span>
<span>  <span class="co"># ylim(0,100) +</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Average Trait Similarity (%)"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Similarity (%)"</span>,</span>
<span>    x <span class="op">=</span> <span class="cn">NULL</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="reference/figures/README-trait-sim-barplot-1.png" alt="Trait-level functional similarity across species." width="100%"><p class="caption">
Trait-level functional similarity across species.
</p>
</div>
</div>
<div class="section level3">
<h3 id="id_72-gower-distance-clustering-and-trait-space-mapping-pcoa">7.2. Gower distance, clustering, and trait space mapping (PCoA)<a class="anchor" aria-label="anchor" href="#id_72-gower-distance-clustering-and-trait-space-mapping-pcoa"></a>
</h3>
<div class="section level4">
<h4 id="id_721-compute-gower-distance-handles-mixed-trait-types">7.2.1. Compute Gower distance (handles mixed trait types)<a class="anchor" aria-label="anchor" href="#id_721-compute-gower-distance-handles-mixed-trait-types"></a>
</h4>
<p>Trait-based approaches require robust dissimilarity measures for mixed data types (continuous, categorical, ordinal). Here, we compute <strong>pairwise Gower distances</strong> among species, which accommodates all variable types, and use hierarchical clustering to visualize functional similarity structure within the community.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute Gower dissimilarity matrix (excluding species column)</span></span>
<span><span class="va">sbt_gower</span> <span class="op">&lt;-</span> <span class="fu">cluster</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cluster/man/daisy.html" class="external-link">daisy</a></span><span class="op">(</span><span class="va">spp_trait</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span>, metric <span class="op">=</span> <span class="st">"gower"</span><span class="op">)</span></span>
<span><span class="va">trait_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/coerce.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">sbt_gower</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="id_722-hierarchical-clustering-visualise-trait-based-groupings">7.2.2. Hierarchical clustering (visualise trait-based groupings)<a class="anchor" aria-label="anchor" href="#id_722-hierarchical-clustering-visualise-trait-based-groupings"></a>
</h4>
<p><code>{r  hclust-traits, fig.cap="Species clustering by functional traits (Gower distance, hierarchical clustering)." # Hierarchical clustering and dendrogram visualization of functional similarity # Hierarchical clustering gower_hc = hclust(as.dist(sbt_gower)) # Dendrogram fviz_dend(   gower_hc,   k = 4,   cex = 0.5,   k_colors = viridis(4, option = "D"), # k_colors = c("red","blue","green","purple"),   color_labels_by_k = TRUE,   rect = TRUE,   rect_border = "grey40",   main = "Gower Cluster Dendrogram") +    guides(scale = "none")</code></p>
</div>
<div class="section level4">
<h4 id="id_723-pcoa-ordination-map-species-into-a-reduced-dimensional-trait-space">7.2.3. PCoA ordination (map species into a reduced-dimensional trait space)<a class="anchor" aria-label="anchor" href="#id_723-pcoa-ordination-map-species-into-a-reduced-dimensional-trait-space"></a>
</h4>
<p>Principal Coordinates Analysis (PCoA) enables <strong>ordination</strong> of species in a reduced, low-dimensional trait space, preserving pairwise dissimilarities. This is used to visualize the overall structure of the functional trait space and examine density and clustering patterns.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># PCoA ordination</span></span>
<span><span class="va">pcoa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cmdscale.html" class="external-link">cmdscale</a></span><span class="op">(</span><span class="va">sbt_gower</span>, eig <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">scores_species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">pcoa</span><span class="op">$</span><span class="va">points</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">scores_species</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PCoA1"</span>, <span class="st">"PCoA2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize trait space density using kernel density estimation</span></span>
<span><span class="va">xlims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">scores_species</span><span class="op">$</span><span class="va">PCoA1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="fl">0.1</span> <span class="op">*</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">scores_species</span><span class="op">$</span><span class="va">PCoA1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ylims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">scores_species</span><span class="op">$</span><span class="va">PCoA2</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span> <span class="op">*</span> <span class="fl">0.1</span> <span class="op">*</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">scores_species</span><span class="op">$</span><span class="va">PCoA2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">grid_density</span> <span class="op">&lt;-</span> <span class="fu">MASS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MASS/man/kde2d.html" class="external-link">kde2d</a></span><span class="op">(</span><span class="va">scores_species</span><span class="op">$</span><span class="va">PCoA1</span>,</span>
<span>  <span class="va">scores_species</span><span class="op">$</span><span class="va">PCoA2</span>,</span>
<span>  n <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  lims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">xlims</span>, <span class="va">ylims</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/filled.contour.html" class="external-link">filled.contour</a></span><span class="op">(</span></span>
<span>  <span class="va">grid_density</span>,</span>
<span>  color.palette <span class="op">=</span> <span class="va">viridis</span>,</span>
<span>  xlim <span class="op">=</span> <span class="va">xlims</span>, ylim <span class="op">=</span> <span class="va">ylims</span>,</span>
<span>  plot.title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span></span>
<span>    main <span class="op">=</span> <span class="st">"Trait Space Density Contours"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"PCoA1"</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"PCoA2"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  plot.axes <span class="op">=</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/axis.html" class="external-link">axis</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/axis.html" class="external-link">axis</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rspatial.github.io/terra/reference/lines.html" class="external-link">points</a></span><span class="op">(</span><span class="va">scores_species</span>, pch <span class="op">=</span> <span class="fl">19</span>, cex <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span>    <span class="co"># Draw all contours (thin)</span></span>
<span>    <span class="fu"><a href="https://rspatial.github.io/terra/reference/contour.html" class="external-link">contour</a></span><span class="op">(</span></span>
<span>      x <span class="op">=</span> <span class="va">grid_density</span><span class="op">$</span><span class="va">x</span>, y <span class="op">=</span> <span class="va">grid_density</span><span class="op">$</span><span class="va">y</span>, z <span class="op">=</span> <span class="va">grid_density</span><span class="op">$</span><span class="va">z</span>,</span>
<span>      add <span class="op">=</span> <span class="cn">TRUE</span>, drawlabels <span class="op">=</span> <span class="cn">FALSE</span>, lwd <span class="op">=</span> <span class="fl">0.7</span>, col <span class="op">=</span> <span class="st">"grey60"</span></span>
<span>    <span class="op">)</span></span>
<span>    <span class="co"># Highlight the major contour (e.g. highest density level)</span></span>
<span>    <span class="fu"><a href="https://rspatial.github.io/terra/reference/contour.html" class="external-link">contour</a></span><span class="op">(</span></span>
<span>      x <span class="op">=</span> <span class="va">grid_density</span><span class="op">$</span><span class="va">x</span>, y <span class="op">=</span> <span class="va">grid_density</span><span class="op">$</span><span class="va">y</span>, z <span class="op">=</span> <span class="va">grid_density</span><span class="op">$</span><span class="va">z</span>,</span>
<span>      add <span class="op">=</span> <span class="cn">TRUE</span>, drawlabels <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>      levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">grid_density</span><span class="op">$</span><span class="va">z</span><span class="op">)</span> <span class="op">*</span> <span class="fl">0.5</span>, <span class="co"># 50% of max density</span></span>
<span>      lwd <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"black"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">}</span>,</span>
<span>  key.title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span>main <span class="op">=</span> <span class="st">"Density"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="reference/figures/README-pcoa-traitspace-1.png" alt="Kernel density in trait space (PCoA axes 1-2)." width="100%"><p class="caption">
Kernel density in trait space (PCoA axes 1-2).
</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="id_73-trait-centrality">7.3. Trait centrality<a class="anchor" aria-label="anchor" href="#id_73-trait-centrality"></a>
</h3>
<p>Trait <strong>centrality</strong> quantifies how close each species is to the â€œcoreâ€ of the communityâ€™s trait space. Peripheral species may be ecologically distinct and potentially more likely to become invaders or to escape biotic resistance.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate the community trait centroid in reduced trait-space (PCoA axes)</span></span>
<span><span class="va">centroid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/rowSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">scores_species</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute each species' Euclidean distance to the centroid (trait centrality)</span></span>
<span><span class="va">scores_species</span><span class="op">$</span><span class="va">centrality</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="op">(</span><span class="va">scores_species</span> <span class="op">-</span> <span class="va">centroid</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add centrality to the main trait data frame for further analysis/plotting</span></span>
<span><span class="va">spp_trt_cent</span> <span class="op">&lt;-</span> <span class="va">spp_trait</span></span>
<span><span class="va">spp_trt_cent</span><span class="op">$</span><span class="va">centrality</span> <span class="op">&lt;-</span> <span class="va">scores_species</span><span class="op">$</span><span class="va">centrality</span></span>
<span></span>
<span><span class="co"># Histogram of distribution of trait centrality (core vs peripheral species)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">spp_trt_cent</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">centrality</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">20</span>, fill <span class="op">=</span> <span class="st">"steelblue"</span>, color <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="st">"Distance to community-centroid"</span>, y <span class="op">=</span> <span class="st">"Number of species"</span>,</span>
<span>    title <span class="op">=</span> <span class="st">"Trait Centrality (Community Edge vs Core)"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="reference/figures/README-trait-centrality-1.png" alt="Distribution of trait centrality (distance to centroid) among species." width="100%"><p class="caption">
Distribution of trait centrality (distance to centroid) among species.
</p>
</div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># # OPTIONAL</span></span>
<span><span class="co"># # Scatterplot of trait-space (PCoA1 vs PCoA2), coloured by trait centrality</span></span>
<span><span class="co"># ggplot(scores_species, aes(x = PCoA1, y = PCoA2, colour = centrality)) +</span></span>
<span><span class="co">#   geom_point(size = 3) +</span></span>
<span><span class="co">#   scale_colour_viridis_c(option = "magma") +</span></span>
<span><span class="co">#   labs(colour = "Centrality",</span></span>
<span><span class="co">#        title = "Species Position in Trait Space (PCoA axes 1-2)")</span></span></code></pre></div>
<blockquote>
<p>This histogram shows how far each species is from the <strong>center</strong> of the communityâ€™s trait space.</p>
<ul>
<li>The <strong>x-axis</strong> is <em>distance to the centroid</em> (central trait combination of the community).</li>
<li>The <strong>y-axis</strong> is the number of species at each distance.</li>
</ul>
<p><strong>Interpretation:</strong></p>
<ul>
<li>Most species are clustered at <strong>intermediate distances</strong> (~0.18-0.22), meaning their traits are moderately similar to the community average.</li>
<li>A few species are <strong>very close</strong> to the centroid (low distances) - these are â€œcoreâ€ species with typical trait values.<br>
</li>
<li>Others lie <strong>further out</strong> (higher distances) - these are â€œperipheralâ€ species with more unusual trait combinations, which might indicate unique ecological roles or specialisations.</li>
</ul>
<p><strong>Summary</strong>: the community is centred around a typical trait set, but also includes a handful of species that are either very similar or quite distinct from that average. Â </p>
</blockquote>
</div>
<div class="section level3">
<h3 id="id_74-trait-dispersion">7.4. Trait dispersion<a class="anchor" aria-label="anchor" href="#id_74-trait-dispersion"></a>
</h3>
<p>We calculate key functional diversity metrics at the community scale:</p>
<ul>
<li>
<strong>FDis</strong>: functional <strong>dispersion</strong> (average distance to centroid)</li>
<li>
<strong>FRic</strong>: functional richness (trait-space convex hull volume)</li>
<li>
<strong>RaoQ</strong>: Raoâ€™s quadratic entropy (total abundance-weighted trait dissimilarity)</li>
</ul>
<p>These summarize the functional structure and ecological breadth of the community.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># FDis: Functional dispersion (mean distance to centroid in trait space)</span></span>
<span><span class="va">FDis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">scores_species</span><span class="op">$</span><span class="va">centrality</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># FRic: Functional richness (convex hull volume in PCoA space)</span></span>
<span><span class="va">hull</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/geometry/man/convhulln.html" class="external-link">convhulln</a></span><span class="op">(</span><span class="va">scores_species</span>, options <span class="op">=</span> <span class="st">"FA"</span><span class="op">)</span></span>
<span><span class="va">FRic</span> <span class="op">&lt;-</span> <span class="va">hull</span><span class="op">$</span><span class="va">vol</span></span>
<span></span>
<span><span class="co"># Rao's Q: Rao's quadratic entropy (abundance-weighted pairwise trait diversity)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">scores_species</span><span class="op">)</span></span>
<span><span class="va">dmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/coerce.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="va">scores_species</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">/</span> <span class="va">n</span>, <span class="va">n</span><span class="op">)</span></span>
<span><span class="va">RaoQ</span> <span class="op">&lt;-</span> <span class="fl">0.5</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/outer.html" class="external-link">outer</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">p</span><span class="op">)</span> <span class="op">*</span> <span class="va">dmat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Assemble all community-level trait dispersion metrics for comparison</span></span>
<span><span class="va">dispersion_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  Metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FDis"</span>, <span class="st">"FRic"</span>, <span class="st">"RaoQ"</span><span class="op">)</span>,</span>
<span>  Value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">FDis</span>, <span class="va">FRic</span>, <span class="va">RaoQ</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Bar plot: community-level trait dispersion metrics</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">dispersion_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Metric</span>, y <span class="op">=</span> <span class="va">Value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.6</span>, fill <span class="op">=</span> <span class="st">"firebrick"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Community-Level Trait Dispersion"</span>, y <span class="op">=</span> <span class="st">"Metric value"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="reference/figures/README-dispersion-metrics-1.png" alt="Community-level functional diversity metrics." width="100%"><p class="caption">
Community-level functional diversity metrics.
</p>
</div>
<blockquote>
<p>This bar chart summarises three ways of describing the communityâ€™s overall functional diversity:</p>
<ul>
<li><p><strong>FDis (Functional Dispersion)</strong> - Highest value here (~0.20). Shows that, on average, species are moderately spread out from the communityâ€™s trait centroid, meaning there is a fair amount of variation in trait combinations.</p></li>
<li><p><strong>FRic (Functional Richness)</strong> - Very low value (~0.02). Indicates that the total â€œvolumeâ€ of trait space occupied by the community is quite small â€” species collectively use only a limited portion of the possible trait combinations.</p></li>
<li><p><strong>RaoQ (Raoâ€™s Quadratic Entropy)</strong> - Intermediate value (~0.14). Measures total abundance-weighted trait dissimilarity. A moderate value here means that, while traits &gt;&gt; differ between species, much of the communityâ€™s abundance is concentrated in species that are not &gt;&gt; extremely dissimilar.</p></li>
</ul>
<p><strong>Summary</strong>: The community has <strong>moderate spread</strong> of traits (FDis), <strong>low coverage</strong> of the potential trait space (FRic), and <strong>moderate abundance-weighted diversity</strong> (RaoQ). This suggests that while individual species differ, the community as a whole is functionally constrained. Â </p>
</blockquote>
</div>
<div class="section level3">
<h3 id="id_75-combined-functional-trait-space-workflow">7.5. Combined functional trait space workflow<a class="anchor" aria-label="anchor" href="#id_75-combined-functional-trait-space-workflow"></a>
</h3>
<p>Trait-based community analyses often require multiple sequential steps: computing per-trait similarity, calculating trait dissimilarities across species, ordination, mapping trait space, quantifying species centrality, and summarising community-level diversity metrics.</p>
<p>The <code><a href="reference/compute_trait_space.html">compute_trait_space()</a></code> function unifies these steps into a single call, producing <strong>both</strong> per-trait similarity summaries and community-level dispersion outputs as follows:</p>
<ul>
<li><p><strong>Per-Trait Similarity Table</strong>: <code>out$similarity</code> holds percentage similarity (0-100) for each trait column, computed as either i) <em>Numeric traits</em> as the scaled mean pairwise similarity; or ii) <em>Categorical traits</em> as the proportion of identical pairs.</p></li>
<li><p><strong>Trait Dissimilarity &amp; Clustering</strong>: <code>out$dispersion$plots$dend</code> contains hierarchical clustering results based on pairwise Gower distances across all traits, highlighting functional similarity groupings.</p></li>
<li><p><strong>Trait Space Ordination (PCoA)</strong>: <code>out$dispersion$plots$density_gg</code> shows species positioned in reduced-dimensional trait space using PCoA, with kernel density contours to reveal clusters and gaps.</p></li>
<li><p><strong>Species Centrality</strong>: <code>out$dispersion$plots$centrality_hist</code> displays each speciesâ€™ Euclidean distance from the trait-space centroid. Shorter distances indicate core species; longer distances indicate peripheral, functionally distinct species.</p></li>
<li><p><strong>Functional Diversity Metrics Table</strong>: <code>out$dispersion$metrics_df</code> provides tabulated values for all computed community-level functional diversity indices.</p></li>
<li>
<p><strong>Functional Diversity Summary Plots</strong>: <code>out$dispersion$plots$metrics_bar</code> presents bar plots for three key community-level metrics:</p>
<ul>
<li>
<strong>FDis</strong> - functional dispersion (mean distance to centroid);<br>
</li>
<li>
<strong>FRic</strong> - functional richness (convex hull volume/area);<br>
</li>
<li>
<strong>RaoQ</strong> - abundance-weighted trait dissimilarity (Raoâ€™s quadratic entropy).</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># res = compute_trait_dispersion(spp_trait,</span></span>
<span><span class="co">#                                species_col = 1,</span></span>
<span><span class="co">#                                k = 4,</span></span>
<span><span class="co">#                                pcoa_dims = 2,</span></span>
<span><span class="co">#                                abundance = rep(1, nrow(spp_trait)),  # equal weights</span></span>
<span><span class="co">#                                kde_n = 100,</span></span>
<span><span class="co">#                                viridis_option = "D",</span></span>
<span><span class="co">#                                show_plots = TRUE,                    # combined patchwork output</span></span>
<span><span class="co">#                                show_density_plot = FALSE,</span></span>
<span><span class="co">#                                seed = NULL)</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># str(res, max.level=1)</span></span>
<span></span>
<span><span class="co"># same inputs and space as your manual pipeline</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/compute_trait_space.html">compute_trait_space</a></span><span class="op">(</span></span>
<span>  trait_df <span class="op">=</span> <span class="va">spp_trait</span>,</span>
<span>  species_col <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  do_similarity <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># you don't need per-trait similarity here</span></span>
<span>  k <span class="op">=</span> <span class="fl">4</span>, <span class="co"># only affects dendrogram; metrics donâ€™t use k</span></span>
<span>  pcoa_dims <span class="op">=</span> <span class="fl">2</span>, <span class="co"># keep 2 axes</span></span>
<span>  abundance <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">spp_trait</span><span class="op">)</span><span class="op">)</span>, <span class="co"># equal weights â‡’ unweighted centroid</span></span>
<span>  show_density_plot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_plots <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-func-test-1.png" width="100%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">res</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 1</span></span>
<span><span class="co">#&gt;  $ dispersion:List of 7</span></span>
<span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">similarity</span><span class="op">)</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">dispersion</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 7</span></span>
<span><span class="co">#&gt;  $ distance_matrix: num [1:27, 1:27] 0 0.398 0.416 0.534 0.607 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;  $ hc             :List of 7</span></span>
<span><span class="co">#&gt;   ..- attr(*, "class")= chr "hclust"</span></span>
<span><span class="co">#&gt;  $ pcoa           :List of 5</span></span>
<span><span class="co">#&gt;  $ scores         :'data.frame': 27 obs. of  3 variables:</span></span>
<span><span class="co">#&gt;  $ centroid       : num [1:2] -1.44e-17 -5.42e-20</span></span>
<span><span class="co">#&gt;  $ metrics_df     :'data.frame': 3 obs. of  2 variables:</span></span>
<span><span class="co">#&gt;  $ plots          :List of 4</span></span></code></pre></div>
<blockquote>
<p><strong>Note</strong>: By pairing <code><a href="reference/compute_trait_similarity.html">compute_trait_similarity()</a></code> (per-trait conservation/lability) with <code><a href="reference/compute_trait_dispersion.html">compute_trait_dispersion()</a></code> (whole-community functional structure), you get a complete, repeatable workflow for trait-based invasion and coexistence studies. All plots are stored in <code>res$plots</code> for flexible reuse, while <code>res$metrics_df</code> provides ready-to-use numerical summaries for statistical modelling. Â </p>
</blockquote>
<p>You can also rearrange or customise the plots in <code>res$plots</code> using patchwork or other layout tools. For example, the code below recreates the combined layout used when <code>show_plot = TRUE</code>, but omits the base <code><a href="https://rdrr.io/r/graphics/filled.contour.html" class="external-link">filled.contour()</a></code> plot for simplicity.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">dispersion</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 7</span></span>
<span><span class="co">#&gt;  $ distance_matrix: num [1:27, 1:27] 0 0.398 0.416 0.534 0.607 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;  $ hc             :List of 7</span></span>
<span><span class="co">#&gt;   ..- attr(*, "class")= chr "hclust"</span></span>
<span><span class="co">#&gt;  $ pcoa           :List of 5</span></span>
<span><span class="co">#&gt;  $ scores         :'data.frame': 27 obs. of  3 variables:</span></span>
<span><span class="co">#&gt;  $ centroid       : num [1:2] -1.44e-17 -5.42e-20</span></span>
<span><span class="co">#&gt;  $ metrics_df     :'data.frame': 3 obs. of  2 variables:</span></span>
<span><span class="co">#&gt;  $ plots          :List of 4</span></span>
<span></span>
<span><span class="co"># # Custom combined layout without base filled.contour</span></span>
<span><span class="co"># combined = res$dispersion$plots$dend /</span></span>
<span><span class="co">#   (res$dispersion$plots$centrality_hist | res$dispersion$plots$metrics_bar) /</span></span>
<span><span class="co">#   res$dispersion$plots$density_gg +</span></span>
<span><span class="co">#   patchwork::plot_layout(heights = c(1, 2, 1))</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># print(combined)  # display in console</span></span></code></pre></div>
<blockquote>
<p><strong>Note</strong>: - In this way you can change the order or arrangement of panels - Replace individual plots with customised versions (e.g., change themes or colours) - Combine them with other figures in your workflow Â </p>
</blockquote>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="id_8-trait-environment-response">8. Trait-Environment Response<a class="anchor" aria-label="anchor" href="#id_8-trait-environment-response"></a>
</h2>
<p>To evaluate how <strong>species functional traits, environmental conditions</strong>, and their <strong>interactions</strong> influence species abundances, we use a <strong>Generalized Linear Mixed Model (GLMM)</strong>. This framework:</p>
<ul>
<li>Quantifies the separate and combined effects of traits and environment.</li>
<li>Controls for repeated observations of the same species and sites via <strong>random intercepts</strong>, accounting for non-independence and spatial structure.</li>
<li>Is flexible enough to predict how hypothetical (invader) species might perform in new environments.</li>
</ul>
<p>We use the <strong>Tweedie</strong> error distribution, which is well-suited to ecological count data because it handles <strong>overdispersion</strong> and <strong>many zeros</strong>.</p>
<div class="section level3">
<h3 id="id_81-prepare-the-long-format-dataset">8.1. Prepare the long-format dataset<a class="anchor" aria-label="anchor" href="#id_81-prepare-the-long-format-dataset"></a>
</h3>
<p>We first create a long-format table, where each row is a single species-at-site observation, with all associated predictors attached:</p>
<ul>
<li>
<strong>Site metadata</strong>: <code>site_id</code>, spatial coordinates (<code>x</code>, <code>y</code>)</li>
<li><strong>Species ID and count/abundance</strong></li>
<li>
<strong>Environmental predictors</strong>: e.g., <code>env1</code>-<code>env10</code>
</li>
<li>
<strong>Species traits</strong>: continuous (<code>trait_cont1</code>-<code>trait_cont10</code>), categorical (<code>trait_cat11</code>-<code>trait_cat15</code>), and ordinal (<code>trait_ord16</code>-<code>trait_ord20</code>)</li>
</ul>
<p>We then convert all character variables to factors so theyâ€™re correctly handled by the model.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Prepare long-format data</span></span>
<span><span class="co"># Use simulated traits</span></span>
<span><span class="va">longDF</span> <span class="op">&lt;-</span> <span class="va">site_env_spp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>    <span class="va">site_id</span>, <span class="va">x</span>, <span class="va">y</span>, <span class="va">species</span>, <span class="va">count</span>, <span class="co"># Metadata + response</span></span>
<span>    <span class="va">env1</span><span class="op">:</span><span class="va">env10</span>, <span class="co"># Environment variables</span></span>
<span>    <span class="va">trait_cont1</span><span class="op">:</span><span class="va">trait_cont10</span>, <span class="co"># Continuous traits</span></span>
<span>    <span class="va">trait_cat11</span><span class="op">:</span><span class="va">trait_cat15</span>, <span class="co"># Categorical traits</span></span>
<span>    <span class="va">trait_ord16</span><span class="op">:</span><span class="va">trait_ord20</span> <span class="co"># Ordinal traits</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html" class="external-link">where</a></span><span class="op">(</span><span class="va">is.character</span><span class="op">)</span>, <span class="va">as.factor</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># # Use real enviro and simulated traits</span></span>
<span><span class="co"># longDF = site_env_spp %&gt;%</span></span>
<span><span class="co">#   dplyr::select(</span></span>
<span><span class="co">#     site_id, x, y, species, count,            # Metadata + response</span></span>
<span><span class="co">#     bio01:bio19,                              # Environment variables</span></span>
<span><span class="co">#     trt1:trt20</span></span>
<span><span class="co">#   ) %&gt;%</span></span>
<span><span class="co">#   mutate(across(where(is.character), as.factor))</span></span>
<span><span class="co"># names(longDF)</span></span>
<span><span class="co"># head(longDF)</span></span>
<span></span>
<span><span class="co"># # Use `grid_obs` from `dissmapr` imports instead</span></span>
<span><span class="co"># longDF = grid_obs %&gt;%</span></span>
<span><span class="co">#   mutate(site_id = as.character(site_id)) %&gt;%</span></span>
<span><span class="co">#   left_join(site_env %&gt;% dplyr::select(-x, -y) %&gt;%</span></span>
<span><span class="co">#               mutate(site_id = as.character(site_id)), by = "site_id") %&gt;%</span></span>
<span><span class="co">#   left_join(spp_trait %&gt;%</span></span>
<span><span class="co">#               mutate(species = as.character(species)), by = "species") %&gt;%</span></span>
<span><span class="co">#   mutate(across(where(is.character), as.factor))  # Ensure all character fields are treated as factors</span></span>
<span><span class="co"># # head(longDF)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="id_82-build-the-model-formula">8.2. Build the model formula<a class="anchor" aria-label="anchor" href="#id_82-build-the-model-formula"></a>
</h3>
<p>We use <code><a href="reference/build_glmm_formula.html">build_glmm_formula()</a></code> to <strong>automatically detect</strong> trait and environmental predictor columns based on naming conventions (prefixes like â€œ<em>trait_</em>â€ or â€œ<em>env_</em>â€) or by excluding known metadata columns (<code>site_id</code>, <code>x</code>, <code>y</code>, <code>species</code>, <code>count</code>).</p>
<p>The function then:</p>
<ul>
<li>
<strong>Generates main effect terms</strong> for all detected traits and all detected environment variables.</li>
<li>
<strong>Optionally expands</strong> these into <strong>all pairwise trait Ã— environment interactions</strong> (e.g., <code>trait_cont1:env3</code>), capturing environment-dependent trait effects.</li>
<li>
<strong>Appends random intercepts</strong> for the grouping variables specified in <code>random_effects</code> (default: <code>(1 | species)</code> and <code>(1 | site_id)</code>).</li>
<li>
<strong>Returns a valid R formula object</strong> ready for model fitting.</li>
</ul>
<p>This automatic approach ensures that the model always reflects the full trait-environment structure without hard-coding variable names.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Automatically build the GLMM formula</span></span>
<span><span class="va">fml</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/build_glmm_formula.html">build_glmm_formula</a></span><span class="op">(</span></span>
<span>  data                 <span class="op">=</span> <span class="va">longDF</span>,</span>
<span>  response             <span class="op">=</span> <span class="st">"count"</span>, <span class="co"># Response variable</span></span>
<span>  species_col          <span class="op">=</span> <span class="st">"species"</span>, <span class="co"># Random effect grouping</span></span>
<span>  site_col             <span class="op">=</span> <span class="st">"site_id"</span>, <span class="co"># Random effect grouping</span></span>
<span>  <span class="co"># env_cols             = names(longDF[,6:24]),</span></span>
<span>  include_interactions <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># Add all trait Ã— environment terms</span></span>
<span>  random_effects       <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"(1 | species)"</span>, <span class="st">"(1 | site_id)"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">fml</span></span>
<span><span class="co">#&gt; count ~ trait_cont1 + trait_cont2 + trait_cont3 + trait_cont4 + </span></span>
<span><span class="co">#&gt;     trait_cont5 + trait_cont6 + trait_cont7 + trait_cont8 + trait_cont9 + </span></span>
<span><span class="co">#&gt;     trait_cont10 + trait_cat11 + trait_cat12 + trait_cat13 + </span></span>
<span><span class="co">#&gt;     trait_cat14 + trait_cat15 + trait_ord16 + trait_ord17 + trait_bin18 + </span></span>
<span><span class="co">#&gt;     trait_bin19 + trait_ord20 + env1 + env2 + env3 + env4 + env5 + </span></span>
<span><span class="co">#&gt;     env6 + env7 + env8 + env9 + env10 + (trait_cont1 + trait_cont2 + </span></span>
<span><span class="co">#&gt;     trait_cont3 + trait_cont4 + trait_cont5 + trait_cont6 + trait_cont7 + </span></span>
<span><span class="co">#&gt;     trait_cont8 + trait_cont9 + trait_cont10 + trait_cat11 + </span></span>
<span><span class="co">#&gt;     trait_cat12 + trait_cat13 + trait_cat14 + trait_cat15 + trait_ord16 + </span></span>
<span><span class="co">#&gt;     trait_ord17 + trait_bin18 + trait_bin19 + trait_ord20):(env1 + </span></span>
<span><span class="co">#&gt;     env2 + env3 + env4 + env5 + env6 + env7 + env8 + env9 + env10) + </span></span>
<span><span class="co">#&gt;     (1 | species) + (1 | site_id)</span></span>
<span><span class="co">#&gt; &lt;environment: 0x000001b1cc7a3b28&gt;</span></span>
<span><span class="co"># Example output:</span></span>
<span><span class="co"># count ~ trait_cont1 + trait_cont2 + ... + env1 + env2 + ... +</span></span>
<span><span class="co">#         (trait_cont1 + ... + trait_cat15):(env1 + ... + env10) +</span></span>
<span><span class="co">#         (1 | species) + (1 | site_id)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="id_83-fit-the-glmm">8.3. Fit the GLMM<a class="anchor" aria-label="anchor" href="#id_83-fit-the-glmm"></a>
</h3>
<p>We fit the model using <code><a href="https://rdrr.io/pkg/glmmTMB/man/glmmTMB.html" class="external-link">glmmTMB::glmmTMB()</a></code>, which supports a wide range of distributions and correlation structures.</p>
<p>In this case:</p>
<ul>
<li>
<strong>Family</strong>: <code>tweedie(link = "log")</code>
<ul>
<li>Handles overdispersed count data and zero-inflation without requiring a separate zero-inflated component.</li>
<li>The <strong>log link</strong> models multiplicative effects of predictors on expected abundance.</li>
</ul>
</li>
<li>
<strong>Data</strong>: The prepared long-format table (<code>longDF</code>).</li>
<li>
<strong>Formula</strong>: The formula, automatically generated <code>fml</code> from <code><a href="reference/build_glmm_formula.html">build_glmm_formula()</a></code> or user customised.</li>
</ul>
<p>The model is fit via maximum likelihood, estimating both fixed-effect coefficients (for traits, environments, and their interactions) and variance components for the random effects.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Fit Tweedie GLMM</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu">glmmTMB</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/glmmTMB/man/glmmTMB.html" class="external-link">glmmTMB</a></span><span class="op">(</span></span>
<span>  formula <span class="op">=</span> <span class="va">fml</span>,</span>
<span>  data    <span class="op">=</span> <span class="va">longDF</span>,</span>
<span>  family  <span class="op">=</span> <span class="fu">glmmTMB</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/glmmTMB/man/nbinom2.html" class="external-link">tweedie</a></span><span class="op">(</span>link <span class="op">=</span> <span class="st">"log"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mod</span><span class="op">)</span></span>
<span><span class="co">#&gt;  Family: tweedie  ( log )</span></span>
<span><span class="co">#&gt; Formula:          </span></span>
<span><span class="co">#&gt; count ~ trait_cont1 + trait_cont2 + trait_cont3 + trait_cont4 +  </span></span>
<span><span class="co">#&gt;     trait_cont5 + trait_cont6 + trait_cont7 + trait_cont8 + trait_cont9 +  </span></span>
<span><span class="co">#&gt;     trait_cont10 + trait_cat11 + trait_cat12 + trait_cat13 +  </span></span>
<span><span class="co">#&gt;     trait_cat14 + trait_cat15 + trait_ord16 + trait_ord17 + trait_bin18 +  </span></span>
<span><span class="co">#&gt;     trait_bin19 + trait_ord20 + env1 + env2 + env3 + env4 + env5 +  </span></span>
<span><span class="co">#&gt;     env6 + env7 + env8 + env9 + env10 + (trait_cont1 + trait_cont2 +  </span></span>
<span><span class="co">#&gt;     trait_cont3 + trait_cont4 + trait_cont5 + trait_cont6 + trait_cont7 +  </span></span>
<span><span class="co">#&gt;     trait_cont8 + trait_cont9 + trait_cont10 + trait_cat11 +  </span></span>
<span><span class="co">#&gt;     trait_cat12 + trait_cat13 + trait_cat14 + trait_cat15 + trait_ord16 +  </span></span>
<span><span class="co">#&gt;     trait_ord17 + trait_bin18 + trait_bin19 + trait_ord20):(env1 +  </span></span>
<span><span class="co">#&gt;     env2 + env3 + env4 + env5 + env6 + env7 + env8 + env9 + env10) +  </span></span>
<span><span class="co">#&gt;     (1 | species) + (1 | site_id)</span></span>
<span><span class="co">#&gt; Data: longDF</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;       AIC       BIC    logLik -2*log(L)  df.resid </span></span>
<span><span class="co">#&gt;   38196.0   40239.4  -18819.0   37638.0     10926 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Random effects:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Conditional model:</span></span>
<span><span class="co">#&gt;  Groups  Name        Variance  Std.Dev.</span></span>
<span><span class="co">#&gt;  species (Intercept) 0.0007666 0.02769 </span></span>
<span><span class="co">#&gt;  site_id (Intercept) 0.0067449 0.08213 </span></span>
<span><span class="co">#&gt; Number of obs: 11205, groups:  species, 27; site_id, 415</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Dispersion parameter for tweedie family (): 7.98 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Conditional model:</span></span>
<span><span class="co">#&gt;                                 Estimate Std. Error z value Pr(&gt;|z|)    </span></span>
<span><span class="co">#&gt; (Intercept)                    1.4704424  0.2628156   5.595 2.21e-08 ***</span></span>
<span><span class="co">#&gt; trait_cont1                   -0.5101186  0.1596020  -3.196 0.001393 ** </span></span>
<span><span class="co">#&gt; trait_cont2                   -0.2327202  0.0678723  -3.429 0.000606 ***</span></span>
<span><span class="co">#&gt; trait_cont3                   -0.0341002  0.1094644  -0.312 0.755407    </span></span>
<span><span class="co">#&gt; trait_cont4                    0.0130361  0.0849067   0.154 0.877977    </span></span>
<span><span class="co">#&gt; trait_cont5                    0.0729463  0.0994374   0.734 0.463199    </span></span>
<span><span class="co">#&gt; trait_cont6                    0.0968116  0.1181499   0.819 0.412560    </span></span>
<span><span class="co">#&gt; trait_cont7                   -0.0688003  0.0874939  -0.786 0.431666    </span></span>
<span><span class="co">#&gt; trait_cont8                    0.1176131  0.1168984   1.006 0.314361    </span></span>
<span><span class="co">#&gt; trait_cont9                   -0.0103170  0.0824478  -0.125 0.900418    </span></span>
<span><span class="co">#&gt; trait_cont10                   0.1304327  0.0805313   1.620 0.105307    </span></span>
<span><span class="co">#&gt; trait_cat11grassland           0.0306508  0.1759612   0.174 0.861716    </span></span>
<span><span class="co">#&gt; trait_cat11wetland            -0.1600557  0.2324683  -0.689 0.491134    </span></span>
<span><span class="co">#&gt; trait_cat12nocturnal          -0.2304255  0.1110776  -2.074 0.038037 *  </span></span>
<span><span class="co">#&gt; trait_cat13multivoltine       -0.0010464  0.0923656  -0.011 0.990961    </span></span>
<span><span class="co">#&gt; trait_cat13univoltine         -0.2102297  0.1253302  -1.677 0.093463 .  </span></span>
<span><span class="co">#&gt; trait_cat14generalist          0.0071383  0.2479466   0.029 0.977032    </span></span>
<span><span class="co">#&gt; trait_cat14nectarivore         0.2505427  0.1645857   1.522 0.127943    </span></span>
<span><span class="co">#&gt; trait_cat15resident           -0.0296779  0.1293586  -0.229 0.818540    </span></span>
<span><span class="co">#&gt; trait_ord16                    0.0100260  0.0466039   0.215 0.829665    </span></span>
<span><span class="co">#&gt; trait_ord17                   -0.0374465  0.0314953  -1.189 0.234457    </span></span>
<span><span class="co">#&gt; trait_bin18                    0.0873029  0.0915280   0.954 0.340165    </span></span>
<span><span class="co">#&gt; trait_bin19                    0.0315366  0.1155415   0.273 0.784895    </span></span>
<span><span class="co">#&gt; trait_ord20medium              0.1233020  0.2515604   0.490 0.624029    </span></span>
<span><span class="co">#&gt; trait_ord20small              -0.1603479  0.1475234  -1.087 0.277067    </span></span>
<span><span class="co">#&gt; env1                           0.5422358  0.5433334   0.998 0.318289    </span></span>
<span><span class="co">#&gt; env2                          -1.0286903  0.5629959  -1.827 0.067674 .  </span></span>
<span><span class="co">#&gt; env3                          -0.4640513  0.8122554  -0.571 0.567788    </span></span>
<span><span class="co">#&gt; env4                           0.5304439  0.6134961   0.865 0.387245    </span></span>
<span><span class="co">#&gt; env5                           0.4719702  0.6761339   0.698 0.485151    </span></span>
<span><span class="co">#&gt; env6                           0.3931038  0.6982169   0.563 0.573427    </span></span>
<span><span class="co">#&gt; env7                           0.3775495  0.9974476   0.379 0.705048    </span></span>
<span><span class="co">#&gt; env8                          -0.2439993  1.1758269  -0.208 0.835609    </span></span>
<span><span class="co">#&gt; env9                          -0.6138702  1.2123491  -0.506 0.612613    </span></span>
<span><span class="co">#&gt; env10                         -0.4518295  1.2848177  -0.352 0.725087    </span></span>
<span><span class="co">#&gt; trait_cont1:env1               0.4542883  0.3307548   1.373 0.169600    </span></span>
<span><span class="co">#&gt; trait_cont1:env2              -0.1657942  0.3429966  -0.483 0.628833    </span></span>
<span><span class="co">#&gt; trait_cont1:env3              -0.0672174  0.4913060  -0.137 0.891178    </span></span>
<span><span class="co">#&gt; trait_cont1:env4               0.3310830  0.3695503   0.896 0.370302    </span></span>
<span><span class="co">#&gt; trait_cont1:env5              -0.0086909  0.4129877  -0.021 0.983211    </span></span>
<span><span class="co">#&gt; trait_cont1:env6              -0.0522244  0.4194517  -0.125 0.900914    </span></span>
<span><span class="co">#&gt; trait_cont1:env7              -0.2542360  0.6089449  -0.418 0.676311    </span></span>
<span><span class="co">#&gt; trait_cont1:env8              -0.1987654  0.7093235  -0.280 0.779310    </span></span>
<span><span class="co">#&gt; trait_cont1:env9               0.0132204  0.7344460   0.018 0.985638    </span></span>
<span><span class="co">#&gt; trait_cont1:env10              0.2976046  0.7825323   0.380 0.703716    </span></span>
<span><span class="co">#&gt; trait_cont2:env1               0.1224603  0.1401736   0.874 0.382318    </span></span>
<span><span class="co">#&gt; trait_cont2:env2               0.1257619  0.1457562   0.863 0.388234    </span></span>
<span><span class="co">#&gt; trait_cont2:env3               0.0510393  0.2097336   0.243 0.807732    </span></span>
<span><span class="co">#&gt; trait_cont2:env4              -0.0545704  0.1554956  -0.351 0.725630    </span></span>
<span><span class="co">#&gt; trait_cont2:env5              -0.0523391  0.1758109  -0.298 0.765931    </span></span>
<span><span class="co">#&gt; trait_cont2:env6              -0.0236931  0.1776132  -0.133 0.893879    </span></span>
<span><span class="co">#&gt; trait_cont2:env7               0.0193329  0.2592569   0.075 0.940556    </span></span>
<span><span class="co">#&gt; trait_cont2:env8              -0.0035217  0.3032631  -0.012 0.990735    </span></span>
<span><span class="co">#&gt; trait_cont2:env9               0.0114218  0.3123773   0.037 0.970833    </span></span>
<span><span class="co">#&gt; trait_cont2:env10             -0.1187163  0.3333231  -0.356 0.721721    </span></span>
<span><span class="co">#&gt; trait_cont3:env1               0.0224383  0.2268712   0.099 0.921215    </span></span>
<span><span class="co">#&gt; trait_cont3:env2               0.0197380  0.2373843   0.083 0.933734    </span></span>
<span><span class="co">#&gt; trait_cont3:env3               0.1063384  0.3360710   0.316 0.751686    </span></span>
<span><span class="co">#&gt; trait_cont3:env4               0.2250068  0.2545347   0.884 0.376700    </span></span>
<span><span class="co">#&gt; trait_cont3:env5               0.0438643  0.2815770   0.156 0.876206    </span></span>
<span><span class="co">#&gt; trait_cont3:env6               0.0931172  0.2868353   0.325 0.745456    </span></span>
<span><span class="co">#&gt; trait_cont3:env7              -0.3698838  0.4171191  -0.887 0.375209    </span></span>
<span><span class="co">#&gt; trait_cont3:env8              -0.2242540  0.4868664  -0.461 0.645081    </span></span>
<span><span class="co">#&gt; trait_cont3:env9              -0.1435396  0.5033282  -0.285 0.775506    </span></span>
<span><span class="co">#&gt; trait_cont3:env10              0.1991328  0.5354910   0.372 0.709990    </span></span>
<span><span class="co">#&gt; trait_cont4:env1               0.0662273  0.1761968   0.376 0.707013    </span></span>
<span><span class="co">#&gt; trait_cont4:env2               0.0482631  0.1848803   0.261 0.794054    </span></span>
<span><span class="co">#&gt; trait_cont4:env3               0.0475713  0.2618060   0.182 0.855815    </span></span>
<span><span class="co">#&gt; trait_cont4:env4               0.0558466  0.1984742   0.281 0.778419    </span></span>
<span><span class="co">#&gt; trait_cont4:env5              -0.0079881  0.2180135  -0.037 0.970772    </span></span>
<span><span class="co">#&gt; trait_cont4:env6               0.1322373  0.2240845   0.590 0.555109    </span></span>
<span><span class="co">#&gt; trait_cont4:env7               0.0030004  0.3240575   0.009 0.992613    </span></span>
<span><span class="co">#&gt; trait_cont4:env8              -0.0559403  0.3797565  -0.147 0.882891    </span></span>
<span><span class="co">#&gt; trait_cont4:env9              -0.0373883  0.3919177  -0.095 0.923998    </span></span>
<span><span class="co">#&gt; trait_cont4:env10             -0.2119408  0.4155868  -0.510 0.610066    </span></span>
<span><span class="co">#&gt; trait_cont5:env1              -0.3157162  0.2076197  -1.521 0.128348    </span></span>
<span><span class="co">#&gt; trait_cont5:env2              -0.0020636  0.2157308  -0.010 0.992368    </span></span>
<span><span class="co">#&gt; trait_cont5:env3              -0.0658514  0.3077739  -0.214 0.830578    </span></span>
<span><span class="co">#&gt; trait_cont5:env4               0.0134922  0.2354324   0.057 0.954300    </span></span>
<span><span class="co">#&gt; trait_cont5:env5               0.0261295  0.2553548   0.102 0.918498    </span></span>
<span><span class="co">#&gt; trait_cont5:env6               0.2519034  0.2647715   0.951 0.341402    </span></span>
<span><span class="co">#&gt; trait_cont5:env7              -0.4533944  0.3789051  -1.197 0.231466    </span></span>
<span><span class="co">#&gt; trait_cont5:env8               0.0661443  0.4466947   0.148 0.882284    </span></span>
<span><span class="co">#&gt; trait_cont5:env9               0.0510576  0.4586971   0.111 0.911371    </span></span>
<span><span class="co">#&gt; trait_cont5:env10              0.5789141  0.4873776   1.188 0.234907    </span></span>
<span><span class="co">#&gt; trait_cont6:env1              -0.1881089  0.2447584  -0.769 0.442161    </span></span>
<span><span class="co">#&gt; trait_cont6:env2               0.1347617  0.2537602   0.531 0.595378    </span></span>
<span><span class="co">#&gt; trait_cont6:env3               0.1762242  0.3631814   0.485 0.627518    </span></span>
<span><span class="co">#&gt; trait_cont6:env4              -0.1655427  0.2731632  -0.606 0.544501    </span></span>
<span><span class="co">#&gt; trait_cont6:env5              -0.0584509  0.3052173  -0.192 0.848129    </span></span>
<span><span class="co">#&gt; trait_cont6:env6               0.1809909  0.3105673   0.583 0.560045    </span></span>
<span><span class="co">#&gt; trait_cont6:env7               0.0136916  0.4483815   0.031 0.975640    </span></span>
<span><span class="co">#&gt; trait_cont6:env8              -0.0287384  0.5247439  -0.055 0.956324    </span></span>
<span><span class="co">#&gt; trait_cont6:env9               0.1670829  0.5425820   0.308 0.758128    </span></span>
<span><span class="co">#&gt; trait_cont6:env10             -0.0216889  0.5775002  -0.038 0.970041    </span></span>
<span><span class="co">#&gt; trait_cont7:env1               0.0117718  0.1832154   0.064 0.948770    </span></span>
<span><span class="co">#&gt; trait_cont7:env2              -0.2494385  0.1908231  -1.307 0.191155    </span></span>
<span><span class="co">#&gt; trait_cont7:env3              -0.2403364  0.2736175  -0.878 0.379745    </span></span>
<span><span class="co">#&gt; trait_cont7:env4               0.1897891  0.2073779   0.915 0.360095    </span></span>
<span><span class="co">#&gt; trait_cont7:env5               0.0307101  0.2285526   0.134 0.893112    </span></span>
<span><span class="co">#&gt; trait_cont7:env6               0.2824628  0.2343997   1.205 0.228185    </span></span>
<span><span class="co">#&gt; trait_cont7:env7              -0.1028858  0.3374557  -0.305 0.760452    </span></span>
<span><span class="co">#&gt; trait_cont7:env8              -0.0941266  0.3973214  -0.237 0.812732    </span></span>
<span><span class="co">#&gt; trait_cont7:env9               0.0188233  0.4101149   0.046 0.963392    </span></span>
<span><span class="co">#&gt; trait_cont7:env10              0.0654068  0.4330113   0.151 0.879936    </span></span>
<span><span class="co">#&gt; trait_cont8:env1              -0.1719497  0.2425198  -0.709 0.478316    </span></span>
<span><span class="co">#&gt; trait_cont8:env2              -0.2490949  0.2505595  -0.994 0.320148    </span></span>
<span><span class="co">#&gt; trait_cont8:env3              -0.1298487  0.3597634  -0.361 0.718153    </span></span>
<span><span class="co">#&gt; trait_cont8:env4              -0.0541162  0.2704139  -0.200 0.841384    </span></span>
<span><span class="co">#&gt; trait_cont8:env5               0.1029516  0.3019067   0.341 0.733100    </span></span>
<span><span class="co">#&gt; trait_cont8:env6               0.2066006  0.3083429   0.670 0.502835    </span></span>
<span><span class="co">#&gt; trait_cont8:env7               0.0301432  0.4444334   0.068 0.945926    </span></span>
<span><span class="co">#&gt; trait_cont8:env8               0.0388618  0.5189531   0.075 0.940306    </span></span>
<span><span class="co">#&gt; trait_cont8:env9               0.0657501  0.5389646   0.122 0.902904    </span></span>
<span><span class="co">#&gt; trait_cont8:env10              0.1524384  0.5715950   0.267 0.789708    </span></span>
<span><span class="co">#&gt; trait_cont9:env1               0.0491704  0.1729237   0.284 0.776144    </span></span>
<span><span class="co">#&gt; trait_cont9:env2              -0.1516373  0.1800992  -0.842 0.399807    </span></span>
<span><span class="co">#&gt; trait_cont9:env3               0.0404439  0.2559539   0.158 0.874447    </span></span>
<span><span class="co">#&gt; trait_cont9:env4               0.1468507  0.1926070   0.762 0.445799    </span></span>
<span><span class="co">#&gt; trait_cont9:env5               0.1078422  0.2144027   0.503 0.614972    </span></span>
<span><span class="co">#&gt; trait_cont9:env6               0.1373539  0.2175442   0.631 0.527790    </span></span>
<span><span class="co">#&gt; trait_cont9:env7               0.0709711  0.3175835   0.223 0.823168    </span></span>
<span><span class="co">#&gt; trait_cont9:env8              -0.1044847  0.3705180  -0.282 0.777946    </span></span>
<span><span class="co">#&gt; trait_cont9:env9              -0.0241682  0.3826458  -0.063 0.949638    </span></span>
<span><span class="co">#&gt; trait_cont9:env10             -0.0707903  0.4081366  -0.173 0.862300    </span></span>
<span><span class="co">#&gt; trait_cont10:env1             -0.0161651  0.1675005  -0.097 0.923117    </span></span>
<span><span class="co">#&gt; trait_cont10:env2             -0.0123918  0.1748189  -0.071 0.943490    </span></span>
<span><span class="co">#&gt; trait_cont10:env3             -0.1351959  0.2499217  -0.541 0.588540    </span></span>
<span><span class="co">#&gt; trait_cont10:env4             -0.0179706  0.1863072  -0.096 0.923158    </span></span>
<span><span class="co">#&gt; trait_cont10:env5             -0.0960179  0.2093890  -0.459 0.646548    </span></span>
<span><span class="co">#&gt; trait_cont10:env6              0.0523432  0.2122691   0.247 0.805227    </span></span>
<span><span class="co">#&gt; trait_cont10:env7              0.1417525  0.3084044   0.460 0.645780    </span></span>
<span><span class="co">#&gt; trait_cont10:env8              0.0911406  0.3615158   0.252 0.800959    </span></span>
<span><span class="co">#&gt; trait_cont10:env9              0.1438915  0.3728547   0.386 0.699557    </span></span>
<span><span class="co">#&gt; trait_cont10:env10            -0.1892620  0.3969231  -0.477 0.633488    </span></span>
<span><span class="co">#&gt; trait_cat11grassland:env1     -0.2166084  0.3633703  -0.596 0.551102    </span></span>
<span><span class="co">#&gt; trait_cat11wetland:env1       -0.2864278  0.4824768  -0.594 0.552739    </span></span>
<span><span class="co">#&gt; trait_cat11grassland:env2      0.3342700  0.3784489   0.883 0.377094    </span></span>
<span><span class="co">#&gt; trait_cat11wetland:env2       -0.0904607  0.5010602  -0.181 0.856730    </span></span>
<span><span class="co">#&gt; trait_cat11grassland:env3      0.0548656  0.5379108   0.102 0.918759    </span></span>
<span><span class="co">#&gt; trait_cat11wetland:env3        0.0121837  0.7162052   0.017 0.986427    </span></span>
<span><span class="co">#&gt; trait_cat11grassland:env4     -0.1520046  0.4004652  -0.380 0.704265    </span></span>
<span><span class="co">#&gt; trait_cat11wetland:env4        0.4100571  0.5381755   0.762 0.446096    </span></span>
<span><span class="co">#&gt; trait_cat11grassland:env5     -0.0669323  0.4505183  -0.149 0.881895    </span></span>
<span><span class="co">#&gt; trait_cat11wetland:env5       -0.0382453  0.6020639  -0.064 0.949350    </span></span>
<span><span class="co">#&gt; trait_cat11grassland:env6     -0.4388071  0.4552669  -0.964 0.335123    </span></span>
<span><span class="co">#&gt; trait_cat11wetland:env6       -0.1781153  0.6077280  -0.293 0.769458    </span></span>
<span><span class="co">#&gt; trait_cat11grassland:env7     -0.1155986  0.6691581  -0.173 0.862846    </span></span>
<span><span class="co">#&gt; trait_cat11wetland:env7        0.0118782  0.8885123   0.013 0.989334    </span></span>
<span><span class="co">#&gt; trait_cat11grassland:env8      0.1636730  0.7758684   0.211 0.832923    </span></span>
<span><span class="co">#&gt; trait_cat11wetland:env8       -0.1067334  1.0347932  -0.103 0.917848    </span></span>
<span><span class="co">#&gt; trait_cat11grassland:env9     -0.1814162  0.8019406  -0.226 0.821029    </span></span>
<span><span class="co">#&gt; trait_cat11wetland:env9       -0.1069967  1.0669591  -0.100 0.920120    </span></span>
<span><span class="co">#&gt; trait_cat11grassland:env10     0.0621179  0.8608062   0.072 0.942473    </span></span>
<span><span class="co">#&gt; trait_cat11wetland:env10       0.1178661  1.1409306   0.103 0.917719    </span></span>
<span><span class="co">#&gt; trait_cat12nocturnal:env1     -0.0754443  0.2309471  -0.327 0.743915    </span></span>
<span><span class="co">#&gt; trait_cat12nocturnal:env2     -0.0296988  0.2400987  -0.124 0.901557    </span></span>
<span><span class="co">#&gt; trait_cat12nocturnal:env3      0.0047181  0.3406933   0.014 0.988951    </span></span>
<span><span class="co">#&gt; trait_cat12nocturnal:env4      0.0006826  0.2569039   0.003 0.997880    </span></span>
<span><span class="co">#&gt; trait_cat12nocturnal:env5     -0.1068315  0.2872327  -0.372 0.709942    </span></span>
<span><span class="co">#&gt; trait_cat12nocturnal:env6      0.2186171  0.2904297   0.753 0.451608    </span></span>
<span><span class="co">#&gt; trait_cat12nocturnal:env7     -0.4323136  0.4256436  -1.016 0.309786    </span></span>
<span><span class="co">#&gt; trait_cat12nocturnal:env8      0.0687424  0.4939589   0.139 0.889319    </span></span>
<span><span class="co">#&gt; trait_cat12nocturnal:env9      0.1660581  0.5108386   0.325 0.745128    </span></span>
<span><span class="co">#&gt; trait_cat12nocturnal:env10     0.5392907  0.5458439   0.988 0.323155    </span></span>
<span><span class="co">#&gt; trait_cat13multivoltine:env1   0.1368508  0.1909488   0.717 0.473566    </span></span>
<span><span class="co">#&gt; trait_cat13univoltine:env1     0.2841505  0.2577352   1.102 0.270249    </span></span>
<span><span class="co">#&gt; trait_cat13multivoltine:env2   0.0681931  0.1971593   0.346 0.729434    </span></span>
<span><span class="co">#&gt; trait_cat13univoltine:env2    -0.0341924  0.2653602  -0.129 0.897474    </span></span>
<span><span class="co">#&gt; trait_cat13multivoltine:env3  -0.0409377  0.2844874  -0.144 0.885580    </span></span>
<span><span class="co">#&gt; trait_cat13univoltine:env3    -0.0574466  0.3794755  -0.151 0.879673    </span></span>
<span><span class="co">#&gt; trait_cat13multivoltine:env4  -0.0495623  0.2158698  -0.230 0.818408    </span></span>
<span><span class="co">#&gt; trait_cat13univoltine:env4     0.1227831  0.2886493   0.425 0.670566    </span></span>
<span><span class="co">#&gt; trait_cat13multivoltine:env5  -0.1020816  0.2363995  -0.432 0.665873    </span></span>
<span><span class="co">#&gt; trait_cat13univoltine:env5    -0.2213424  0.3186085  -0.695 0.487233    </span></span>
<span><span class="co">#&gt; trait_cat13multivoltine:env6  -0.2882216  0.2432146  -1.185 0.235997    </span></span>
<span><span class="co">#&gt; trait_cat13univoltine:env6     0.2284586  0.3278184   0.697 0.485862    </span></span>
<span><span class="co">#&gt; trait_cat13multivoltine:env7  -0.0209723  0.3503083  -0.060 0.952261    </span></span>
<span><span class="co">#&gt; trait_cat13univoltine:env7     0.0771967  0.4670323   0.165 0.868714    </span></span>
<span><span class="co">#&gt; trait_cat13multivoltine:env8   0.0419555  0.4116171   0.102 0.918814    </span></span>
<span><span class="co">#&gt; trait_cat13univoltine:env8     0.0414558  0.5478425   0.076 0.939681    </span></span>
<span><span class="co">#&gt; trait_cat13multivoltine:env9  -0.0118302  0.4233286  -0.028 0.977705    </span></span>
<span><span class="co">#&gt; trait_cat13univoltine:env9     0.3152771  0.5656912   0.557 0.577301    </span></span>
<span><span class="co">#&gt; trait_cat13multivoltine:env10  0.0253027  0.4497710   0.056 0.955137    </span></span>
<span><span class="co">#&gt; trait_cat13univoltine:env10   -0.2688425  0.6025053  -0.446 0.655447    </span></span>
<span><span class="co">#&gt; trait_cat14generalist:env1     0.2606976  0.5164928   0.505 0.613737    </span></span>
<span><span class="co">#&gt; trait_cat14nectarivore:env1   -0.2360857  0.3408356  -0.693 0.488518    </span></span>
<span><span class="co">#&gt; trait_cat14generalist:env2     0.0418664  0.5331459   0.079 0.937409    </span></span>
<span><span class="co">#&gt; trait_cat14nectarivore:env2   -0.0044629  0.3533997  -0.013 0.989924    </span></span>
<span><span class="co">#&gt; trait_cat14generalist:env3    -0.0139439  0.7682749  -0.018 0.985519    </span></span>
<span><span class="co">#&gt; trait_cat14nectarivore:env3    0.1527119  0.5020463   0.304 0.760992    </span></span>
<span><span class="co">#&gt; trait_cat14generalist:env4     0.1510262  0.5776250   0.261 0.793737    </span></span>
<span><span class="co">#&gt; trait_cat14nectarivore:env4    0.0284546  0.3790514   0.075 0.940161    </span></span>
<span><span class="co">#&gt; trait_cat14generalist:env5    -0.2015317  0.6417709  -0.314 0.753503    </span></span>
<span><span class="co">#&gt; trait_cat14nectarivore:env5   -0.0326672  0.4232302  -0.077 0.938476    </span></span>
<span><span class="co">#&gt; trait_cat14generalist:env6    -0.3943444  0.6552377  -0.602 0.547285    </span></span>
<span><span class="co">#&gt; trait_cat14nectarivore:env6   -0.0958633  0.4293834  -0.223 0.823335    </span></span>
<span><span class="co">#&gt; trait_cat14generalist:env7     0.3523372  0.9512120   0.370 0.711078    </span></span>
<span><span class="co">#&gt; trait_cat14nectarivore:env7    0.2099320  0.6256045   0.336 0.737198    </span></span>
<span><span class="co">#&gt; trait_cat14generalist:env8    -0.0441667  1.1089876  -0.040 0.968232    </span></span>
<span><span class="co">#&gt; trait_cat14nectarivore:env8   -0.2024453  0.7254215  -0.279 0.780189    </span></span>
<span><span class="co">#&gt; trait_cat14generalist:env9     0.0889186  1.1459615   0.078 0.938152    </span></span>
<span><span class="co">#&gt; trait_cat14nectarivore:env9    0.0925423  0.7521012   0.123 0.902071    </span></span>
<span><span class="co">#&gt; trait_cat14generalist:env10   -0.2991912  1.2222034  -0.245 0.806614    </span></span>
<span><span class="co">#&gt; trait_cat14nectarivore:env10   0.0601482  0.8030205   0.075 0.940292    </span></span>
<span><span class="co">#&gt; trait_cat15resident:env1      -0.1558773  0.2696314  -0.578 0.563188    </span></span>
<span><span class="co">#&gt; trait_cat15resident:env2       0.0972675  0.2779438   0.350 0.726373    </span></span>
<span><span class="co">#&gt; trait_cat15resident:env3       0.0959809  0.4010530   0.239 0.810856    </span></span>
<span><span class="co">#&gt; trait_cat15resident:env4      -0.1433718  0.3026179  -0.474 0.635663    </span></span>
<span><span class="co">#&gt; trait_cat15resident:env5       0.0713967  0.3360772   0.212 0.831763    </span></span>
<span><span class="co">#&gt; trait_cat15resident:env6      -0.0006126  0.3434115  -0.002 0.998577    </span></span>
<span><span class="co">#&gt; trait_cat15resident:env7      -0.2809555  0.4952511  -0.567 0.570511    </span></span>
<span><span class="co">#&gt; trait_cat15resident:env8       0.1465885  0.5809911   0.252 0.800803    </span></span>
<span><span class="co">#&gt; trait_cat15resident:env9      -0.1148045  0.6007147  -0.191 0.848437    </span></span>
<span><span class="co">#&gt; trait_cat15resident:env10      0.3813862  0.6369512   0.599 0.549327    </span></span>
<span><span class="co">#&gt; trait_ord16:env1               0.0443858  0.0967309   0.459 0.646336    </span></span>
<span><span class="co">#&gt; trait_ord16:env2               0.1264998  0.0999672   1.265 0.205723    </span></span>
<span><span class="co">#&gt; trait_ord16:env3               0.0089911  0.1434370   0.063 0.950018    </span></span>
<span><span class="co">#&gt; trait_ord16:env4              -0.1249036  0.1092007  -1.144 0.252708    </span></span>
<span><span class="co">#&gt; trait_ord16:env5              -0.0327199  0.1190084  -0.275 0.783364    </span></span>
<span><span class="co">#&gt; trait_ord16:env6              -0.1532371  0.1232035  -1.244 0.213583    </span></span>
<span><span class="co">#&gt; trait_ord16:env7               0.0864534  0.1769544   0.489 0.625151    </span></span>
<span><span class="co">#&gt; trait_ord16:env8               0.0561132  0.2077514   0.270 0.787085    </span></span>
<span><span class="co">#&gt; trait_ord16:env9               0.0409145  0.2138390   0.191 0.848265    </span></span>
<span><span class="co">#&gt; trait_ord16:env10             -0.1615596  0.2272823  -0.711 0.477188    </span></span>
<span><span class="co">#&gt; trait_ord17:env1              -0.0508782  0.0650626  -0.782 0.434221    </span></span>
<span><span class="co">#&gt; trait_ord17:env2               0.0673156  0.0678696   0.992 0.321277    </span></span>
<span><span class="co">#&gt; trait_ord17:env3               0.0981880  0.0966602   1.016 0.309722    </span></span>
<span><span class="co">#&gt; trait_ord17:env4               0.0150704  0.0720106   0.209 0.834230    </span></span>
<span><span class="co">#&gt; trait_ord17:env5              -0.0461649  0.0803691  -0.574 0.565690    </span></span>
<span><span class="co">#&gt; trait_ord17:env6               0.0522241  0.0822010   0.635 0.525218    </span></span>
<span><span class="co">#&gt; trait_ord17:env7              -0.1098775  0.1196980  -0.918 0.358642    </span></span>
<span><span class="co">#&gt; trait_ord17:env8              -0.0805842  0.1399392  -0.576 0.564715    </span></span>
<span><span class="co">#&gt; trait_ord17:env9               0.1282283  0.1437985   0.892 0.372542    </span></span>
<span><span class="co">#&gt; trait_ord17:env10              0.1133910  0.1545754   0.734 0.463214    </span></span>
<span><span class="co">#&gt; trait_bin18:env1              -0.1179621  0.1877592  -0.628 0.529832    </span></span>
<span><span class="co">#&gt; trait_bin18:env2               0.0410970  0.1953517   0.210 0.833375    </span></span>
<span><span class="co">#&gt; trait_bin18:env3              -0.1774229  0.2786485  -0.637 0.524303    </span></span>
<span><span class="co">#&gt; trait_bin18:env4               0.2552379  0.2089121   1.222 0.221803    </span></span>
<span><span class="co">#&gt; trait_bin18:env5               0.1704284  0.2339985   0.728 0.466411    </span></span>
<span><span class="co">#&gt; trait_bin18:env6              -0.1018293  0.2381233  -0.428 0.668919    </span></span>
<span><span class="co">#&gt; trait_bin18:env7              -0.1197166  0.3440251  -0.348 0.727849    </span></span>
<span><span class="co">#&gt; trait_bin18:env8               0.2359866  0.4016228   0.588 0.556812    </span></span>
<span><span class="co">#&gt; trait_bin18:env9              -0.2244515  0.4160808  -0.539 0.589582    </span></span>
<span><span class="co">#&gt; trait_bin18:env10              0.1134193  0.4424246   0.256 0.797674    </span></span>
<span><span class="co">#&gt; trait_bin19:env1              -0.1452246  0.2360668  -0.615 0.538433    </span></span>
<span><span class="co">#&gt; trait_bin19:env2               0.0131558  0.2494912   0.053 0.957946    </span></span>
<span><span class="co">#&gt; trait_bin19:env3               0.1914155  0.3544992   0.540 0.589224    </span></span>
<span><span class="co">#&gt; trait_bin19:env4              -0.2372310  0.2625524  -0.904 0.366231    </span></span>
<span><span class="co">#&gt; trait_bin19:env5              -0.0486417  0.2962394  -0.164 0.869576    </span></span>
<span><span class="co">#&gt; trait_bin19:env6               0.0364531  0.2983436   0.122 0.902753    </span></span>
<span><span class="co">#&gt; trait_bin19:env7               0.1866480  0.4367544   0.427 0.669123    </span></span>
<span><span class="co">#&gt; trait_bin19:env8               0.0871527  0.5111420   0.171 0.864612    </span></span>
<span><span class="co">#&gt; trait_bin19:env9               0.1295414  0.5262485   0.246 0.805558    </span></span>
<span><span class="co">#&gt; trait_bin19:env10              0.0354184  0.5621364   0.063 0.949761    </span></span>
<span><span class="co">#&gt; trait_ord20medium:env1        -0.1718096  0.5230806  -0.328 0.742566    </span></span>
<span><span class="co">#&gt; trait_ord20small:env1          0.2169684  0.3054409   0.710 0.477490    </span></span>
<span><span class="co">#&gt; trait_ord20medium:env2        -0.1134164  0.5411585  -0.210 0.833995    </span></span>
<span><span class="co">#&gt; trait_ord20small:env2          0.1796733  0.3127285   0.575 0.565606    </span></span>
<span><span class="co">#&gt; trait_ord20medium:env3        -0.0666643  0.7776583  -0.086 0.931685    </span></span>
<span><span class="co">#&gt; trait_ord20small:env3         -0.0540712  0.4546263  -0.119 0.905327    </span></span>
<span><span class="co">#&gt; trait_ord20medium:env4        -0.2657073  0.5851281  -0.454 0.649756    </span></span>
<span><span class="co">#&gt; trait_ord20small:env4         -0.3051708  0.3432778  -0.889 0.374008    </span></span>
<span><span class="co">#&gt; trait_ord20medium:env5        -0.0077351  0.6507966  -0.012 0.990517    </span></span>
<span><span class="co">#&gt; trait_ord20small:env5         -0.2466345  0.3796726  -0.650 0.515952    </span></span>
<span><span class="co">#&gt; trait_ord20medium:env6         0.4155222  0.6624804   0.627 0.530514    </span></span>
<span><span class="co">#&gt; trait_ord20small:env6          0.0448119  0.3876854   0.116 0.907979    </span></span>
<span><span class="co">#&gt; trait_ord20medium:env7        -0.0706120  0.9624806  -0.073 0.941516    </span></span>
<span><span class="co">#&gt; trait_ord20small:env7          0.0312668  0.5610285   0.056 0.955556    </span></span>
<span><span class="co">#&gt; trait_ord20medium:env8         0.1203152  1.1207031   0.107 0.914506    </span></span>
<span><span class="co">#&gt; trait_ord20small:env8          0.2136062  0.6563351   0.325 0.744838    </span></span>
<span><span class="co">#&gt; trait_ord20medium:env9         0.0828510  1.1587777   0.071 0.943001    </span></span>
<span><span class="co">#&gt; trait_ord20small:env9          0.0302249  0.6752415   0.045 0.964297    </span></span>
<span><span class="co">#&gt; trait_ord20medium:env10        0.0412495  1.2358430   0.033 0.973373    </span></span>
<span><span class="co">#&gt; trait_ord20small:env10        -0.3558163  0.7188480  -0.495 0.620613    </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code></pre></div>
<blockquote>
<p><strong>Note</strong>: This model forms the core of the invasion fitness workflow, allowing us to predict how novel trait-environment combinations are likely to perform (abundance-wise) in different sites, and enables robust estimation of both direct and interactive drivers of community assembly. Â </p>
</blockquote>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="id_9-simulate--predict-invaders">9. Simulate &amp; Predict Invaders<a class="anchor" aria-label="anchor" href="#id_9-simulate--predict-invaders"></a>
</h2>
<p>This section demonstrates how to <strong>simulate trait profiles for hypothetical invaders</strong> and then use the fitted abundance model to <strong>predict their expected performance across all sites</strong>. This forms a crucial part of modern invasion biology workflows: moving from theory-driven invader trait generation to data-driven, site-level predictions of establishment and performance.</p>
<div class="section level3">
<h3 id="id_91-simulate-hypothetical-invaders">9.1. Simulate hypothetical invaders<a class="anchor" aria-label="anchor" href="#id_91-simulate-hypothetical-invaders"></a>
</h3>
<p>To evaluate how potential invaders might perform, we generate novel species by resampling trait values from the empirical distribution of the resident pool using the <code><a href="reference/simulate_invaders.html">simulate_invaders()</a></code> function. This approach allows you to:</p>
<ul>
<li>
<strong>Preserve realism</strong> by keeping trait ranges and levels identical to those observed in the resident community.</li>
<li>
<strong>Explore unobserved combinations</strong> (when using column-wise sampling) or maintain natural trait correlations (when using row-wise sampling).</li>
<li>Control how numeric traits are drawn - bootstrap from observed values, draw from a truncated normal distribution, or sample from a uniform distribution within observed bounds.</li>
</ul>
<p>By default, traits are sampled <strong>independently by column</strong> (<code>mode = "columnwise"</code>), which generates novel combinations. Alternatively, setting <code>mode = "rowwise"</code> samples entire resident profiles, preserving observed cross-trait covariance.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example: simulate 10 invaders from the resident trait table</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">inv_traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/simulate_invaders.html">simulate_invaders</a></span><span class="op">(</span></span>
<span>  resident_traits <span class="op">=</span> <span class="va">spp_trait</span>, <span class="co"># your resident species traits table</span></span>
<span>  n_inv           <span class="op">=</span> <span class="fl">10</span>, <span class="co"># number of invaders to create</span></span>
<span>  species_col     <span class="op">=</span> <span class="st">"species"</span>, <span class="co"># ID column for species</span></span>
<span>  mode            <span class="op">=</span> <span class="st">"columnwise"</span>, <span class="co"># or "rowwise"</span></span>
<span>  numeric_method  <span class="op">=</span> <span class="st">"bootstrap"</span>, <span class="co"># or "normal", "uniform"</span></span>
<span>  seed            <span class="op">=</span> <span class="fl">42</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">inv_traits</span><span class="op">)</span></span>
<span><span class="co">#&gt;      species trait_cont1 trait_cont2 trait_cont3 trait_cont4 trait_cont5</span></span>
<span><span class="co">#&gt; inv1    inv1  0.47317663  -0.9853317  0.02881542  -0.8287759  0.07475339</span></span>
<span><span class="co">#&gt; inv2    inv2  0.31398458   0.8114763  0.65631697   0.9252160  0.70096550</span></span>
<span><span class="co">#&gt; inv3    inv3  0.02842357   0.5693856  0.69937944   0.1272937  0.70096550</span></span>
<span><span class="co">#&gt; inv4    inv4 -0.08451645  -0.5846821  0.43871168   0.4797106  0.20694817</span></span>
<span><span class="co">#&gt; inv5    inv5  0.86934449   0.6658322  0.16320801   0.8660683  0.65788426</span></span>
<span><span class="co">#&gt; inv6    inv6 -0.21959307  -0.1060607  0.51908854   0.3348530  0.14695180</span></span>
<span><span class="co">#&gt;      trait_cont6 trait_cont7 trait_cont8 trait_cont9 trait_cont10 trait_cat11</span></span>
<span><span class="co">#&gt; inv1  -0.6596750  -0.3658933  0.06374887   0.9931054 -0.001454239      forest</span></span>
<span><span class="co">#&gt; inv2   0.3603284   0.8361128 -0.77538352  -0.5086517 -0.957099543      forest</span></span>
<span><span class="co">#&gt; inv3   0.4670559   0.8358081 -0.09737830  -0.8268388 -0.460567643     wetland</span></span>
<span><span class="co">#&gt; inv4   0.3571857   0.4947222 -0.89374103  -0.8647264  0.881129756   grassland</span></span>
<span><span class="co">#&gt; inv5  -0.8394711   0.4845329 -0.42450050  -0.6819552 -0.957099543   grassland</span></span>
<span><span class="co">#&gt; inv6  -0.9418284  -0.3658933  0.67751007  -0.3726324 -0.623131341     wetland</span></span>
<span><span class="co">#&gt;      trait_cat12  trait_cat13 trait_cat14 trait_cat15 trait_ord16 trait_ord17</span></span>
<span><span class="co">#&gt; inv1   nocturnal multivoltine detritivore   migratory           1           2</span></span>
<span><span class="co">#&gt; inv2   nocturnal   univoltine detritivore   migratory           3           2</span></span>
<span><span class="co">#&gt; inv3   nocturnal    bivoltine nectarivore   migratory           1           2</span></span>
<span><span class="co">#&gt; inv4   nocturnal multivoltine  generalist   migratory           1           2</span></span>
<span><span class="co">#&gt; inv5     diurnal    bivoltine nectarivore    resident           2           1</span></span>
<span><span class="co">#&gt; inv6     diurnal   univoltine  generalist   migratory           1           3</span></span>
<span><span class="co">#&gt;      trait_bin18 trait_bin19 trait_ord20</span></span>
<span><span class="co">#&gt; inv1           1           1       large</span></span>
<span><span class="co">#&gt; inv2           0           1       small</span></span>
<span><span class="co">#&gt; inv3           1           1      medium</span></span>
<span><span class="co">#&gt; inv4           1           1      medium</span></span>
<span><span class="co">#&gt; inv5           0           1      medium</span></span>
<span><span class="co">#&gt; inv6           0           1       large</span></span></code></pre></div>
<blockquote>
<p><strong>Note</strong>: Use <code>mode = "rowwise"</code> when you want to preserve realistic trait correlations.<br>
Use <code>mode = "columnwise"</code> when exploring hypothetical trait combinations beyond those observed. Â </p>
</blockquote>
</div>
<div class="section level3">
<h3 id="id_92-predict-abundances-across-all-sites">9.2. Predict abundances across all sites<a class="anchor" aria-label="anchor" href="#id_92-predict-abundances-across-all-sites"></a>
</h3>
<p>With simulated invaders in hand, we use the <code><a href="reference/predict_invader_response.html">predict_invader_response()</a></code> function to generate expected abundances (our performance proxy) for every <strong>species-site</strong> combination.</p>
<p>This function:<br>
- Takes a fitted model (e.g., from <strong>glmmTMB</strong>, <strong>lme4</strong>, <strong>glm</strong>, or <strong>gam</strong>)<br>
- Crosses the <strong>species traits table</strong> (residents + simulated invaders) with the <strong>site environment table</strong><br>
- Harmonizes factor levels with the modelâ€™s training data<br>
- Calls the modelâ€™s native <code><a href="https://rspatial.github.io/terra/reference/predict.html" class="external-link">predict()</a></code> method to return: - The full prediction grid (<code>newdata</code>) - A long table of predictions (<code>predictions</code>) - A wide <strong>site Ã— species</strong> prediction matrix (<code>prediction_matrix</code>)</p>
<blockquote>
<p><strong>Note</strong>: Why use population-level (fixed-effects) predictions? - For novel invaders, the random-effect levels (e.g., species- or site-specific intercepts) are unknown. - Accordingly, <code><a href="reference/predict_invader_response.html">predict_invader_response()</a></code> defaults to setting random effects to zero, producing <em>population-level</em> predictions driven solely by traits, environmental covariates, and their interactions. This is the appropriate scale for invasion screening and comparative risk assessment. - Set <code>include_random = TRUE</code> only if you require conditional predictions that incorporate known random-effect levels from the fitted model (and, where supported, allow estimation for new levels).<br>
Â </p>
</blockquote>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Combine residents and invaders into a single trait table</span></span>
<span><span class="va">all_traits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span></span>
<span>  <span class="va">spp_trait</span>, <span class="co"># Resident species traits (must include 'species' column)</span></span>
<span>  <span class="va">inv_traits</span> <span class="co"># Output from simulate_invaders()</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">all_traits</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                               species trait_cont1 trait_cont2</span></span>
<span><span class="co">#&gt; Utetheisa pulchella               Utetheisa pulchella  0.02842357  -0.2030292</span></span>
<span><span class="co">#&gt; Danaus chrysippus orientis Danaus chrysippus orientis  0.03819190  -0.2237834</span></span>
<span><span class="co">#&gt; Telchinia serena                     Telchinia serena -0.83512488  -0.3065035</span></span>
<span><span class="co">#&gt; Vanessa cardui                         Vanessa cardui -0.21959307   0.5693856</span></span>
<span><span class="co">#&gt; Hypolimnas misippus               Hypolimnas misippus  0.31398458   0.6658322</span></span>
<span><span class="co">#&gt; Pieris brassicae                     Pieris brassicae -0.76502528  -0.1364975</span></span>
<span><span class="co">#&gt;                            trait_cont3 trait_cont4 trait_cont5 trait_cont6</span></span>
<span><span class="co">#&gt; Utetheisa pulchella        -0.99685889   0.4797106  0.87477170   0.8696459</span></span>
<span><span class="co">#&gt; Danaus chrysippus orientis  0.02882587  -0.5325932 -0.09453685  -0.7031068</span></span>
<span><span class="co">#&gt; Telchinia serena            0.02881542   0.9252160  0.26301460   0.3603284</span></span>
<span><span class="co">#&gt; Vanessa cardui              0.16320801   0.4664918  0.70096550   0.1009882</span></span>
<span><span class="co">#&gt; Hypolimnas misippus         0.51908854  -0.3895633 -0.99723831   0.5587363</span></span>
<span><span class="co">#&gt; Pieris brassicae           -0.71904181   0.4879493 -0.21005391   0.5576190</span></span>
<span><span class="co">#&gt;                            trait_cont7 trait_cont8 trait_cont9 trait_cont10</span></span>
<span><span class="co">#&gt; Utetheisa pulchella          0.5863824  -0.0973783   0.1228759  0.850091782</span></span>
<span><span class="co">#&gt; Danaus chrysippus orientis  -0.3658933  -0.8554938  -0.6567358 -0.001454239</span></span>
<span><span class="co">#&gt; Telchinia serena             0.8358081  -0.8781005  -0.8647264 -0.589900776</span></span>
<span><span class="co">#&gt; Vanessa cardui              -0.7333408   0.6775101  -0.8585562  0.773507196</span></span>
<span><span class="co">#&gt; Hypolimnas misippus          0.4594602  -0.7753835  -0.3726324 -0.623131341</span></span>
<span><span class="co">#&gt; Pieris brassicae             0.5380649   0.9354673  -0.9649034  0.429707387</span></span>
<span><span class="co">#&gt;                            trait_cat11 trait_cat12  trait_cat13 trait_cat14</span></span>
<span><span class="co">#&gt; Utetheisa pulchella            wetland     diurnal multivoltine detritivore</span></span>
<span><span class="co">#&gt; Danaus chrysippus orientis   grassland     diurnal   univoltine detritivore</span></span>
<span><span class="co">#&gt; Telchinia serena                forest   nocturnal multivoltine  generalist</span></span>
<span><span class="co">#&gt; Vanessa cardui                  forest     diurnal   univoltine  generalist</span></span>
<span><span class="co">#&gt; Hypolimnas misippus             forest     diurnal   univoltine nectarivore</span></span>
<span><span class="co">#&gt; Pieris brassicae             grassland   nocturnal   univoltine  generalist</span></span>
<span><span class="co">#&gt;                            trait_cat15 trait_ord16 trait_ord17 trait_bin18</span></span>
<span><span class="co">#&gt; Utetheisa pulchella          migratory           3           4           1</span></span>
<span><span class="co">#&gt; Danaus chrysippus orientis   migratory           1           4           1</span></span>
<span><span class="co">#&gt; Telchinia serena             migratory           2           5           1</span></span>
<span><span class="co">#&gt; Vanessa cardui                resident           3           3           0</span></span>
<span><span class="co">#&gt; Hypolimnas misippus           resident           1           4           0</span></span>
<span><span class="co">#&gt; Pieris brassicae              resident           1           2           0</span></span>
<span><span class="co">#&gt;                            trait_bin19 trait_ord20</span></span>
<span><span class="co">#&gt; Utetheisa pulchella                  1       small</span></span>
<span><span class="co">#&gt; Danaus chrysippus orientis           1      medium</span></span>
<span><span class="co">#&gt; Telchinia serena                     1       large</span></span>
<span><span class="co">#&gt; Vanessa cardui                       0       large</span></span>
<span><span class="co">#&gt; Hypolimnas misippus                  0       large</span></span>
<span><span class="co">#&gt; Pieris brassicae                     0       large</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">all_traits</span><span class="op">)</span></span>
<span><span class="co">#&gt;       species trait_cont1 trait_cont2 trait_cont3 trait_cont4 trait_cont5</span></span>
<span><span class="co">#&gt; inv5     inv5   0.8693445   0.6658322  0.16320801   0.8660683  0.65788426</span></span>
<span><span class="co">#&gt; inv6     inv6  -0.2195931  -0.1060607  0.51908854   0.3348530  0.14695180</span></span>
<span><span class="co">#&gt; inv7     inv7   0.4382245   0.9151532  0.38640964  -0.5656846 -0.99723831</span></span>
<span><span class="co">#&gt; inv8     inv8   0.2834910   0.6221103  0.51908854   0.5516467  0.70096550</span></span>
<span><span class="co">#&gt; inv9     inv9   0.4731766   0.8132028 -0.56722917   0.8849114 -0.09453685</span></span>
<span><span class="co">#&gt; inv10   inv10   0.6608953   0.4751912  0.02882587   0.1272937 -0.28866809</span></span>
<span><span class="co">#&gt;       trait_cont6 trait_cont7 trait_cont8 trait_cont9 trait_cont10 trait_cat11</span></span>
<span><span class="co">#&gt; inv5   -0.8394711  0.48453290 -0.42450050  -0.6819552  -0.95709954   grassland</span></span>
<span><span class="co">#&gt; inv6   -0.9418284 -0.36589330  0.67751007  -0.3726324  -0.62313134     wetland</span></span>
<span><span class="co">#&gt; inv7    0.2472269  0.72525955 -0.11440746   0.6410291   0.09924083      forest</span></span>
<span><span class="co">#&gt; inv8   -0.6596750  0.53806485  0.03422211  -0.2932992  -0.33153735      forest</span></span>
<span><span class="co">#&gt; inv9    0.5576190 -0.09622701  0.70386197  -0.3726324  -0.37790076   grassland</span></span>
<span><span class="co">#&gt; inv10   0.5576190 -0.14501107 -0.11535072  -0.1572432  -0.33153735     wetland</span></span>
<span><span class="co">#&gt;       trait_cat12  trait_cat13 trait_cat14 trait_cat15 trait_ord16 trait_ord17</span></span>
<span><span class="co">#&gt; inv5      diurnal    bivoltine nectarivore    resident           2           1</span></span>
<span><span class="co">#&gt; inv6      diurnal   univoltine  generalist   migratory           1           3</span></span>
<span><span class="co">#&gt; inv7    nocturnal multivoltine  generalist    resident           4           4</span></span>
<span><span class="co">#&gt; inv8      diurnal    bivoltine detritivore    resident           1           5</span></span>
<span><span class="co">#&gt; inv9      diurnal    bivoltine nectarivore   migratory           1           2</span></span>
<span><span class="co">#&gt; inv10     diurnal multivoltine detritivore    resident           1           4</span></span>
<span><span class="co">#&gt;       trait_bin18 trait_bin19 trait_ord20</span></span>
<span><span class="co">#&gt; inv5            0           1      medium</span></span>
<span><span class="co">#&gt; inv6            0           1       large</span></span>
<span><span class="co">#&gt; inv7            1           0       large</span></span>
<span><span class="co">#&gt; inv8            0           1       small</span></span>
<span><span class="co">#&gt; inv9            0           1      medium</span></span>
<span><span class="co">#&gt; inv10           1           1      medium</span></span>
<span></span>
<span><span class="co"># Predict across all sites Ã— all species using predict_invader_response()</span></span>
<span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/predict_invader_response.html">predict_invader_response</a></span><span class="op">(</span></span>
<span>  model          <span class="op">=</span> <span class="va">mod</span>,</span>
<span>  species_traits <span class="op">=</span> <span class="va">all_traits</span>, <span class="co"># residents + invaders (traits)</span></span>
<span>  site_env       <span class="op">=</span> <span class="va">site_env</span>, <span class="co"># base site predictors</span></span>
<span>  species_col    <span class="op">=</span> <span class="st">"species"</span>,</span>
<span>  site_col       <span class="op">=</span> <span class="st">"site_id"</span>,</span>
<span>  response_type  <span class="op">=</span> <span class="st">"response"</span>,</span>
<span>  include_random <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  site_aug       <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">spp_rich_obs</span>, <span class="va">site_id</span>, <span class="va">obs_sum</span>, <span class="va">spp_rich</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect outputs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">pred</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 3</span></span>
<span><span class="co">#&gt;  $ newdata          : tibble [15,355 Ã— 36] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;  $ predictions      :'data.frame':   15355 obs. of  3 variables:</span></span>
<span><span class="co">#&gt;  $ prediction_matrix: num [1:415, 1:37] 2.27 1.77 2.99 2.08 1.78 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pred</span><span class="op">$</span><span class="va">newdata</span><span class="op">)</span> <span class="co"># long table: site_id, species, pred</span></span>
<span><span class="co">#&gt;  [1] "species"      "site_id"      "trait_cont1"  "trait_cont2"  "trait_cont3" </span></span>
<span><span class="co">#&gt;  [6] "trait_cont4"  "trait_cont5"  "trait_cont6"  "trait_cont7"  "trait_cont8" </span></span>
<span><span class="co">#&gt; [11] "trait_cont9"  "trait_cont10" "trait_cat11"  "trait_cat12"  "trait_cat13" </span></span>
<span><span class="co">#&gt; [16] "trait_cat14"  "trait_cat15"  "trait_ord16"  "trait_ord17"  "trait_bin18" </span></span>
<span><span class="co">#&gt; [21] "trait_bin19"  "trait_ord20"  "x"            "y"            "env1"        </span></span>
<span><span class="co">#&gt; [26] "env2"         "env3"         "env4"         "env5"         "env6"        </span></span>
<span><span class="co">#&gt; [31] "env7"         "env8"         "env9"         "env10"        "obs_sum"     </span></span>
<span><span class="co">#&gt; [36] "spp_rich"</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pred</span><span class="op">$</span><span class="va">predictions</span><span class="op">)</span> <span class="co"># long table: site_id, species, pred</span></span>
<span><span class="co">#&gt;   site_id             species     pred</span></span>
<span><span class="co">#&gt; 1    1026 Utetheisa pulchella 1.761536</span></span>
<span><span class="co">#&gt; 2    1027 Utetheisa pulchella 1.561990</span></span>
<span><span class="co">#&gt; 3    1028 Utetheisa pulchella 1.771892</span></span>
<span><span class="co">#&gt; 4    1029 Utetheisa pulchella 1.364692</span></span>
<span><span class="co">#&gt; 5    1030 Utetheisa pulchella 1.843295</span></span>
<span><span class="co">#&gt; 6    1031 Utetheisa pulchella 1.359066</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pred</span><span class="op">$</span><span class="va">prediction_matrix</span><span class="op">)</span> <span class="co"># sites Ã— species matrix (for fitness/impact/risk calcs)</span></span>
<span><span class="co">#&gt; [1] 415  37</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="id_10-biotic--abiotic-constraints">10. Biotic &amp; Abiotic Constraints<a class="anchor" aria-label="anchor" href="#id_10-biotic--abiotic-constraints"></a>
</h2>
<p>Understanding how <strong>trait similarity</strong> (biotic constraint) and <strong>environmental matching</strong> (abiotic constraint) shape invasion outcomes requires explicit, quantitative representations of both processes. This section details three linked steps:</p>
<ol style="list-style-type: decimal">
<li>Constructing the general <strong>trait-based interaction strength</strong> matrix <a href="#def-gall"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>g</mi><mrow><mi>i</mi><mi>j</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">a</mi><mi mathvariant="normal">l</mi><mi mathvariant="normal">l</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><annotation encoding="application/x-tex">g^{\mathrm{(all)}}_{ij}</annotation></semantics></math></a>
</li>
<li>Deriving <strong>competition coefficients</strong> <a href="#def-alpha"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\alpha_{ij}</annotation></semantics></math></a> from the underlying trait dissimilarities <a href="#def-dij"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>d</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">d_{ij}</annotation></semantics></math></a>
</li>
<li>Quantifying <strong>environmental matching</strong> via speciesâ€™ optima <a href="#def-Ke"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>K</mi><mi>e</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><mo>;</mo><msub><mi>Ïƒ</mi><mi>e</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K_e(\Delta_{js};\sigma_e)</annotation></semantics></math></a>
</li>
</ol>
<div class="section level3">
<h3 id="id_101-interaction-strength">10.1 Interaction Strength<a class="anchor" aria-label="anchor" href="#id_101-interaction-strength"></a>
</h3>
<p>The general trait-based interaction strength <a href="#def-gall"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>g</mi><mrow><mi>i</mi><mi>j</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">a</mi><mi mathvariant="normal">l</mi><mi mathvariant="normal">l</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><annotation encoding="application/x-tex">g^{\mathrm{(all)}}_{ij}</annotation></semantics></math></a> (<code>g_all</code>) quantifies the dissimilarity between the trait vectors <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ğ­</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\mathbf{t}_i</annotation></semantics></math> and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ğ­</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\mathbf{t}_j</annotation></semantics></math> for any two species <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>. When <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>g</mi><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">a</mi><mi mathvariant="normal">l</mi><mi mathvariant="normal">l</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msup><annotation encoding="application/x-tex">g^{\mathrm{(all)}}</annotation></semantics></math> is computed as a Gower distance, it is numerically identical to <a href="#def-dij"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>d</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">d_{ij}</annotation></semantics></math></a>.</p>
<p>We also compute the <strong>resident Ã— site abundance matrix</strong> <a href="#def-Nstar"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>N</mi><mrow><mi>j</mi><mi>s</mi></mrow><mo>*</mo></msubsup><annotation encoding="application/x-tex">N^{*}_{js}</annotation></semantics></math></a> (<code>Nstar</code>), representing the predicted equilibrium abundance of resident species <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math> at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.</p>
<p><strong><code><a href="reference/compute_interaction_strength.html">compute_interaction_strength()</a></code> outputs:</strong></p>
<ul>
<li>
<code>d_ij</code> â€” trait dissimilarity matrix (residents + invaders)</li>
<li>
<code>Nstar</code> â€” resident Ã— site abundance matrix from <code>pred$predictions</code>
</li>
</ul>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify all resident species (excluding invaders)</span></span>
<span><span class="va">residents</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spp_trait</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute interaction strengths and resident abundance matrix</span></span>
<span><span class="va">cis</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/compute_interaction_strength.html">compute_interaction_strength</a></span><span class="op">(</span></span>
<span>  traits      <span class="op">=</span> <span class="va">all_traits</span>, <span class="co"># residents + invaders</span></span>
<span>  predDF      <span class="op">=</span> <span class="va">pred</span><span class="op">$</span><span class="va">predictions</span>, <span class="co"># predictions from predict_invader_response()</span></span>
<span>  method      <span class="op">=</span> <span class="st">"gower"</span>, <span class="co"># mixed traits</span></span>
<span>  kernel      <span class="op">=</span> <span class="st">"distance"</span>, <span class="co"># keep as dissimilarity</span></span>
<span>  standardise <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># Gower is already [0,1]</span></span>
<span>  sparsify_k  <span class="op">=</span> <span class="cn">NULL</span> <span class="co"># no sparsification for now</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># site Ã— resident abundance matrix</span></span>
<span><span class="co"># - Nstar is built from predictions and restricted to resident species.</span></span>
<span><span class="va">Nstar</span> <span class="op">&lt;-</span> <span class="va">cis</span><span class="op">$</span><span class="va">Nstar</span> <span class="co"># site Ã— residents</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Nstar</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                1026     1027     1028     1029</span></span>
<span><span class="co">#&gt; Utetheisa pulchella        1.761536 1.561990 1.771892 1.364692</span></span>
<span><span class="co">#&gt; Danaus chrysippus orientis 3.239507 3.001678 3.677266 3.446281</span></span>
<span></span>
<span><span class="co"># trait-based distances/dissimilarities (interaction strengths)</span></span>
<span><span class="co"># - g_all[i, j] is the trait dissimilarity between species i and j.</span></span>
<span><span class="va">g_all</span> <span class="op">&lt;-</span> <span class="va">cis</span><span class="op">$</span><span class="va">g_all</span> <span class="co"># species Ã— species</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">g_all</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                            Utetheisa pulchella Danaus chrysippus orientis</span></span>
<span><span class="co">#&gt; Utetheisa pulchella                   0.000000                   0.398072</span></span>
<span><span class="co">#&gt; Danaus chrysippus orientis            0.398072                   0.000000</span></span>
<span><span class="co">#&gt;                            Telchinia serena Vanessa cardui</span></span>
<span><span class="co">#&gt; Utetheisa pulchella               0.4164589      0.5341222</span></span>
<span><span class="co">#&gt; Danaus chrysippus orientis        0.4409018      0.5248909</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="id_102-competition">10.2 Competition<a class="anchor" aria-label="anchor" href="#id_102-competition"></a>
</h3>
<p>From the trait dissimilarities <a href="#def-dij"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>d</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">d_{ij}</annotation></semantics></math></a> (obtained from <a href="#def-gall"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>g</mi><mrow><mi>i</mi><mi>j</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">a</mi><mi mathvariant="normal">l</mi><mi mathvariant="normal">l</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><annotation encoding="application/x-tex">g^{\mathrm{(all)}}_{ij}</annotation></semantics></math></a> when it is a distance matrix), we derive <strong>trait-based competition coefficients</strong> <a href="#def-alpha"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\alpha_{ij}</annotation></semantics></math></a> using the Gaussian trait kernel.</p>
<p><strong><code><a href="reference/compute_competition_kernel.html">compute_competition_kernel()</a></code> outputs:</strong></p>
<ul>
<li>
<code>a_ij</code> â€” invader Ã— resident competition coefficients <a href="#def-alpha"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\alpha_{ij}</annotation></semantics></math></a>
</li>
<li>
<code>sigma_t</code> â€” bandwidth parameter used <a href="#def-sigmat"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Ïƒ</mi><mi>t</mi></msub><annotation encoding="application/x-tex">\sigma_t</annotation></semantics></math></a>
</li>
<li>
<code>d_ij</code> â€” distances used in kernel</li>
</ul>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># --- Competition coefficients from trait distances ---</span></span>
<span><span class="va">comp</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/compute_competition_kernel.html">compute_competition_kernel</a></span><span class="op">(</span></span>
<span>  g_all <span class="op">=</span> <span class="va">g_all</span>, <span class="co"># trait distance matrix (e.g., Gower)</span></span>
<span>  residents <span class="op">=</span> <span class="va">residents</span>, <span class="co"># resident IDs</span></span>
<span>  <span class="co"># invaders = paste0("inv", 1:n_inv),  # optional; default = non-residents</span></span>
<span>  <span class="co"># sigma_t  = NULL,             # optional; auto from resident-resident distances</span></span>
<span>  sigma_method <span class="op">=</span> <span class="st">"sd"</span> <span class="co"># or "median", "iqr"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">sigma_t</span> <span class="op">&lt;-</span> <span class="va">comp</span><span class="op">$</span><span class="va">sigma_t</span> <span class="co"># bandwidth used</span></span>
<span><span class="va">d_ij</span> <span class="op">&lt;-</span> <span class="va">comp</span><span class="op">$</span><span class="va">d_ij</span> <span class="co"># invader-resident distances</span></span>
<span><span class="va">a_ij</span> <span class="op">&lt;-</span> <span class="va">comp</span><span class="op">$</span><span class="va">a_ij</span> <span class="co"># Gaussian competition coefficients</span></span>
<span></span>
<span><span class="va">sigma_t</span></span>
<span><span class="co">#&gt; [1] 0.08758267</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">d_ij</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;      Utetheisa pulchella Danaus chrysippus orientis Telchinia serena</span></span>
<span><span class="co">#&gt; inv1           0.4428054                  0.3371734        0.3626310</span></span>
<span><span class="co">#&gt; inv2           0.4224789                  0.4655103        0.3639053</span></span>
<span><span class="co">#&gt; inv3           0.4136122                  0.4075364        0.3784909</span></span>
<span><span class="co">#&gt; inv4           0.3899103                  0.3182808        0.2564511</span></span>
<span><span class="co">#&gt;      Vanessa cardui</span></span>
<span><span class="co">#&gt; inv1      0.5756327</span></span>
<span><span class="co">#&gt; inv2      0.4594073</span></span>
<span><span class="co">#&gt; inv3      0.5894313</span></span>
<span><span class="co">#&gt; inv4      0.5408265</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">a_ij</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;      Utetheisa pulchella Danaus chrysippus orientis Telchinia serena</span></span>
<span><span class="co">#&gt; inv1        2.814174e-06               6.049429e-04     1.894032e-04</span></span>
<span><span class="co">#&gt; inv2        8.856219e-06               7.337331e-07     1.783117e-04</span></span>
<span><span class="co">#&gt; inv3        1.435855e-05               1.987671e-05     8.803305e-05</span></span>
<span><span class="co">#&gt; inv4        4.968745e-05               1.356017e-03     1.374756e-02</span></span>
<span><span class="co">#&gt;      Vanessa cardui</span></span>
<span><span class="co">#&gt; inv1   4.167436e-10</span></span>
<span><span class="co">#&gt; inv2   1.060064e-06</span></span>
<span><span class="co">#&gt; inv3   1.461412e-10</span></span>
<span><span class="co">#&gt; inv4   5.247268e-09</span></span>
<span></span>
<span><span class="co"># Example use with N* (from Section 7.1):</span></span>
<span><span class="co"># For invader i at site s, total expected competitive pressure (scalar) could be:</span></span>
<span><span class="co">#   pressure_{i,s} = sum_j a_{ij} * N*_j,s</span></span>
<span><span class="co"># which combines trait overlap (a_ij) with resident context (N*).</span></span></code></pre></div>
<blockquote>
<p><strong>Note:</strong> * High trait similarity (low <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>d</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">d_{ij}</annotation></semantics></math>) â†’ strong competition (high <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\alpha_{ij}</annotation></semantics></math>) * High trait difference (high <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>d</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">d_{ij}</annotation></semantics></math>) â†’ weak competition (low <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\alpha_{ij}</annotation></semantics></math>) * Lowering <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Ïƒ</mi><mi>t</mi></msub><annotation encoding="application/x-tex">\sigma_t</annotation></semantics></math> sharpens niche separation; useful for sensitivity tests * Ensure <code>g_all</code> is a <em>distance</em> matrix before applying the Gaussian kernel; if itâ€™s already kernelised, retrieve the original <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>d</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">d_{ij}</annotation></semantics></math> first Â </p>
</blockquote>
</div>
<div class="section level3">
<h3 id="id_103-environmental-filtering">10.3 Environmental Filtering<a class="anchor" aria-label="anchor" href="#id_103-environmental-filtering"></a>
</h3>
<p>We estimate each speciesâ€™ <strong>environmental optimum</strong> from predicted abundances and site-level environmental variables, then compute siteâ€“species distances <a href="#def-delta"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\Delta_{js}</annotation></semantics></math></a>. These are converted to the Gaussian environmental similarity kernel <a href="#def-Ke"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>K</mi><mi>e</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><mo>;</mo><msub><mi>Ïƒ</mi><mi>e</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K_e(\Delta_{js};\sigma_e)</annotation></semantics></math></a> using bandwidth <a href="#def-sigmae"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Ïƒ</mi><mi>e</mi></msub><annotation encoding="application/x-tex">\sigma_e</annotation></semantics></math></a>.</p>
<p><strong><code><a href="reference/compute_environment_kernel.html">compute_environment_kernel()</a></code> outputs:</strong></p>
<ul>
<li>
<code>env_opt</code> â€” species Ã— environmental optima</li>
<li>
<code>env_dist</code> â€” site Ã— species environmental distances</li>
<li>
<code>K_env</code> â€” Gaussian environmental similarity kernel</li>
</ul>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Build environmental kernel and distances</span></span>
<span><span class="va">ek</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/compute_environment_kernel.html">compute_environment_kernel</a></span><span class="op">(</span></span>
<span>  site_env <span class="op">=</span> <span class="va">site_env</span>, <span class="co"># site_id, x, y, env1:env10, ...</span></span>
<span>  predictions <span class="op">=</span> <span class="va">pred</span><span class="op">$</span><span class="va">predictions</span>, <span class="co"># species, site_id, pred (from predict_invader_response)</span></span>
<span>  site_col <span class="op">=</span> <span class="st">"site_id"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"gower"</span>,</span>
<span>  kernel <span class="op">=</span> <span class="st">"gaussian"</span>, <span class="co"># "distance" or "similarity" also supported</span></span>
<span>  sigma_method <span class="op">=</span> <span class="st">"sd"</span> <span class="co"># or "median", "iqr"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">sigma_e</span> <span class="op">&lt;-</span> <span class="va">ek</span><span class="op">$</span><span class="va">sigma_e</span> <span class="co"># bandwidth used for Gaussian kernel</span></span>
<span><span class="va">env_opt</span> <span class="op">&lt;-</span> <span class="va">ek</span><span class="op">$</span><span class="va">env_opt</span> <span class="co"># species Ã— env (optima)</span></span>
<span><span class="va">env_dist</span> <span class="op">&lt;-</span> <span class="va">ek</span><span class="op">$</span><span class="va">env_dist</span> <span class="co"># sites Ã— species (distance)</span></span>
<span><span class="va">g_env</span> <span class="op">&lt;-</span> <span class="va">ek</span><span class="op">$</span><span class="va">K_env</span> <span class="co"># sites Ã— species (Gaussian similarity), if requested</span></span>
<span></span>
<span><span class="va">sigma_e</span></span>
<span><span class="co">#&gt; [1] 0.0495915</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">env_opt</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                   env1        env2        env3       env4</span></span>
<span><span class="co">#&gt; Utetheisa pulchella        -0.45572568 -0.25326724 -0.05813116  0.1956375</span></span>
<span><span class="co">#&gt; Danaus chrysippus orientis -0.09139371 -0.19942460  0.22554528  0.1762052</span></span>
<span><span class="co">#&gt; Telchinia serena           -0.15295732 -0.10760847  0.16373671  0.2630910</span></span>
<span><span class="co">#&gt; Vanessa cardui              0.17546879 -0.03901351  0.02578274 -0.1113054</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">env_dist</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;      Utetheisa pulchella Danaus chrysippus orientis Telchinia serena</span></span>
<span><span class="co">#&gt; 1026           0.2919083                  0.2901035        0.2891394</span></span>
<span><span class="co">#&gt; 1027           0.3004139                  0.2986090        0.2976450</span></span>
<span><span class="co">#&gt; 1028           0.3051308                  0.3033259        0.3023619</span></span>
<span><span class="co">#&gt; 1029           0.3230038                  0.3211989        0.3202349</span></span>
<span><span class="co">#&gt;      Vanessa cardui</span></span>
<span><span class="co">#&gt; 1026      0.2443649</span></span>
<span><span class="co">#&gt; 1027      0.2528705</span></span>
<span><span class="co">#&gt; 1028      0.2575874</span></span>
<span><span class="co">#&gt; 1029      0.2754604</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">g_env</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;      Utetheisa pulchella Danaus chrysippus orientis Telchinia serena</span></span>
<span><span class="co">#&gt; 1026        2.994181e-08               3.707048e-08     4.152730e-08</span></span>
<span><span class="co">#&gt; 1027        1.075079e-08               1.339372e-08     1.505410e-08</span></span>
<span><span class="co">#&gt; 1028        6.015103e-09               7.519820e-09     8.467670e-09</span></span>
<span><span class="co">#&gt; 1029        6.137287e-10               7.773871e-10     8.815289e-10</span></span>
<span><span class="co">#&gt;      Vanessa cardui</span></span>
<span><span class="co">#&gt; 1026   5.339339e-06</span></span>
<span><span class="co">#&gt; 1027   2.259750e-06</span></span>
<span><span class="co">#&gt; 1028   1.385049e-06</span></span>
<span><span class="co">#&gt; 1029   1.996438e-07</span></span></code></pre></div>
<blockquote>
<p><strong>Note</strong>: For environmental kernel context, these matrices are directly used in the invasion fitness computation in Section 8. Â </p>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="id_11-invasion-fitness">11. Invasion Fitness<a class="anchor" aria-label="anchor" href="#id_11-invasion-fitness"></a>
</h2>
<p>The invasion fitness of each hypothetical invader at each site is determined by combining predicted growth potential <a href="#def-r"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>r</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">r_{is}</annotation></semantics></math></a>, the summed effect of trait-based competition <a href="#def-Craw"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><annotation encoding="application/x-tex">C^{\mathrm{(raw)}}_{is}</annotation></semantics></math></a>, and the degree of environmental matching <a href="#def-Ke"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>K</mi><mi>e</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><mo>;</mo><msub><mi>Ïƒ</mi><mi>e</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K_e(\Delta_{js};\sigma_e)</annotation></semantics></math></a>.</p>
<div class="section level3">
<h3 id="id_111-competitive-pressure-with-environmental-context">11.1 Competitive Pressure with Environmental Context<a class="anchor" aria-label="anchor" href="#id_111-competitive-pressure-with-environmental-context"></a>
</h3>
<p>For site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>, invader <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>, resident <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>, the <strong>impact tensor</strong> <a href="#def-I"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mrow><mi>i</mi><mi>j</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">I_{ijs}</annotation></semantics></math></a> combines:</p>
<ul>
<li>competition coefficient <a href="#def-alpha"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\alpha_{ij}</annotation></semantics></math></a>
</li>
<li>environmental similarity kernel <a href="#def-Ke"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>K</mi><mi>e</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><mo>;</mo><msub><mi>Ïƒ</mi><mi>e</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K_e(\Delta_{js};\sigma_e)</annotation></semantics></math></a>
</li>
<li>equilibrium abundance <a href="#def-Nstar"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>N</mi><mrow><mi>j</mi><mi>s</mi></mrow><mo>*</mo></msubsup><annotation encoding="application/x-tex">N^{*}_{js}</annotation></semantics></math></a>
</li>
</ul>
<p><code><a href="reference/assemble_matrices.html">assemble_matrices()</a></code> outputs:</p>
<ul>
<li>
<code>I_raw</code> â€” <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>I</mi><mrow><mi>i</mi><mi>j</mi><mi>s</mi></mrow></msub><mo stretchy="true" form="postfix">]</mo></mrow><annotation encoding="application/x-tex">[I_{ijs}]</annotation></semantics></math> impact tensor</li>
<li>
<code>pressure_inv_site</code> â€” total competitive pressure per invader Ã— site</li>
<li>
<code>C_raw</code> â€” total raw competitive penalty <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><mo stretchy="true" form="postfix">]</mo></mrow><annotation encoding="application/x-tex">[C^{\mathrm{(raw)}}_{is}]</annotation></semantics></math>
</li>
</ul>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assemble site- and species-specific competition/impact matrices</span></span>
<span><span class="va">am</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/assemble_matrices.html">assemble_matrices</a></span><span class="op">(</span></span>
<span>  a_ij            <span class="op">=</span> <span class="va">comp</span><span class="op">$</span><span class="va">a_ij</span>, <span class="co"># invader Ã— resident competition coefficients</span></span>
<span>  Nstar           <span class="op">=</span> <span class="va">cis</span><span class="op">$</span><span class="va">Nstar</span>, <span class="co"># resident abundances (residents Ã— sites)</span></span>
<span>  predictions     <span class="op">=</span> <span class="va">pred</span><span class="op">$</span><span class="va">predictions</span>, <span class="co"># provides invader r_is (sites Ã— invaders) if needed</span></span>
<span>  K_env           <span class="op">=</span> <span class="va">ek</span><span class="op">$</span><span class="va">K_env</span> <span class="co"># environmental kernel (sites Ã— residents), or env_dist + sigma_e</span></span>
<span>  <span class="co"># env_dist        = ek$env_dist,      # OR Pass distances and Ïƒ to compute the Gaussian kernel internally</span></span>
<span>  <span class="co"># sigma_e         = ek$sigma_e</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">am</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 3</span></span>
<span><span class="co">#&gt;  $ I_raw            : num [1:10, 1:27, 1:415] 1.35e-12 1.06e-11 3.40e-12 7.14e-12 2.01e-16 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 3</span></span>
<span><span class="co">#&gt;  $ pressure_inv_site: num [1:10, 1:415] 5.39e-07 1.45e-06 1.23e-06 5.49e-07 1.57e-05 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;  $ meta             :List of 6</span></span>
<span></span>
<span><span class="co"># 3D impact tensor: [invader, resident, site]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">am</span><span class="op">$</span><span class="va">I_raw</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  10  27 415</span></span>
<span></span>
<span><span class="co"># Total competitive pressure per invader Ã— site (sum over residents)</span></span>
<span><span class="va">pressure</span> <span class="op">&lt;-</span> <span class="va">am</span><span class="op">$</span><span class="va">pressure_inv_site</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pressure</span><span class="op">)</span> <span class="co"># invaders Ã— sites</span></span>
<span><span class="co">#&gt; [1]  10 415</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="id_112-invasion-fitness-calculation">11.2 Invasion Fitness Calculation<a class="anchor" aria-label="anchor" href="#id_112-invasion-fitness-calculation"></a>
</h3>
<p>Given <a href="#def-Craw"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><annotation encoding="application/x-tex">C^{\mathrm{(raw)}}_{is}</annotation></semantics></math></a> and predicted growth <a href="#def-r"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>r</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">r_{is}</annotation></semantics></math></a>, invasion fitness <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> is computed following the definition in Section 1, with possible scaling or transformation variants:</p>
<ul>
<li>
<strong>Raw penalty</strong>: <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mrow><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi></mrow><mo stretchy="true" form="postfix">)</mo></mrow></msubsup><mo>=</mo><msub><mi>r</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><mo>âˆ’</mo><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mrow><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi></mrow><mo stretchy="true" form="postfix">)</mo></mrow></msubsup></mrow><annotation encoding="application/x-tex">\lambda^{(\mathrm{raw})}_{is} = r_{is} - C^{(\mathrm{raw})}_{is}</annotation></semantics></math><br>
Â </li>
<li>
<strong>Richness-scaled</strong>: <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mrow><mi mathvariant="normal">s</mi><mi mathvariant="normal">c</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">l</mi><mi mathvariant="normal">e</mi><mi mathvariant="normal">d</mi></mrow><mo stretchy="true" form="postfix">)</mo></mrow></msubsup><mo>=</mo><msub><mi>r</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><mo>âˆ’</mo><mfrac><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mrow><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi></mrow><mo stretchy="true" form="postfix">)</mo></mrow></msubsup><msub><mi>J</mi><mi>s</mi></msub></mfrac></mrow><annotation encoding="application/x-tex">\lambda^{(\mathrm{scaled})}_{is} = r_{is} - \frac{C^{(\mathrm{raw})}_{is}}{J_s}</annotation></semantics></math>, where <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>J</mi><mi>s</mi></msub><annotation encoding="application/x-tex">J_s</annotation></semantics></math> is resident richness at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>. Â </li>
<li>
<strong>Relative-abundance weighted</strong>: <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mrow><mi mathvariant="normal">r</mi><mi mathvariant="normal">e</mi><mi mathvariant="normal">l</mi></mrow><mo stretchy="true" form="postfix">)</mo></mrow></msubsup><mo>=</mo><msub><mi>r</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><mo>âˆ’</mo><mrow><mo stretchy="true" form="prefix">(</mo><msub><mo>âˆ‘</mo><mi>j</mi></msub><msub><mi>a</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msubsup><mi>N</mi><mrow><mi>j</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mrow><mi mathvariant="normal">r</mi><mi mathvariant="normal">e</mi><mi mathvariant="normal">l</mi></mrow><mo stretchy="true" form="postfix">)</mo></mrow></msubsup><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\lambda^{(\mathrm{rel})}_{is} = r_{is} - \left( \sum_j a_{ij} N^{(\mathrm{rel})}_{js} \right)</annotation></semantics></math> Â </li>
<li>
<strong>Logistic penalty</strong> (bounded): <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mrow><mi mathvariant="normal">l</mi><mi mathvariant="normal">o</mi><mi mathvariant="normal">g</mi><mi mathvariant="normal">i</mi><mi mathvariant="normal">s</mi></mrow><mo stretchy="true" form="postfix">)</mo></mrow></msubsup><mo>=</mo><msub><mi>r</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><mo>âˆ’</mo><mrow><mi mathvariant="normal">i</mi><mi mathvariant="normal">n</mi><mi mathvariant="normal">v</mi><mi mathvariant="normal">_</mi><mi mathvariant="normal">l</mi><mi mathvariant="normal">o</mi><mi mathvariant="normal">g</mi><mi mathvariant="normal">i</mi><mi mathvariant="normal">t</mi></mrow><mspace width="-0.167em"></mspace><mrow><mo stretchy="true" form="prefix">(</mo><mi>k</mi><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>C</mi><mrow><mrow><mi mathvariant="normal">t</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">r</mi><mi mathvariant="normal">g</mi><mi mathvariant="normal">e</mi><mi mathvariant="normal">t</mi></mrow><mo>,</mo><mi>i</mi><mi>s</mi></mrow></msub><mo>âˆ’</mo><msub><mi>x</mi><mn>0</mn></msub><mo stretchy="true" form="postfix">]</mo></mrow><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\lambda^{(\mathrm{logis})}_{is} = r_{is} - \mathrm{inv\_logit}\!\left(k \left[C_{\mathrm{target},is} - x_0\right]\right)</annotation></semantics></math> Â </li>
</ul>
<p><code><a href="reference/compute_invasion_fitness.html">compute_invasion_fitness()</a></code> outputs:</p>
<ul>
<li>
<code>C_raw</code> â€” <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">[</mo><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><mo stretchy="true" form="postfix">]</mo></mrow><annotation encoding="application/x-tex">[C^{\mathrm{(raw)}}_{is}]</annotation></semantics></math> penalty matrix (depending on variant)</li>
<li>
<code>lambda</code> â€” chosen fitness variant <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><mo stretchy="true" form="postfix">]</mo></mrow><annotation encoding="application/x-tex">[\lambda_{is}]</annotation></semantics></math>
</li>
</ul>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute invasion fitness (uses am$I_raw from assemble_matrices and pred$predictions)</span></span>
<span><span class="va">fitness</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/compute_invasion_fitness.html">compute_invasion_fitness</a></span><span class="op">(</span></span>
<span>  I_raw        <span class="op">=</span> <span class="va">am</span><span class="op">$</span><span class="va">I_raw</span>, <span class="co"># or pressure_inv_site = am$pressure_inv_site</span></span>
<span>  predictions  <span class="op">=</span> <span class="va">pred</span><span class="op">$</span><span class="va">predictions</span>, <span class="co"># builds r_mat internally</span></span>
<span>  a_ij         <span class="op">=</span> <span class="va">comp</span><span class="op">$</span><span class="va">a_ij</span>, <span class="co"># needed for lambda_rel / logistic_on = "rel"</span></span>
<span>  Nstar        <span class="op">=</span> <span class="va">cis</span><span class="op">$</span><span class="va">Nstar</span>,</span>
<span>  logistic_on  <span class="op">=</span> <span class="st">"rel"</span>, <span class="co"># cap the A %*% N_rel penalty</span></span>
<span>  k            <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  x0           <span class="op">=</span> <span class="cn">NULL</span>, <span class="co"># auto: median penalty</span></span>
<span>  prefer       <span class="op">=</span> <span class="st">"logis"</span> <span class="co"># returned as $lambda</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">fitness</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 8</span></span>
<span><span class="co">#&gt;  $ C_raw        : num [1:10, 1:415] 5.39e-07 1.45e-06 1.23e-06 5.49e-07 1.57e-05 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;  $ r_mat        : num [1:10, 1:415] 9.09 22.79 4.49 2.73 7.63 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;  $ lambda_raw   : num [1:10, 1:415] 9.09 22.79 4.49 2.73 7.63 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;  $ lambda_scaled: num [1:10, 1:415] 9.09 22.79 4.49 2.73 7.63 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;  $ lambda_rel   : num [1:10, 1:415] 9.09 22.79 4.49 2.72 7.63 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;  $ lambda_logis : num [1:10, 1:415] 8.59 22.29 3.99 2.23 7.13 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;  $ lambda       : num [1:10, 1:415] 8.59 22.29 3.99 2.23 7.13 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;  $ meta         :List of 5</span></span>
<span></span>
<span><span class="co"># Key outputs</span></span>
<span><span class="va">C_raw</span> <span class="op">&lt;-</span> <span class="va">fitness</span><span class="op">$</span><span class="va">C_raw</span> <span class="co"># invader Ã— site penalty</span></span>
<span><span class="va">lambda_mat</span> <span class="op">&lt;-</span> <span class="va">fitness</span><span class="op">$</span><span class="va">lambda</span> <span class="co"># final invasion fitness (invader Ã— site)</span></span>
<span></span>
<span><span class="co"># Inspect invasion fitness matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">lambda_mat</span><span class="op">)</span> <span class="co"># invaders Ã— sites</span></span>
<span><span class="co">#&gt; [1]  10 415</span></span></code></pre></div>
<p>The diagram below shows the information flow from raw interaction strengths to final invasion fitness.</p>
<p><strong>Conceptual flow</strong></p>
<div class="sourceCode" id="cb37"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="co">#&gt; I_raw  (invader x resident x site)</span></span>
<span id="cb37-2"><a href="#cb37-2" tabindex="-1"></a><span class="co">#&gt;     â”‚  Sum over residents (j)</span></span>
<span id="cb37-3"><a href="#cb37-3" tabindex="-1"></a><span class="co">#&gt;     â–¼</span></span>
<span id="cb37-4"><a href="#cb37-4" tabindex="-1"></a><span class="co">#&gt;  C_raw  (invader x site)</span></span>
<span id="cb37-5"><a href="#cb37-5" tabindex="-1"></a><span class="co">#&gt;     â”‚  Subtract from predicted growth r_mat</span></span>
<span id="cb37-6"><a href="#cb37-6" tabindex="-1"></a><span class="co">#&gt;     â–¼</span></span>
<span id="cb37-7"><a href="#cb37-7" tabindex="-1"></a><span class="co">#&gt;  Î» (invasion fitness) variants</span></span>
<span id="cb37-8"><a href="#cb37-8" tabindex="-1"></a><span class="co">#&gt;     â”‚</span></span>
<span id="cb37-9"><a href="#cb37-9" tabindex="-1"></a><span class="co">#&gt;     â”œâ”€â”€ Î»_raw       = r - C_raw</span></span>
<span id="cb37-10"><a href="#cb37-10" tabindex="-1"></a><span class="co">#&gt;     â”œâ”€â”€ Î»_scaled    = r - (C_raw / J)</span></span>
<span id="cb37-11"><a href="#cb37-11" tabindex="-1"></a><span class="co">#&gt;     â”œâ”€â”€ Î»_rel       = r - (A %*% N_rel)</span></span>
<span id="cb37-12"><a href="#cb37-12" tabindex="-1"></a><span class="co">#&gt;     â””â”€â”€ Î»_logis     = r - inv_logit( k * (C_target - x0) )</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="id_12-visualisation-ranking--clustering">12. Visualisation, Ranking &amp; Clustering<a class="anchor" aria-label="anchor" href="#id_12-visualisation-ranking--clustering"></a>
</h2>
<p>This section summarises invasion fitness outcomes using plots and rankings that diagnose which invaders and locations pose the greatest establishment risk, show how that risk is distributed across the study region, and support both hypothesis generation and management prioritisation.</p>
<div class="section level3">
<h3 id="id_121-invasion-fitness-matrix-invader--site">12.1 Invasion Fitness Matrix (invader Ã— site)<a class="anchor" aria-label="anchor" href="#id_121-invasion-fitness-matrix-invader--site"></a>
</h3>
<p>We visualise the <strong>invasion fitness matrix</strong> <code>Î»_mat</code>, where each entry <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> represents the low-density per-capita growth rate of invader <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> in site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math> as computed by <a href="#"><code>compute_invasion_fitness()</code></a>.</p>
<p>This matrix is central to the invasion framework because it enables:</p>
<ul>
<li>
<strong>Direct comparison</strong> of all simulated invaders across all sites.</li>
<li>Identification of <strong>susceptible sites</strong> (rows/columns with consistently high <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a>).</li>
<li>Detection of <strong>broadly adapted invaders</strong> versus specialists.</li>
</ul>
<p>Ecological interpretation:</p>
<ul>
<li>
<strong>Higher</strong> <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> (lighter colours) â†’ favourable conditions and/or weak competition.</li>
<li>
<strong>Lower</strong> <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> (darker colours) â†’ strong competitive exclusion and/or poor environmental match.</li>
</ul>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert the invasion fitness matrix to long format for ggplot and attach spatial coordinates</span></span>
<span><span class="va">lambda_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">lambda_mat</span>, rownames <span class="op">=</span> <span class="st">"invader"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># Reshape from wide (invader Ã— site) to long (invader, site, fitness value)</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="op">-</span><span class="va">invader</span>, names_to <span class="op">=</span> <span class="st">"site_id"</span>, values_to <span class="op">=</span> <span class="st">"lambda"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># Add spatial coordinates for each site, matched by site_id</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">site_xy</span><span class="op">$</span><span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">site_id</span>, <span class="va">site_xy</span><span class="op">$</span><span class="va">site_id</span><span class="op">)</span><span class="op">]</span>, <span class="co"># Longitude/Easting</span></span>
<span>    y <span class="op">=</span> <span class="va">site_xy</span><span class="op">$</span><span class="va">y</span><span class="op">[</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">site_id</span>, <span class="va">site_xy</span><span class="op">$</span><span class="va">site_id</span><span class="op">)</span><span class="op">]</span> <span class="co"># Latitude/Northing</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># # Visualize the full invader Ã— site invasion fitness matrix as a heatmap</span></span>
<span><span class="co"># ggplot(lambda_df, aes(x = site_id, y = invader, fill = lambda)) +</span></span>
<span><span class="co">#   geom_tile(color = "grey90") +  # Draw grid with light borders for readability</span></span>
<span><span class="co">#   scale_fill_viridis_c(option = "magma", # Use perceptually uniform color palette (magma)</span></span>
<span><span class="co">#     direction = -1, # Reverse so high = bright/light</span></span>
<span><span class="co">#     name = expression("Invasion fitness" ~ lambda[i*k])) +</span></span>
<span><span class="co">#   labs(title = "Invasion Fitness Matrix: Invaders Across Sites",  # Plot title</span></span>
<span><span class="co">#        x = "Site",                                                # X-axis: sites</span></span>
<span><span class="co">#        y = "Invader") +                                           # Y-axis: invaders</span></span>
<span><span class="co">#   theme_minimal(base_size = 12) +</span></span>
<span><span class="co">#   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, size = 8),  # Vertical site labels</span></span>
<span><span class="co">#         axis.text.y = element_text(size = 8),                           # Compact invader labels</span></span>
<span><span class="co">#         panel.grid = element_blank()                                    # Remove grid lines for clarity</span></span>
<span><span class="co">#   )</span></span>
<span></span>
<span><span class="co"># Orders: sites by space; invaders by mean lambda</span></span>
<span><span class="va">site_order</span> <span class="op">&lt;-</span> <span class="va">site_xy</span><span class="op">$</span><span class="va">site_id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">site_xy</span><span class="op">$</span><span class="va">x</span>, <span class="va">site_xy</span><span class="op">$</span><span class="va">y</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">inv_order</span> <span class="op">&lt;-</span> <span class="va">lambda_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">invader</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>mu <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">lambda</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, .groups <span class="op">=</span> <span class="st">"drop"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">mu</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">invader</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lambda_df</span> <span class="op">&lt;-</span> <span class="va">lambda_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    site_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">site_id</span>, levels <span class="op">=</span> <span class="va">site_order</span><span class="op">)</span>,</span>
<span>    invader <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">invader</span>, levels <span class="op">=</span> <span class="va">inv_order</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># 2) Color scaling: winsorize high tail (or set trans = "sqrt")</span></span>
<span><span class="va">cap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">lambda_df</span><span class="op">$</span><span class="va">lambda</span>, <span class="fl">0.99</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># library(scales)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">lambda_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">site_id</span>, y <span class="op">=</span> <span class="va">invader</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="va">lambda</span>, <span class="va">cap</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_raster</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="co"># faster, cleaner than geom_tile with borders</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span></span>
<span>    option <span class="op">=</span> <span class="st">"magma"</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>    limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">cap</span><span class="op">)</span>, oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/oob.html" class="external-link">squish</a></span>,</span>
<span>    name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="st">"Invasion fitness"</span> <span class="op">~</span> <span class="va">lambda</span><span class="op">[</span><span class="va">i</span> <span class="op">*</span> <span class="va">k</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Invasion Fitness Matrix: Invaders Across Sites"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Site"</span>, y <span class="op">=</span> <span class="st">"Invader"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">0.5</span>, size <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># 3) show every Nth site label to avoid overlap</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_x_discrete</a></span><span class="op">(</span></span>
<span>    breaks <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, by <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_colourbar.html" class="external-link">guide_colorbar</a></span><span class="op">(</span></span>
<span>    title.position <span class="op">=</span> <span class="st">"top"</span>,</span>
<span>    barheight <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">40</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>    barwidth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"mm"</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-plot-fitness-1.png" width="100%" style="display: block; margin: auto;"></p>
<p>This figure serves as a diagnostic map of invasibility and invader performance, feeding into later steps like risk ranking, vulnerability hotspot mapping, or scenario testing.</p>
<hr>
</div>
<div class="section level3">
<h3 id="id_122-invader-invasiveness-ranking">12.2 Invader Invasiveness Ranking<a class="anchor" aria-label="anchor" href="#id_122-invader-invasiveness-ranking"></a>
</h3>
<p>From the invasion fitness matrix <code>Î»_mat</code>, we derive the <strong>species-level invasiveness</strong> <a href="#def-Ii"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mi>i</mi></msub><annotation encoding="application/x-tex">I_i</annotation></semantics></math></a>, defined as the mean (or sum) of <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> over all sites for invader <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>. This integrates both <strong>biotic resistance</strong> and <strong>environmental filtering</strong> across the full landscape, producing a single quantitative measure of establishment potential.</p>
<p>Ranking invaders by <a href="#def-Ii"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mi>i</mi></msub><annotation encoding="application/x-tex">I_i</annotation></semantics></math></a> reveals:</p>
<ul>
<li>
<strong>High-risk â€œsuper-invadersâ€</strong> â€” species with high and geographically broad <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a>, posing widespread establishment potential.</li>
<li>
<strong>Low-risk invaders</strong> â€” species predicted to be strongly excluded in most or all sites.</li>
</ul>
<p>Such rankings are valuable for <strong>risk prioritisation</strong>, focusing management and surveillance on species most likely to succeed under current conditions. In plots, bar height represents cumulative <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> (summed over sites), ordered from highest to lowest <a href="#def-Ii"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mi>i</mi></msub><annotation encoding="application/x-tex">I_i</annotation></semantics></math></a>.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Sum invasion fitness across sites for each invader and order from highest to lowest</span></span>
<span><span class="va">rank_df</span> <span class="op">&lt;-</span> <span class="va">lambda_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">invader</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>total_lambda <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">lambda</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">total_lambda</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>invader <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">invader</span>, levels <span class="op">=</span> <span class="va">invader</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Bar plot of total invasion fitness by invader</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>  <span class="va">rank_df</span>,</span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">invader</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">total_lambda</span><span class="op">)</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">total_lambda</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.7</span>, color <span class="op">=</span> <span class="st">"grey50"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span></span>
<span>    option <span class="op">=</span> <span class="st">"inferno"</span>,</span>
<span>    name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="st">"Total fitness"</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">lambda</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Invader Ranking by Total Growth Potential"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Invader"</span>,</span>
<span>    y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="st">"Total invasion fitness "</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">lambda</span><span class="op">[</span><span class="va">i</span> <span class="op">*</span> <span class="va">k</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>    panel.grid.major.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>    legend.position <span class="op">=</span> <span class="st">"top"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-invader-rank-1.png" width="100%" style="display: block; margin: auto;"></p>
<hr>
</div>
<div class="section level3">
<h3 id="id_123-site-level-invasibility-metrics">12.3 Site-level Invasibility Metrics<a class="anchor" aria-label="anchor" href="#id_123-site-level-invasibility-metrics"></a>
</h3>
<p>We summarise <code>Î»_mat</code> across invaders to quantify <strong>site-level invasibility</strong> <a href="#def-Vs"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mi>s</mi></msub><annotation encoding="application/x-tex">V_s</annotation></semantics></math></a> â€” the openness of each community to establishment. For each site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>, we compute:</p>
<ul>
<li>
<strong>Proportion of invaders with positive fitness</strong> â€” fraction of invaders with <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><mo>&gt;</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\lambda_{is} &gt; 0</annotation></semantics></math></a>.</li>
<li>
<strong>Count of successful invaders</strong> â€” number of <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> with <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><mo>&gt;</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\lambda_{is} &gt; 0</annotation></semantics></math></a>.</li>
<li>
<strong>Mean invasion fitness</strong> â€” average <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> across invaders, indicating general establishment potential.</li>
</ul>
<p>Spatial mapping of <a href="#def-Vs"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mi>s</mi></msub><annotation encoding="application/x-tex">V_s</annotation></semantics></math></a> reveals:</p>
<ul>
<li>
<strong>Hotspots</strong> â€” sites with high proportions of positive <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a>, indicating low biotic resistance and/or favourable abiotic conditions.</li>
<li>
<strong>Potential refugia</strong> â€” sites with low proportions, where community structure or environmental context limits establishment.</li>
</ul>
<p>Tile labels often show counts of successful invaders to complement proportional risk with absolute numbers.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Summarize invasion fitness at each site:</span></span>
<span><span class="va">site_sum</span> <span class="op">&lt;-</span> <span class="va">lambda_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">site_id</span>, <span class="va">x</span>, <span class="va">y</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span></span>
<span>    mean_l   <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">lambda</span><span class="op">)</span>, <span class="co"># Mean invasion fitness across invaders</span></span>
<span>    prop_pos <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">lambda</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>, <span class="co"># Proportion of invaders with positive fitness</span></span>
<span>    cnt_pos  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">lambda</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>, <span class="co"># Number of successful invaders</span></span>
<span>    cnt_neg  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">lambda</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">)</span> <span class="co"># Number of unsuccessful invaders</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Map proportion of successful invaders at each site</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">site_sum</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, fill <span class="op">=</span> <span class="va">prop_pos</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span>color <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">rsa</span>, inherit.aes <span class="op">=</span> <span class="cn">FALSE</span>, fill <span class="op">=</span> <span class="cn">NA</span>, color <span class="op">=</span> <span class="st">"black"</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span></span>
<span>    option <span class="op">=</span> <span class="st">"rocket"</span>,</span>
<span>    labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent_format</a></span><span class="op">(</span>accuracy <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>    name <span class="op">=</span> <span class="st">"% Invaders with\nPositive Fitness"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">cnt_pos</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"white"</span>, size <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">coord_sf</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Site-level Invasibility (Proportion of Successful Invaders)"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Longitude"</span>, y <span class="op">=</span> <span class="st">"Latitude"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-site-invasibility-1.png" width="100%" style="display: block; margin: auto;"></p>
<hr>
</div>
<div class="section level3">
<h3 id="id_124-mean-invasion-fitness-by-site">12.4 Mean Invasion Fitness by Site<a class="anchor" aria-label="anchor" href="#id_124-mean-invasion-fitness-by-site"></a>
</h3>
<p>We map the <strong>mean invasion fitness</strong> across invaders at each site â€” the average <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> over all <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> for a given site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>. This provides a continuous index of <strong>community openness</strong>, complementing the proportion-based <a href="#def-Vs"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mi>s</mi></msub><annotation encoding="application/x-tex">V_s</annotation></semantics></math></a>:</p>
<ul>
<li>
<strong>High mean values</strong> â€” communities generally favourable to a broad range of potential invaders.</li>
<li>
<strong>Low mean values</strong> â€” communities broadly resistant due to environmental mismatch, strong resident competition, or both.</li>
</ul>
<p>Sites with moderately high mean values but low <a href="#def-Vs"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mi>s</mi></msub><annotation encoding="application/x-tex">V_s</annotation></semantics></math></a> may be vulnerable to a few highly adapted invaders â€” a pattern worth targeted monitoring.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Map the mean invasion fitness of all invaders at each site</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">site_sum</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, fill <span class="op">=</span> <span class="va">mean_l</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span>color <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">rsa</span>, inherit.aes <span class="op">=</span> <span class="cn">FALSE</span>, fill <span class="op">=</span> <span class="cn">NA</span>, color <span class="op">=</span> <span class="st">"black"</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"turbo"</span>, name <span class="op">=</span> <span class="st">"Mean invasion\nfitness"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># Optional: annotate with number of successful invaders</span></span>
<span>  <span class="co"># geom_text(aes(label = cnt_pos), color = "white", size = 1.5) +</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">coord_sf</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Mean Species Invasiveness by Site (All Invaders)"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Longitude"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Latitude"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-spp-invasiness-1.png" width="100%" style="display: block; margin: auto;"></p>
<hr>
</div>
<div class="section level3">
<h3 id="id_125-clustering-and-risk-scenarios">12.5 Clustering and Risk Scenarios<a class="anchor" aria-label="anchor" href="#id_125-clustering-and-risk-scenarios"></a>
</h3>
<p>We identify <strong>clusters of sites and invaders</strong> with similar invasion fitness patterns using hierarchical clustering of <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> from <code>Î»_mat</code>. This reveals ecological â€œsyndromesâ€ â€” types of sites most at risk, or types of invaders most threatening â€” enabling targeted surveillance and intervention.</p>
<div class="section level4">
<h4 id="id_1251-hierarchical-clustering-of-invaders-and-sites">12.5.1 Hierarchical Clustering of Invaders and Sites<a class="anchor" aria-label="anchor" href="#id_1251-hierarchical-clustering-of-invaders-and-sites"></a>
</h4>
<p>Hierarchical clustering groups:</p>
<ul>
<li>
<strong>Rows</strong> (invaders) by similarity of their <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math> profiles across sites.</li>
<li>
<strong>Columns</strong> (sites) by similarity of their <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math> profiles across invaders.</li>
</ul>
<p>The heatmap displays <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> values as cells, with dendrograms indicating clusters. This approach distinguishes:</p>
<ul>
<li>
<strong>Invader types</strong> â€” e.g., broad-spectrum high-risk vs.Â site-specialists.</li>
<li>
<strong>Site types</strong> â€” e.g., broadly open vs.Â strongly resistant communities.</li>
</ul>
<p>These groups can be used for categorical risk classification and scenario exploration in downstream analyses.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Remove invaders (rows) and sites (columns) with all NA values (if present)</span></span>
<span><span class="co"># This ensures only meaningful data are visualized and clustered</span></span>
<span><span class="va">lambda_mat_noNA</span> <span class="op">&lt;-</span> <span class="va">lambda_mat</span><span class="op">[</span></span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">lambda_mat</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">lambda_mat</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/rowSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">lambda_mat</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">lambda_mat</span><span class="op">)</span></span>
<span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Clustered heatmap of invasion fitness matrix (invader Ã— site)</span></span>
<span><span class="co"># - Each cell shows the invasion fitness (lambda) for a given invader-site pair</span></span>
<span><span class="co"># - Hierarchical clustering groups similar invaders and similar sites</span></span>
<span><span class="fu">pheatmap</span><span class="op">(</span><span class="va">lambda_mat_noNA</span>,</span>
<span>  color <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/flip.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">100</span>, option <span class="op">=</span> <span class="st">"viridis"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  clustering_distance_rows <span class="op">=</span> <span class="st">"euclidean"</span>,</span>
<span>  clustering_distance_cols <span class="op">=</span> <span class="st">"euclidean"</span>,</span>
<span>  clustering_method <span class="op">=</span> <span class="st">"complete"</span>,</span>
<span>  fontsize_row <span class="op">=</span> <span class="fl">8</span>, fontsize_col <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Clustered Invasion Fitness Matrix (Invader Ã— Site)"</span>,</span>
<span>  angle_col <span class="op">=</span> <span class="fl">45</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-fitness-cluster-1.png" width="100%" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Assign invaders and sites to discrete clusters (syndromes) for downstream visualization</span></span>
<span><span class="co"># Sites: cluster by invasion fitness profile (columns)</span></span>
<span><span class="co"># Donâ€™t force equal-width bands (let the data pick k) --&gt;</span></span>
<span><span class="va">site_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/transpose.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># standardize first then compute pairwise distances between sites</span></span>
<span><span class="va">site_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="va">site_dist</span>, method <span class="op">=</span> <span class="st">"ward.D2"</span><span class="op">)</span> <span class="co"># Hierarchical clustering</span></span>
<span><span class="va">kj</span> <span class="op">&lt;-</span> <span class="fu">factoextra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/factoextra/man/fviz_nbclust.html" class="external-link">fviz_nbclust</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/transpose.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span><span class="op">)</span>, <span class="va">kmeans</span>, method <span class="op">=</span> <span class="st">"silhouette"</span><span class="op">)</span> <span class="co"># or gap</span></span>
<span><span class="va">site_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html" class="external-link">cutree</a></span><span class="op">(</span><span class="va">site_clust</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="co"># Assign each site to one of 5 clusters</span></span>
<span></span>
<span><span class="co"># Invaders: cluster by fitness profile across sites (rows)</span></span>
<span><span class="va">invader_dist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span><span class="op">)</span> <span class="co"># Scale then compute pairwise distances between invaders</span></span>
<span><span class="va">invader_clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="va">invader_dist</span>, method <span class="op">=</span> <span class="st">"ward.D2"</span><span class="op">)</span></span>
<span><span class="va">ki</span> <span class="op">&lt;-</span> <span class="fu">factoextra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/factoextra/man/fviz_nbclust.html" class="external-link">fviz_nbclust</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/transpose.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span><span class="op">)</span>, <span class="va">kmeans</span>, method <span class="op">=</span> <span class="st">"silhouette"</span><span class="op">)</span> <span class="co"># or gap</span></span>
<span><span class="va">invader_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html" class="external-link">cutree</a></span><span class="op">(</span><span class="va">invader_clust</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="co"># Assign each invader to one of 5 clusters</span></span>
<span></span>
<span><span class="co"># Attach cluster labels to site and invader summary dataframes</span></span>
<span><span class="co"># Add site cluster group to site summary dataframe (site_sum)</span></span>
<span><span class="co"># site_id must match the column names of lambda_mat_noNA</span></span>
<span><span class="va">site_sum</span><span class="op">$</span><span class="va">site_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">site_groups</span><span class="op">[</span><span class="va">site_sum</span><span class="op">$</span><span class="va">site_id</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add invader cluster group to long-form fitness dataframe (lambda_df)</span></span>
<span><span class="co"># invader must match the row names of lambda_mat_noNA</span></span>
<span><span class="va">lambda_df</span><span class="op">$</span><span class="va">invader_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">invader_groups</span><span class="op">[</span><span class="va">lambda_df</span><span class="op">$</span><span class="va">invader</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># These cluster assignments can be used for:</span></span>
<span><span class="co"># - Color-coding sites and invaders in subsequent plots</span></span>
<span><span class="co"># - Group-based analyses or statistical comparisons</span></span>
<span><span class="co"># - Mapping spatial patterns of invasion risk by ecological â€œsyndromeâ€</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="id_1252-mapping-site-level-risk-categories">12.5.2 Mapping Site-level Risk Categories<a class="anchor" aria-label="anchor" href="#id_1252-mapping-site-level-risk-categories"></a>
</h4>
<p>Cluster assignments are converted into <strong>descriptive invasibility categories</strong> (e.g., <em>very-high</em>, <em>high</em>, <em>medium</em>, <em>low</em>, <em>very-low</em>) based on mean <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> within each cluster.</p>
<ul>
<li>
<strong>High-risk site categories</strong> â€” clusters with generally high <a href="#def-Vs"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mi>s</mi></msub><annotation encoding="application/x-tex">V_s</annotation></semantics></math></a> and high mean <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math>.</li>
<li>
<strong>Low-risk site categories</strong> â€” clusters with low <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mi>s</mi></msub><annotation encoding="application/x-tex">V_s</annotation></semantics></math> and consistently negative or low <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math>.</li>
</ul>
<p>Mapping these categories provides a spatial overview of invasibility syndromes, highlighting geographic concentrations of risk. Parallel invader-category assignments allow alignment of high-risk invaders with high-risk sites for targeted management scenarios.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># -------------------------------</span></span>
<span><span class="co"># 1) INVADER CATEGORIES (unchanged)</span></span>
<span><span class="co"># -------------------------------</span></span>
<span><span class="va">invader_summary</span> <span class="op">&lt;-</span> <span class="va">lambda_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">invader</span>, <span class="va">invader_cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>mean_lambda <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">lambda</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, .groups <span class="op">=</span> <span class="st">"drop"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cluster_means</span> <span class="op">&lt;-</span> <span class="va">invader_summary</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">invader_cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>cluster_mean <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">mean_lambda</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">cluster_mean</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">category_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"very-high"</span>, <span class="st">"high"</span>, <span class="st">"medium"</span>, <span class="st">"low"</span>, <span class="st">"very-low"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">cluster_means</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">cluster_map</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">category_labels</span>, <span class="va">cluster_means</span><span class="op">$</span><span class="va">invader_cluster</span><span class="op">)</span></span>
<span></span>
<span><span class="va">invader_summary</span> <span class="op">&lt;-</span> <span class="va">invader_summary</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>invader_category <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cluster_map</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">invader_cluster</span><span class="op">)</span><span class="op">]</span>,</span>
<span>    levels <span class="op">=</span> <span class="va">category_labels</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lambda_df</span> <span class="op">&lt;-</span> <span class="va">lambda_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>invader_category <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cluster_map</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">invader_cluster</span><span class="op">)</span><span class="op">]</span>,</span>
<span>    levels <span class="op">=</span> <span class="va">category_labels</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># -------------------------------------------------------------</span></span>
<span><span class="co"># 2) SITE GROUPING: pick *one* method to reduce "stripey" maps</span></span>
<span><span class="co">#    Choices:</span></span>
<span><span class="co">#      "quantile"   â€“ gradient bins of mean fitness (recommended default)</span></span>
<span><span class="co">#      "pca_kmeans" â€“ PCA (on Î» profiles) + k-means (k via silhouette)</span></span>
<span><span class="co">#      "clustgeo"   â€“ spatially-constrained Ward clustering (needs ClustGeo)</span></span>
<span><span class="co">#      "hc_sil"     â€“ Ward hierarchical on standardized Î» with k via silhouette</span></span>
<span><span class="co"># -------------------------------------------------------------</span></span>
<span><span class="va">site_grouping_method</span> <span class="op">&lt;-</span> <span class="st">"quantile"</span> <span class="co"># = change to "pca_kmeans", "clustgeo", or "hc_sil"</span></span>
<span></span>
<span><span class="co"># Build Î» matrix with aligned columns = site_id, rows = invaders</span></span>
<span><span class="co"># (Assumes you have lambda_mat already; otherwise widen lambda_df)</span></span>
<span><span class="va">lambda_mat_noNA</span> <span class="op">&lt;-</span> <span class="va">lambda_mat</span><span class="op">[</span></span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">lambda_mat</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">lambda_mat</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/rowSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">lambda_mat</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">lambda_mat</span><span class="op">)</span></span>
<span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Utility: safe match coords for columns (sites)</span></span>
<span><span class="va">site_coords</span> <span class="op">&lt;-</span> <span class="va">site_sum</span><span class="op">[</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span>, <span class="va">site_sum</span><span class="op">$</span><span class="va">site_id</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># -------------------</span></span>
<span><span class="co"># A) Quantile method</span></span>
<span><span class="co"># -------------------</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">site_grouping_method</span> <span class="op">==</span> <span class="st">"quantile"</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">brks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">site_sum</span><span class="op">$</span><span class="va">mean_l</span>, probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.2</span><span class="op">)</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="va">labs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"very-low"</span>, <span class="st">"low"</span>, <span class="st">"medium"</span>, <span class="st">"high"</span>, <span class="st">"very-high"</span><span class="op">)</span></span>
<span>  <span class="va">site_sum</span> <span class="op">&lt;-</span> <span class="va">site_sum</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>site_category <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">mean_l</span>,</span>
<span>      breaks <span class="op">=</span> <span class="va">brks</span>,</span>
<span>      include.lowest <span class="op">=</span> <span class="cn">TRUE</span>, labels <span class="op">=</span> <span class="va">labs</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># ------------------------------</span></span>
<span><span class="co"># B) PCA + k-means (silhouette)</span></span>
<span><span class="co"># ------------------------------</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">site_grouping_method</span> <span class="op">==</span> <span class="st">"pca_kmeans"</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># standardize Î» columns (sites) across invaders before PCA</span></span>
<span>  <span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/transpose.html" class="external-link">t</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span><span class="op">)</span> <span class="co"># sites Ã— invaders</span></span>
<span>  <span class="va">pcs</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="va">X</span>, center <span class="op">=</span> <span class="cn">TRUE</span>, scale. <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span>  <span class="co"># choose k via silhouette (2..8); fall back to k=4 if package missing</span></span>
<span>  <span class="va">choose_k</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Z</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"factoextra"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="va">sil</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">8</span>, <span class="kw">function</span><span class="op">(</span><span class="va">k</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="va">Z</span>, centers <span class="op">=</span> <span class="va">k</span>, nstart <span class="op">=</span> <span class="fl">50</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span>        <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu">cluster</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cluster/man/silhouette.html" class="external-link">silhouette</a></span><span class="op">(</span><span class="va">cl</span>, <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="va">Z</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span></span>
<span>      <span class="op">}</span><span class="op">)</span></span>
<span>      <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">sil</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>      <span class="fl">4</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">k</span> <span class="op">&lt;-</span> <span class="fu">choose_k</span><span class="op">(</span><span class="va">pcs</span><span class="op">)</span></span>
<span>  <span class="va">site_groups</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="va">pcs</span>, centers <span class="op">=</span> <span class="va">k</span>, nstart <span class="op">=</span> <span class="fl">50</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span></span>
<span></span>
<span>  <span class="co"># Order clusters by *mean site fitness* so labels match risk gradient</span></span>
<span>  <span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    site_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span>,</span>
<span>    cluster <span class="op">=</span> <span class="va">site_groups</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">site_sum</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"site_id"</span>, <span class="st">"mean_l"</span><span class="op">)</span><span class="op">]</span>, by <span class="op">=</span> <span class="st">"site_id"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>mu <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">mean_l</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, .groups <span class="op">=</span> <span class="st">"drop"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">mu</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">site_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"very-high"</span>, <span class="st">"high"</span>, <span class="st">"medium"</span>, <span class="st">"low"</span>, <span class="st">"very-low"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">remap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">site_labels</span>, <span class="va">tmp</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">site_sum</span> <span class="op">&lt;-</span> <span class="va">site_sum</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>      site_cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">remap</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">site_groups</span><span class="op">[</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">site_id</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">]</span>,</span>
<span>        levels <span class="op">=</span> <span class="va">site_labels</span></span>
<span>      <span class="op">)</span>,</span>
<span>      site_category <span class="op">=</span> <span class="va">site_cluster</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># --------------------------------------------</span></span>
<span><span class="co"># C) Spatially constrained clustering (ClustGeo)</span></span>
<span><span class="co"># --------------------------------------------</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">site_grouping_method</span> <span class="op">==</span> <span class="st">"clustgeo"</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"ClustGeo"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"ClustGeo not installed. Install it or choose another method."</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/transpose.html" class="external-link">t</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span><span class="op">)</span> <span class="co"># sites Ã— invaders</span></span>
<span>  <span class="va">D1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="co"># Î»-profile distance</span></span>
<span>  <span class="va">D0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/coerce.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">site_coords</span><span class="op">)</span><span class="op">)</span> <span class="co"># geographic distance</span></span>
<span></span>
<span>  <span class="co"># alpha controls trade-off (0 = spatial only, 1 = Î» only)</span></span>
<span>  <span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.3</span></span>
<span>  <span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">ClustGeo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ClustGeo/man/hclustgeo.html" class="external-link">hclustgeo</a></span><span class="op">(</span><span class="va">D0</span>, <span class="va">D1</span>, alpha <span class="op">=</span> <span class="va">alpha</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># choose k via silhouette on D1 for guidance; fallback k=5</span></span>
<span>  <span class="va">choose_k_hc</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">tree</span>, <span class="va">D</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"cluster"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="va">ks</span> <span class="op">&lt;-</span> <span class="fl">2</span><span class="op">:</span><span class="fl">8</span></span>
<span>      <span class="va">sil</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ks</span>, <span class="kw">function</span><span class="op">(</span><span class="va">k</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html" class="external-link">cutree</a></span><span class="op">(</span><span class="va">tree</span>, k <span class="op">=</span> <span class="va">k</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu">cluster</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cluster/man/silhouette.html" class="external-link">silhouette</a></span><span class="op">(</span><span class="va">cl</span>, <span class="va">D</span><span class="op">)</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span></span>
<span>      <span class="op">}</span><span class="op">)</span></span>
<span>      <span class="va">ks</span><span class="op">[</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">sil</span><span class="op">)</span><span class="op">]</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>      <span class="fl">5</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">k</span> <span class="op">&lt;-</span> <span class="fu">choose_k_hc</span><span class="op">(</span><span class="va">tree</span>, <span class="va">D1</span><span class="op">)</span></span>
<span>  <span class="va">site_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html" class="external-link">cutree</a></span><span class="op">(</span><span class="va">tree</span>, k <span class="op">=</span> <span class="va">k</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    site_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span>,</span>
<span>    cluster <span class="op">=</span> <span class="va">site_groups</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">site_sum</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"site_id"</span>, <span class="st">"mean_l"</span><span class="op">)</span><span class="op">]</span>, by <span class="op">=</span> <span class="st">"site_id"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>mu <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">mean_l</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, .groups <span class="op">=</span> <span class="st">"drop"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">mu</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">site_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"very-high"</span>, <span class="st">"high"</span>, <span class="st">"medium"</span>, <span class="st">"low"</span>, <span class="st">"very-low"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">remap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">site_labels</span>, <span class="va">tmp</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">site_sum</span> <span class="op">&lt;-</span> <span class="va">site_sum</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>      site_cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">remap</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">site_groups</span><span class="op">[</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">site_id</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">]</span>,</span>
<span>        levels <span class="op">=</span> <span class="va">site_labels</span></span>
<span>      <span class="op">)</span>,</span>
<span>      site_category <span class="op">=</span> <span class="va">site_cluster</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># -------------------------------------------</span></span>
<span><span class="co"># D) Ward HC with k via silhouette (no PCA)</span></span>
<span><span class="co"># -------------------------------------------</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">site_grouping_method</span> <span class="op">==</span> <span class="st">"hc_sil"</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/transpose.html" class="external-link">t</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">hc</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"ward.D2"</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">choose_k_hc</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">hc</span>, <span class="va">X</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"cluster"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="va">ks</span> <span class="op">&lt;-</span> <span class="fl">2</span><span class="op">:</span><span class="fl">8</span></span>
<span>      <span class="va">sil</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ks</span>, <span class="kw">function</span><span class="op">(</span><span class="va">k</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html" class="external-link">cutree</a></span><span class="op">(</span><span class="va">hc</span>, k <span class="op">=</span> <span class="va">k</span><span class="op">)</span></span>
<span>        <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu">cluster</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/cluster/man/silhouette.html" class="external-link">silhouette</a></span><span class="op">(</span><span class="va">cl</span>, <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span></span>
<span>      <span class="op">}</span><span class="op">)</span></span>
<span>      <span class="va">ks</span><span class="op">[</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">sil</span><span class="op">)</span><span class="op">]</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>      <span class="fl">4</span></span>
<span>    <span class="op">}</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">k</span> <span class="op">&lt;-</span> <span class="fu">choose_k_hc</span><span class="op">(</span><span class="va">hc</span>, <span class="va">X</span><span class="op">)</span></span>
<span>  <span class="va">site_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html" class="external-link">cutree</a></span><span class="op">(</span><span class="va">hc</span>, k <span class="op">=</span> <span class="va">k</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    site_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span>,</span>
<span>    cluster <span class="op">=</span> <span class="va">site_groups</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">site_sum</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"site_id"</span>, <span class="st">"mean_l"</span><span class="op">)</span><span class="op">]</span>, by <span class="op">=</span> <span class="st">"site_id"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>mu <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/summarize-generics.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">mean_l</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, .groups <span class="op">=</span> <span class="st">"drop"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">mu</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">site_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"very-high"</span>, <span class="st">"high"</span>, <span class="st">"medium"</span>, <span class="st">"low"</span>, <span class="st">"very-low"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">remap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">site_labels</span>, <span class="va">tmp</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">site_sum</span> <span class="op">&lt;-</span> <span class="va">site_sum</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>      site_cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">remap</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">site_groups</span><span class="op">[</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">site_id</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">lambda_mat_noNA</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">]</span>,</span>
<span>        levels <span class="op">=</span> <span class="va">site_labels</span></span>
<span>      <span class="op">)</span>,</span>
<span>      site_category <span class="op">=</span> <span class="va">site_cluster</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># ---------------------------------------</span></span>
<span><span class="co"># 3) Sanity checks (optional diagnostics)</span></span>
<span><span class="co"># ---------------------------------------</span></span>
<span><span class="co"># Are columns pre-sorted by longitude?</span></span>
<span><span class="co"># plot(match(colnames(lambda_mat_noNA), site_sum$site_id),</span></span>
<span><span class="co">#      order(site_sum$x)[match(colnames(lambda_mat_noNA), site_sum$site_id)],</span></span>
<span><span class="co">#      main = "Col order vs. longitude order")</span></span>
<span></span>
<span><span class="co"># Shuffle columns to confirm clustering isnâ€™t an artifact</span></span>
<span><span class="co"># set.seed(1); shuf = sample(ncol(lambda_mat_noNA))</span></span>
<span><span class="co"># hc_shuf = hclust(dist(t(lambda_mat_noNA[, shuf])))</span></span>
<span></span>
<span><span class="co"># -----------------------------</span></span>
<span><span class="co"># 4) Map the site categories</span></span>
<span><span class="co"># -----------------------------</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">site_sum</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, fill <span class="op">=</span> <span class="va">site_category</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span>color <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">rsa</span>, inherit.aes <span class="op">=</span> <span class="cn">FALSE</span>, fill <span class="op">=</span> <span class="cn">NA</span>, color <span class="op">=</span> <span class="st">"black"</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"RdYlBu"</span>, direction <span class="op">=</span> <span class="fl">1</span>, name <span class="op">=</span> <span class="st">"Site invasibility"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">coord_sf</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>      <span class="st">"Spatial Invasion Risk ("</span>,</span>
<span>      <span class="kw"><a href="https://rdrr.io/r/base/switch.html" class="external-link">switch</a></span><span class="op">(</span><span class="va">site_grouping_method</span>,</span>
<span>        quantile   <span class="op">=</span> <span class="st">"Quantile Bins of Mean \u03BB"</span>,</span>
<span>        pca_kmeans <span class="op">=</span> <span class="st">"PCA + k-means"</span>,</span>
<span>        clustgeo   <span class="op">=</span> <span class="st">"Spatially Constrained ClustGeo"</span>,</span>
<span>        hc_sil     <span class="op">=</span> <span class="st">"Ward HC + Silhouette"</span></span>
<span>      <span class="op">)</span>,</span>
<span>      <span class="st">")"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Longitude"</span>, y <span class="op">=</span> <span class="st">"Latitude"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-map-risk-1.png" width="100%" style="display: block; margin: auto;"></p>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="id_13-synthesising-invasion-fitness-insights">13. Synthesising Invasion Fitness Insights<a class="anchor" aria-label="anchor" href="#id_13-synthesising-invasion-fitness-insights"></a>
</h2>
<p>This section distills the spatial, clustering, and trait-based analyses into targeted summaries of <strong>invasion risk patterns</strong>. All outputs are derived from the invasion fitness matrix <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> computed in <a href="#def-compute-invasion-fitness"><code>compute_invasion_fitness()</code></a>, where <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> indexes invaders and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math> indexes sites.</p>
<div class="section level3">
<h3 id="id_131-distribution-of-invasion-fitness-values">13.1 Distribution of Invasion Fitness Values<a class="anchor" aria-label="anchor" href="#id_131-distribution-of-invasion-fitness-values"></a>
</h3>
<p>A histogram of all <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> values provides a system-wide view of <strong>community openness</strong> to invasion:</p>
<ul>
<li>
<strong>Right-skewed distributions</strong> â†’ prevalence of favourable establishment conditions.</li>
<li>
<strong>Left-skewed distributions</strong> â†’ dominance of exclusionary environments.</li>
</ul>
<p>This complements the site- and invader-specific patterns in Section 12.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the distribution of invasion fitness values (lambda) across all invader-site combinations,</span></span>
<span><span class="co"># using a histogram colored by lambda value for visual emphasis on fitness extremes</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">lambda_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">lambda</span>, fill <span class="op">=</span> <span class="va">..x..</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">40</span>, color <span class="op">=</span> <span class="st">"grey30"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"magma"</span>, guide <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Distribution of Invasion Fitness Values (All Invader Ã— Site)"</span>,</span>
<span>    x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="st">"Invasion Fitness"</span> <span class="op">~</span> <span class="va">lambda</span><span class="op">[</span><span class="va">i</span> <span class="op">*</span> <span class="va">k</span><span class="op">]</span><span class="op">)</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Frequency"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-fitness-dist-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="id_132-top-and-bottom-invaders-and-sites">13.2 Top and Bottom Invaders and Sites<a class="anchor" aria-label="anchor" href="#id_132-top-and-bottom-invaders-and-sites"></a>
</h3>
<p>For rapid prioritisation, we rank invaders and sites by their mean <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math>:</p>
<ul>
<li>
<strong>Top-ranked invaders</strong> â€” consistently high predicted establishment success across sites.</li>
<li>
<strong>Top-ranked sites</strong> â€” consistently permissive to a wide range of invaders.</li>
<li>
<strong>Bottom-ranked invaders/sites</strong> â€” consistently low <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math>, indicating strong biotic or abiotic resistance.</li>
</ul>
<p>These rankings are direct, interpretable summaries of the invasion fitness landscape.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a><span class="co">#&gt; ==== Top 3 Invaders by Mean Invasion Fitness ====</span></span>
<span id="cb46-2"><a href="#cb46-2" tabindex="-1"></a><span class="co">#&gt; 1. inv3: 5.703</span></span>
<span id="cb46-3"><a href="#cb46-3" tabindex="-1"></a><span class="co">#&gt; 2. inv4: 5.611</span></span>
<span id="cb46-4"><a href="#cb46-4" tabindex="-1"></a><span class="co">#&gt; 3. inv9: 5.14</span></span>
<span id="cb46-5"><a href="#cb46-5" tabindex="-1"></a><span class="co">#&gt; ==== Bottom 3 Invaders by Mean Invasion Fitness ====</span></span>
<span id="cb46-6"><a href="#cb46-6" tabindex="-1"></a><span class="co">#&gt; 1. inv2: 2.816</span></span>
<span id="cb46-7"><a href="#cb46-7" tabindex="-1"></a><span class="co">#&gt; 2. inv10: 2.3</span></span>
<span id="cb46-8"><a href="#cb46-8" tabindex="-1"></a><span class="co">#&gt; 3. inv7: 2.289</span></span>
<span id="cb46-9"><a href="#cb46-9" tabindex="-1"></a><span class="co">#&gt; ==== Top 3 Sites by Mean Invasion Fitness ====</span></span>
<span id="cb46-10"><a href="#cb46-10" tabindex="-1"></a><span class="co">#&gt; 1. 558: 8.559</span></span>
<span id="cb46-11"><a href="#cb46-11" tabindex="-1"></a><span class="co">#&gt; 2. 823: 8.338</span></span>
<span id="cb46-12"><a href="#cb46-12" tabindex="-1"></a><span class="co">#&gt; 3. 1026: 8.173</span></span>
<span id="cb46-13"><a href="#cb46-13" tabindex="-1"></a><span class="co">#&gt; ==== Bottom 3 Sites by Mean Invasion Fitness ====</span></span>
<span id="cb46-14"><a href="#cb46-14" tabindex="-1"></a><span class="co">#&gt; 1. 729: 1.046</span></span>
<span id="cb46-15"><a href="#cb46-15" tabindex="-1"></a><span class="co">#&gt; 2. 614: 0.952</span></span>
<span id="cb46-16"><a href="#cb46-16" tabindex="-1"></a><span class="co">#&gt; 3. 802: 0.927</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="id_133-functional-correlates-of-invasion-success">13.3 Functional Correlates of Invasion Success<a class="anchor" aria-label="anchor" href="#id_133-functional-correlates-of-invasion-success"></a>
</h3>
<p>To explore traitâ€“fitness relationships, we join <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> summaries with invader trait data:</p>
<ul>
<li>
<strong>Continuous traits</strong> â€” correlation coefficients with mean <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math> identify traits most positively or negatively associated with invasion success.</li>
<li>
<strong>Categorical traits</strong> â€” mean <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math> per trait level identifies functional syndromes linked to high or low establishment.</li>
</ul>
<p>This supports hypothesis generation about <strong>mechanistic drivers</strong> of invasion potential, complementing the purely spatial and compositional analyses of Section 12.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="co">#&gt; ==== Top 3 Continuous Traits by Correlation with Mean Invasion Fitness ====</span></span>
<span id="cb47-2"><a href="#cb47-2" tabindex="-1"></a><span class="co">#&gt; 1. trait_cont10: 0.403</span></span>
<span id="cb47-3"><a href="#cb47-3" tabindex="-1"></a><span class="co">#&gt; 2. trait_cont6: 0.321</span></span>
<span id="cb47-4"><a href="#cb47-4" tabindex="-1"></a><span class="co">#&gt; 3. trait_cont5: 0.302</span></span>
<span id="cb47-5"><a href="#cb47-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb47-6"><a href="#cb47-6" tabindex="-1"></a><span class="co">#&gt; ==== Bottom 3 Continuous Traits by Correlation with Mean Invasion Fitness ====</span></span>
<span id="cb47-7"><a href="#cb47-7" tabindex="-1"></a><span class="co">#&gt; 1. trait_cont2: -0.278</span></span>
<span id="cb47-8"><a href="#cb47-8" tabindex="-1"></a><span class="co">#&gt; 2. trait_cont9: -0.459</span></span>
<span id="cb47-9"><a href="#cb47-9" tabindex="-1"></a><span class="co">#&gt; 3. trait_cont1: -0.505</span></span>
<span id="cb47-10"><a href="#cb47-10" tabindex="-1"></a><span class="co">#&gt; ==== Categorical Traits: Top Value per Trait by Mean Invasion Fitness ====</span></span>
<span id="cb47-11"><a href="#cb47-11" tabindex="-1"></a><span class="co">#&gt; trait_cat11: grassland (4.53)</span></span>
<span id="cb47-12"><a href="#cb47-12" tabindex="-1"></a><span class="co">#&gt; trait_cat12: nocturnal (4.01)</span></span>
<span id="cb47-13"><a href="#cb47-13" tabindex="-1"></a><span class="co">#&gt; trait_cat13: bivoltine (4.25)</span></span>
<span id="cb47-14"><a href="#cb47-14" tabindex="-1"></a><span class="co">#&gt; trait_cat14: nectarivore (4.56)</span></span>
<span id="cb47-15"><a href="#cb47-15" tabindex="-1"></a><span class="co">#&gt; trait_cat15: migratory (4.36)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="id_134-faceted-maps-for-key-invaders">13.4 Faceted Maps for Key Invaders<a class="anchor" aria-label="anchor" href="#id_134-faceted-maps-for-key-invaders"></a>
</h3>
<p>We map <a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a> for the <strong>top 3</strong> and <strong>bottom 3</strong> invaders from Section 13.2:</p>
<ul>
<li>
<strong>Top invaders</strong> â€” spatially explicit â€œhotspotsâ€ of establishment potential.</li>
<li>
<strong>Bottom invaders</strong> â€” regions where their traits or niches are mismatched to the community or environment.</li>
</ul>
<p>These maps link functional traits, site conditions, and predicted invasion risk in an interpretable, management-relevant format.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify the top 3 and bottom 3 invaders by mean fitness</span></span>
<span><span class="va">key_invaders</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">top3_inv</span><span class="op">$</span><span class="va">invader</span>, <span class="va">bottom3_inv</span><span class="op">$</span><span class="va">invader</span><span class="op">)</span> <span class="co"># select 3 best/worst</span></span>
<span></span>
<span><span class="co"># Filter df amd ensure facet order matches ranking (top first, then bottom)</span></span>
<span><span class="va">lambda_key</span> <span class="op">&lt;-</span> <span class="va">lambda_df</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">invader</span> <span class="op"><a href="https://rspatial.github.io/terra/reference/match.html" class="external-link">%in%</a></span> <span class="va">key_invaders</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>invader <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">invader</span>, levels <span class="op">=</span> <span class="va">key_invaders</span><span class="op">)</span><span class="op">)</span> <span class="co"># enforce desired order</span></span>
<span></span>
<span><span class="co"># Faceted map</span></span>
<span><span class="co"># Spatial invasion fitness for each of the top/bottom invaders, enabling direct visual comparison</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">lambda_key</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, fill <span class="op">=</span> <span class="va">lambda</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsf.html" class="external-link">geom_sf</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">rsa</span>, inherit.aes <span class="op">=</span> <span class="cn">FALSE</span>, fill <span class="op">=</span> <span class="cn">NA</span>, color <span class="op">=</span> <span class="st">"black"</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">invader</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"viridis"</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Spatial Summaries of Invasion Fitness\nfor Top-3 and Bottom-3 Invaders"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Longitude"</span>, y <span class="op">=</span> <span class="st">"Latitude"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span>base_size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-key-invaders-1.png" width="100%" style="display: block; margin: auto;"></p>
<blockquote>
<p><strong>Note:</strong> * Sections 12â€“13 together move from <strong>raw <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></strong> values â†’ <strong>clustered patterns</strong> â†’ <strong>functional correlates</strong> â†’ <strong>site- and invader-specific rankings</strong>. * This progression enables scaling from <strong>broad ecological syndromes</strong> to <strong>actionable risk profiles</strong> for specific species and sites. Â </p>
</blockquote>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="id_14-glossary-table-symbols-r-objects-meaning-units-source">14. Glossary table (symbols, R objects, meaning, units, source)<a class="anchor" aria-label="anchor" href="#id_14-glossary-table-symbols-r-objects-meaning-units-source"></a>
</h2>
<table class="table">
<colgroup>
<col width="11%">
<col width="19%">
<col width="30%">
<col width="19%">
<col width="18%">
</colgroup>
<thead><tr class="header">
<th>Symbol &amp; Link</th>
<th>Equation</th>
<th>Definition</th>
<th>R Object / Slot(s)</th>
<th>Function(s) Producing It</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="#def-dij"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>d</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">d_{ij}</annotation></semantics></math></a></td>
<td>Pairwise trait dissimilarity (e.g.Â Gower)</td>
<td>Functional trait distance between species <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> and <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>; input to Gaussian trait kernel <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>K</mi><mi>t</mi></msub><annotation encoding="application/x-tex">K_t</annotation></semantics></math>
</td>
<td>
<code>comp$d_ij[i, j]</code> or <code>cis$g_all[i, j]</code> if <code>g_all</code> computed as distance</td>
<td>
<code><a href="reference/compute_interaction_strength.html">compute_interaction_strength()</a></code> / <code><a href="reference/compute_competition_kernel.html">compute_competition_kernel()</a></code>
</td>
</tr>
<tr class="even">
<td><a href="#def-gall"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>g</mi><mrow><mi>i</mi><mi>j</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">a</mi><mi mathvariant="normal">l</mi><mi mathvariant="normal">l</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><annotation encoding="application/x-tex">g^{\mathrm{(all)}}_{ij}</annotation></semantics></math></a></td>
<td>Generalised trait distance/similarity depending on kernel choice</td>
<td>Distance/similarity for all species pairs, used in interaction strength; for competition, must be a distance</td>
<td><code>cis$g_all[i, j]</code></td>
<td><code><a href="reference/compute_interaction_strength.html">compute_interaction_strength()</a></code></td>
</tr>
<tr class="odd">
<td><a href="#def-Nstar"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>N</mi><mrow><mi>j</mi><mi>s</mi></mrow><mo>*</mo></msubsup><annotation encoding="application/x-tex">N^{*}_{js}</annotation></semantics></math></a></td>
<td>Equilibrium abundance of resident species <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math> at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>
</td>
<td>Resident abundance context for interaction strength calculations</td>
<td><code>cis$Nstar[j, s]</code></td>
<td><code><a href="reference/compute_interaction_strength.html">compute_interaction_strength()</a></code></td>
</tr>
<tr class="even">
<td><a href="#def-alpha"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\alpha_{ij}</annotation></semantics></math></a></td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><msub><mi>K</mi><mi>t</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>d</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>;</mo><msub><mi>Ïƒ</mi><mi>t</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\alpha_{ij} = K_t(d_{ij}; \sigma_t)</annotation></semantics></math></td>
<td>Competition coefficient between invader <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> and resident <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math> based on trait similarity</td>
<td><code>comp$a_ij[i, j]</code></td>
<td><code><a href="reference/compute_competition_kernel.html">compute_competition_kernel()</a></code></td>
</tr>
<tr class="odd">
<td><a href="#def-sigmat"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Ïƒ</mi><mi>t</mi></msub><annotation encoding="application/x-tex">\sigma_t</annotation></semantics></math></a></td>
<td>Trait kernel bandwidth</td>
<td>Controls width of Gaussian trait kernel <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>K</mi><mi>t</mi></msub><annotation encoding="application/x-tex">K_t</annotation></semantics></math>
</td>
<td><code>comp$sigma_t</code></td>
<td>User-defined or estimated</td>
</tr>
<tr class="even">
<td><a href="#def-theta"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î¸</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\theta_j</annotation></semantics></math></a></td>
<td>Environmental optimum of species <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
</td>
<td>Traitâ€“environment centroid for species <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
</td>
<td><code>ek$env_opt[j, ]</code></td>
<td><code><a href="reference/compute_environment_kernel.html">compute_environment_kernel()</a></code></td>
</tr>
<tr class="odd">
<td><a href="#def-delta"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\Delta_{js}</annotation></semantics></math></a></td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><mo>=</mo><mo stretchy="false" form="postfix">âˆ¥</mo><msub><mi>ğ„</mi><mi>s</mi></msub><mo>âˆ’</mo><msub><mi>Î¸</mi><mi>j</mi></msub><mo stretchy="false" form="postfix">âˆ¥</mo></mrow><annotation encoding="application/x-tex">\Delta_{js} = \| \mathbf{E}_s - \theta_j \|</annotation></semantics></math></td>
<td>Environmental mismatch between site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math> and species <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
</td>
<td><code>ek$env_dist[j, s]</code></td>
<td><code><a href="reference/compute_environment_kernel.html">compute_environment_kernel()</a></code></td>
</tr>
<tr class="even">
<td><a href="#def-Ke"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>K</mi><mi>e</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><mo>;</mo><msub><mi>Ïƒ</mi><mi>e</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K_e(\Delta_{js};\sigma_e)</annotation></semantics></math></a></td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>K</mi><mi>e</mi></msub><mo>=</mo><mo>exp</mo><mspace width="-0.167em"></mspace><mrow><mo stretchy="true" form="prefix">(</mo><mo>âˆ’</mo><mfrac><msubsup><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow><mn>2</mn></msubsup><mrow><mn>2</mn><msubsup><mi>Ïƒ</mi><mi>e</mi><mn>2</mn></msubsup></mrow></mfrac><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">K_e = \exp\!\left( -\frac{\Delta_{js}^2}{2\sigma_e^2} \right)</annotation></semantics></math></td>
<td>Environmental filtering weight</td>
<td><code>ek$K_env[j, s]</code></td>
<td><code><a href="reference/compute_environment_kernel.html">compute_environment_kernel()</a></code></td>
</tr>
<tr class="odd">
<td><a href="#def-sigmae"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Ïƒ</mi><mi>e</mi></msub><annotation encoding="application/x-tex">\sigma_e</annotation></semantics></math></a></td>
<td>Environmental kernel bandwidth</td>
<td>Controls width of Gaussian environment kernel <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>K</mi><mi>e</mi></msub><annotation encoding="application/x-tex">K_e</annotation></semantics></math>
</td>
<td><code>ek$sigma_e</code></td>
<td>User-defined or estimated</td>
</tr>
<tr class="even">
<td><a href="#def-r"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>r</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">r_{is}</annotation></semantics></math></a></td>
<td>Predicted intrinsic growth rate</td>
<td>Growth rate of invader <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math> without biotic/abiotic penalties</td>
<td><code>fitness$r_mat[i, s]</code></td>
<td><code><a href="reference/predict_invader_response.html">predict_invader_response()</a></code></td>
</tr>
<tr class="odd">
<td><a href="#def-I"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mrow><mi>i</mi><mi>j</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">I_{ijs}</annotation></semantics></math></a></td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>I</mi><mrow><mi>i</mi><mi>j</mi><mi>s</mi></mrow></msub><mo>=</mo><msub><mi>Î±</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>â‹…</mo><msub><mi>K</mi><mi>e</mi></msub><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Î”</mi><mrow><mi>j</mi><mi>s</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>â‹…</mo><msubsup><mi>N</mi><mrow><mi>j</mi><mi>s</mi></mrow><mo>*</mo></msubsup></mrow><annotation encoding="application/x-tex">I_{ijs} = \alpha_{ij} \cdot K_e(\Delta_{js}) \cdot N^{*}_{js}</annotation></semantics></math></td>
<td>Impact tensor: per-site, per-resident contribution to invaderâ€™s penalty</td>
<td><code>am$I_raw[i, j, s]</code></td>
<td>
<code><a href="reference/assemble_matrices.html">assemble_matrices()</a></code> / <code><a href="reference/compute_interaction_strength.html">compute_interaction_strength()</a></code>
</td>
</tr>
<tr class="even">
<td><a href="#def-Craw"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><annotation encoding="application/x-tex">C^{\mathrm{(raw)}}_{is}</annotation></semantics></math></a></td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup><mo>=</mo><msub><mo>âˆ‘</mo><mi>j</mi></msub><msub><mi>I</mi><mrow><mi>i</mi><mi>j</mi><mi>s</mi></mrow></msub></mrow><annotation encoding="application/x-tex">C^{\mathrm{(raw)}}_{is} = \sum_{j} I_{ijs}</annotation></semantics></math></td>
<td>Total competitive + environmental penalty experienced by invader <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>
</td>
<td><code>fitness$C_raw[i, s]</code></td>
<td><code><a href="reference/compute_invasion_fitness.html">compute_invasion_fitness()</a></code></td>
</tr>
<tr class="odd">
<td><a href="#def-lambda"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math></a></td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><mo>=</mo><msub><mi>r</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><mo>âˆ’</mo><msubsup><mi>C</mi><mrow><mi>i</mi><mi>s</mi></mrow><mrow><mo stretchy="true" form="prefix" mathvariant="normal">(</mo><mi mathvariant="normal">r</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">w</mi><mo stretchy="true" form="postfix" mathvariant="normal">)</mo></mrow></msubsup></mrow><annotation encoding="application/x-tex">\lambda_{is} = r_{is} - C^{\mathrm{(raw)}}_{is}</annotation></semantics></math></td>
<td>Low-density per-capita growth rate of invader <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math> at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>
</td>
<td>
<code>fitness$lambda[i, s]</code> (also in <code>lambda_mat</code>)</td>
<td><code><a href="reference/compute_invasion_fitness.html">compute_invasion_fitness()</a></code></td>
</tr>
<tr class="even">
<td><a href="#def-Vs"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>V</mi><mi>s</mi></msub><annotation encoding="application/x-tex">V_s</annotation></semantics></math></a></td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>V</mi><mi>s</mi></msub><mo>=</mo><mfrac><mn>1</mn><msub><mi>n</mi><mrow><mi mathvariant="normal">i</mi><mi mathvariant="normal">n</mi><mi mathvariant="normal">v</mi></mrow></msub></mfrac><msub><mo>âˆ‘</mo><mi>i</mi></msub><mn>ğŸ™</mn><mrow><mo stretchy="true" form="prefix">[</mo><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><mo>&gt;</mo><mn>0</mn><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">V_s = \frac{1}{n_\mathrm{inv}} \sum_{i} \mathbb{1}[\lambda_{is} &gt; 0]</annotation></semantics></math></td>
<td>Invasibility: mean or proportion of invaders with positive <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math> at site <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>
</td>
<td><code>summary$Vs[s]</code></td>
<td>Post-processing / summary functions</td>
</tr>
<tr class="odd">
<td><a href="#def-Ii"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>I</mi><mi>i</mi></msub><annotation encoding="application/x-tex">I_i</annotation></semantics></math></a></td>
<td><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>I</mi><mi>i</mi></msub><mo>=</mo><mfrac><mn>1</mn><msub><mi>n</mi><mrow><mi mathvariant="normal">s</mi><mi mathvariant="normal">i</mi><mi mathvariant="normal">t</mi><mi mathvariant="normal">e</mi><mi mathvariant="normal">s</mi></mrow></msub></mfrac><msub><mo>âˆ‘</mo><mi>s</mi></msub><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub></mrow><annotation encoding="application/x-tex">I_i = \frac{1}{n_\mathrm{sites}} \sum_{s} \lambda_{is}</annotation></semantics></math></td>
<td>Invasiveness: mean or sum of <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Î»</mi><mrow><mi>i</mi><mi>s</mi></mrow></msub><annotation encoding="application/x-tex">\lambda_{is}</annotation></semantics></math> over sites for species <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
</td>
<td><code>summary$Ii[i]</code></td>
<td>Post-processing / summary functions</td>
</tr>
</tbody>
</table>
<p><strong>Notes (index orientation youâ€™ll care about in code):</strong></p>
<ul>
<li>
<code>a_ij[i, j]</code> = <strong>invader Ã— resident</strong><br>
</li>
<li>
<code>K_env[s, j]</code> = <strong>sites Ã— residents</strong><br>
</li>
<li>
<code>Nstar[j, s]</code> = <strong>residents Ã— sites</strong><br>
</li>
<li>
<code>prediction_matrix[s, i]</code> = <strong>sites Ã— species</strong>; in <code>fitness$r_mat</code> we use <strong>invader Ã— site</strong> ordering: <code>r_mat[i, s]</code>.</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="id_15-references">15. References<a class="anchor" aria-label="anchor" href="#id_15-references"></a>
</h2>
<ul>
<li>Hui C, Richardson DM (2017). Invasion Dynamics. Oxford University Press.</li>
<li>Hui C, Richardson DM, Landi P. et al.Â (2016). Defining invasiveness and invasibility in ecological networks. Biological Invasions.</li>
<li>Hui C et al.Â (2021). Trait-mediated ecological networks: mechanisms and consequences of invasion. Trends in Ecology &amp; Evolution.</li>
<li><em>Hui C et al.Â (2023). [Title of relevant work, update as needed.]</em></li>
</ul>
<p>This RMarkdown document implements a reproducible, extensible workflow for trait-based invasion ecology, integrating best practices for data wrangling, statistical modeling, and scientific visualization. Please cite the original data and framework sources as appropriate.</p>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/macSands/invasimap/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/macSands/invasimap/issues" class="external-link">Report a bug</a></li>
<li><a href="https://github.com/macSands/invasimap" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/macSands/invasimap/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing invasimap</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>First Last <br><small class="roles"> Author, maintainer </small>   </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/macSands/invasimap/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/macSands/invasimap/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://github.com/macSands/invasimap/actions/workflows/test-coverage.yaml" class="external-link"><img src="https://github.com/macSands/invasimap/actions/workflows/test-coverage.yaml/badge.svg" alt="test-coverage"></a></li>
<li><a href="https://app.codecov.io/gh/macSands/invasimap" class="external-link"><img src="https://codecov.io/gh/macSands/invasimap/graph/badge.svg" alt="codecov"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#stable" class="external-link"><img src="https://img.shields.io/badge/lifecycle-stable-brightgreen.svg" alt="lifecycle"></a></li>
<li><a href="LICENSE.html"><img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="License: MIT"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by First Last.</p>
</div>

<div class="pkgdown-footer-right">
  <p>NULL NULL</p>
</div>

    </footer>
</div>





  </body>
</html>
